---
title: "Chicago"
output: html_notebook
editor_options: 
  markdown: 
    wrap: 72
---


```{r}
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("EnsDb.Mmusculus.v79")
#install.packages("stringr",lib='~/R/x86_64-pc-linux-gnu-library/4.2')
setwd("~/data/chicago_R/chicago")
#save.image(file = "~/enh_capC/chicago.RData")

load("~/enh_capC/chicago.RData")
 
library(stringr,lib='~/R/x86_64-pc-linux-gnu-library/4.2')
library(tidyverse,lib='~/R/x86_64-pc-linux-gnu-library/4.2')
library(data.table)
library(cowplot)
library(scales)
library(zoo)
#library(ggbio) something wrong with this package
#library(ensembldb)
#library(GenomicFeatures)
library(ggrepel)
library(rstatix)
library(ggpubr)
library(ggplotify)
library(patchwork)

library(multidplyr)
library(EnsDb.Mmusculus.v79)
library(gggenes) 

squish_trans <- function(from, to, factor) {
  trans <- function(x) {
    if (any(is.na(x))) return(x)
    # get indices for the relevant regions
    isq <- x > from & x < to
    ito <- x >= to
    # apply transformation
    x[isq] <- from + (x[isq] - from)/factor
    x[ito] <- from + (to - from)/factor + (x[ito] - to)
    return(x)
  }

  inv <- function(x) {
    if (any(is.na(x))) return(x)
    # get indices for the relevant regions
    isq <- x > from & x < from + (to - from)/factor
    ito <- x >= from + (to - from)/factor
    # apply transformation
    x[isq] <- from + (x[isq] - from) * factor
    x[ito] <- to + (x[ito] - (from + (to - from)/factor))
    return(x)
  }
  # return the transformation
  return(trans_new("squished", trans, inv))
}

options(scipen=999) 

```

Find the baits that overlap with promoters, treated them as promoter-capture-Hic

```{r}
outputRds <- file.path("~/enh_capC/track_hub/chicago_round3/chicdiff/all_sample_results.Rds")
countputRds <- file.path("~/enh_capC/track_hub/chicago_round3/chicdiff/all_sample_countput.Rds")
bmapRds <- file.path("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.baitmap")
bmap <- read_tsv(bmapRds,col_names = c("chr","start","end","baitID","name")) %>% mutate(baitID_mid=as.integer((start+end)/2))
rmap <- read_tsv(paste0("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap"),col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2)) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid)

tad <- read_tsv("../20kb-res.tads.mESC_zhan.mm10.bed",col_names = c("chr","start","end")) %>% rownames_to_column("id") %>% as.data.table()
setkey(tad,chr,start,end)

#remove Gm genes, from 205 -> 165 baits overlap with promoters
prom <- read_tsv("ensembl_promoter_target_gene.bed",col_names = c("chr","start","end","gene")) %>% as.data.table()
setkey(prom,chr,start,end)

#use all the RNA-seq data from E10.5 to E12.5, if genes express in any tissues in these stages, then keep these promoter
files <- list.files(path="~/enh_capC/encode/RNA-seq/", pattern="tsv$", recursive=T) 
rm(dataset)
for (path in files){
  name=str_split(path, "_",simplify = T)
  if (length(name)==3) {
    temp_dataset <- read_tsv(paste0("~/enh_capC/encode/RNA-seq/",path),col_types = "ccddddddddddddddd") %>% mutate(tissue = name[1,1], rep=name[1,3], stage=name[1,2]) %>% dplyr::filter(grepl("ENS",gene_id)) %>% dplyr::select(gene_id,expected_count,TPM,FPKM,tissue,stage,rep)
    if (!exists("dataset")) {
      dataset <- temp_dataset 
    } else {
      dataset <- temp_dataset %>% bind_rows(dataset)
    }
    rm(temp_dataset)
  }
}

rna_seq <- dataset %>% group_by(gene_id,tissue,stage) %>% partition(new_cluster(10)) %>% summarise(mean_TPM=mean(TPM)) %>% collect() %>% dplyr::filter(mean_TPM>0.5) %>% dplyr::filter(stage %in% c("E105","E115","E125") ) %>% separate(gene_id,c("gene",NULL),sep="[^[:alnum:]]+") %>% left_join(read_tsv("gene_ensID_vs_name.tsv",col_names = c("gene","gene_name")), by="gene") %>% distinct(gene_name,stage,tissue,mean_TPM)
#dplyr::filter(rna_seq,gene_name=="Gm17750")

#the problem is some micro RNA like mir9 is not in the RNA-seq data, but other non-coding RNAs present (like Gm33366,Gm26803,C130071C03Rik)
expr_prom <- prom %>% mutate(len=str_count(gene,",")) %>% arrange(desc(len)) %>% separate(gene,paste0("gene",seq(1,13)),sep=",") %>% pivot_longer(!c(chr,start,end,len),names_to = "name",values_to = "value") %>% dplyr::filter(!is.na(value)) %>%
  left_join(group_by(rna_seq,gene_name) %>% summarise(TPM=mean(mean_TPM)),by=c("value"="gene_name")) %>% dplyr::filter(!is.na(TPM)) %>% group_by(chr,start,end) %>% summarise(genes = toString(value)) %>% ungroup() %>% as.data.table()
#expr_prom %>% dplyr::filter(grepl("Mir9|Gm33366|C130071C03",genes))
setkey(expr_prom,chr,start,end)
write_tsv(expr_prom,"ensembl_e10-12_expressing_promoter_target_gene.bed")

#176 prom bait regions, 5kb around promoters
prom_bait <- bmap %>% as.data.table() %>% mutate(start=start-5000,end=end+5000) %>% 
  foverlaps(expr_prom,type = "any") %>% dplyr::select(genes,bait_name=name,chr,start,end) %>% #mutate(num=str_count(genes,",")) %>% arrange(desc(num))
  separate(genes,paste0("gene",seq(1,10)),sep=",") %>% pivot_longer(!c(bait_name,chr,start,end),names_to = "num",values_to = "gene") %>% distinct(gene,bait_name,.keep_all = T) %>% dplyr::filter(!is.na(gene)) %>% group_by(bait_name,chr) %>% summarise(genes = toString(gene),start=min(start),end=max(end)) %>% dplyr::select(chr,start,end,bait_name,genes)
#dplyr::filter(prom_bait,bait_name=="hs268")

neg_bait <- read_tsv("../GATC.frags.pos.merged.tiles.mappability.anno.repeats.tsv") %>% distinct(elementID,class,tissue) %>% dplyr::filter((class=="NEG") & ! (elementID %in% prom_bait$bait_name)) %>% select(bait_name=elementID,class)


#931 pos enh bait regions
pos_enh_bait <- bmap %>% dplyr::filter(! (name %in% c(prom_bait$bait_name, neg_bait$bait_name))) %>% dplyr::select(bait_name=name,everything())

#after removing h3k27ac overlapping elements, only 87 left
#remove negative elements that overlap with E11.5 h3k27ac peaks
h3k27ac <- read_tsv("~/enh_capC/encode/E115_H3K27ac_peaks.bed",col_names = c("chr","start","end")) %>% as.data.table()
setkey(h3k27ac,chr,start,end)

neg_bait <- left_join(neg_bait,bmap,by=c("bait_name"="name")) %>% as.data.table() %>% foverlaps(h3k27ac, type="any") %>% dplyr::filter(is.na(start)) %>% dplyr::select(bait_name,chr,start=i.start,end=i.end,baitID,baitID_mid) %>% distinct()

prom_bait %>% arrange(bait_name)
pos_enh_bait %>% arrange(bait_name)
neg_bait %>% arrange(bait_name)

tis_corr <- data.frame(tissue=c("branchialarch","facialmesenchyme","forebrain","heart","hindbrain(rhombencephalon)","limb","limb","midbrain(mesencephalon)","neuraltube","somite","tail"),new=c("CF","CF","FB","HR","HB","FL","HL","MB","NT","TK","TL"))

#add new bait activity to regions that overlap multiple elements
bait_enh_act <- read_tsv("../GATC.frags.pos.merged.tiles.mappability.anno.repeats.tsv") %>% distinct(elementID,class,tissue) %>% bind_rows(dplyr::select(data.frame(elementID=c("hs2094","hs342","hs472","hs654","hs935"),class=c("POS","POS","POS","POS","POS"),alt_ID=c("mm1532","hs1523","hs461","hs1203","hs1032"),tissue=c("cranialnerve;hindbrain(rhombencephalon);","forebrain;midbrain(mesencephalon);","eye;","hindbrain(rhombencephalon);neuraltube;","forebrain;midbrain(mesencephalon);")),elementID,class,tissue)) %>% 
  mutate(len=str_count(tissue,";")) %>% arrange(desc(len)) %>% separate(tissue, paste0("tis_",seq(1:10)),sep=";") %>% pivot_longer(!c(elementID,class,len), names_to = "name", values_to = "tissue") %>% left_join(tis_corr,by="tissue") %>% dplyr::select(elementID,class,tissue=new) %>% dplyr::filter(!is.na(tissue)) %>% distinct(elementID,class,tissue) %>% dplyr::filter(elementID %in% bmap$name)
#test <- bait_enh_act %>% left_join(bmap,by=c("elementID"="name")) %>% dplyr::select(elementID,tissue,chr,start,end)
#write_tsv(test,"vista_enhancer_mm10_activity.bed")

#how many negative baits in promoter category
read_tsv("../GATC.frags.pos.merged.tiles.mappability.anno.repeats.tsv") %>% distinct(elementID,class,tissue) %>% dplyr::filter((class=="NEG") & (elementID %in% prom_bait$bait_name))
```

get the ditag peaks
```{r}
#read read_counts
countput_coord <- readRDS(countputRds) %>% left_join(rmap,by=c("otherEndID"),copy=T) %>% dplyr::select(-oeID_mid.x,oeID_mid=oeID_mid.y)
scale1 <- countput_coord %>% left_join(bmap,by="baitID") %>% dplyr::filter(abs(oeID_mid-baitID_mid) < 10000 & chr==oe_chr) %>% group_by(baitID,condition) %>% summarise(tot_read=sum(Nav)) %>% mutate(scale=1000/tot_read)
count_scaled <- countput_coord %>% left_join(bmap,by="baitID") %>% left_join(scale1,by=c("baitID","condition")) %>% mutate(scaled_Nav=Nav*scale)
rm(countput_coord)

#get adj_ditag peaks
files <- list.files(path="~/enh_capC/track_hub/chicago_round3/test/adj_ditag/", pattern="_adjcent_count.txt$", recursive=T) %>% .[!grepl("/",.)]
rm(ditag_peaks)
for (path in files){
  name=str_split(path,"_",simplify = T)
  print(name[1])
  temp_dataset <- read_table(paste0("~/enh_capC/track_hub/chicago_round3/test/adj_ditag/",path),col_names = c("ID","bait","oe","name","un","score","read_count"),col_types = cols(ID = col_character())) %>% add_count(bait,oe) %>% dplyr::filter(n>1) %>% separate(oe,c("oe_chr","oe_start","oe_end"),sep=",",convert=T) %>% mutate(oeID_mid=as.integer((oe_start+oe_end)/2)) %>% left_join(rmap,by=c("oeID_mid","oe_chr")) %>% distinct(name,oe_chr,oeID_mid,otherEndID,score) %>% add_column(condition=name[1]) 
  if (!exists("ditag_peaks")) {
    ditag_peaks <- temp_dataset 
  } else {
    ditag_peaks <- bind_rows(ditag_peaks,temp_dataset)
  }
}
ditag_peaks %>% dplyr::count(condition,name) %>% arrange(desc(n))

ditag_peaks %>% distinct(name) %>% mutate(bait_type=if_else(name %in% pos_enh_bait$bait_name, "enh", if_else(name %in% prom_bait$bait_name, "pro", "neg"))) %>% count(bait_type)

bmap %>% mutate(bait_type=if_else(name %in% pos_enh_bait$bait_name, "enh", if_else(name %in% prom_bait$bait_name, "pro", "neg"))) %>% count(bait_type)


rmap <- read_tsv(paste0("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap"),col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2)) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid)

#to create ucsc track 
test <- count_scaled %>% dplyr::select(name,condition,oe_chr,oeID_mid,otherEndID,scaled_Nav) %>%
  mutate(scaled_Nav=replace(scaled_Nav,(otherEndID %in% bmap$baitID),0)) %>% #remove bait-to-bait reads, 197,507 rows
  left_join(full_rmap,by=c("oe_chr"="chr","otherEndID"="id","oeID_mid"))

 
for (i in bmap$name) {
  write_tsv(dplyr::filter(test, name==i),paste0("~/enh_capC/track_hub/chicago_round3/merged_bam/chic_output/",i,"_chicdiff_output_across_10_tissues.tsv"))
}

bmap %>% dplyr::filter(name=="hs698")
#there's no viewpoint for hs1192, maybe because there's only 2 baits on it, has few reads on this viewpoint
count_scaled %>% dplyr::filter(name=="hs1192") 
```

Supp fig1 supfig1a capture rates

```{r}
data1 <- read_tsv("~/enh_capC/track_hub/chicago_round3/merged_bam/capture_rate_count.txt",col_names = c("name","read_count")) %>% mutate(on_bait=rep(c("yes","no"),nrow(.)/2),tis=word(name, 1, sep = '/')) %>% arrange(tis) %>% group_by(tis,on_bait) %>% dplyr::filter(read_count==max(read_count)) %>% distinct(tis,on_bait,read_count) %>%
  pivot_wider(names_from = on_bait, values_from=read_count) %>% mutate(tot=yes+no,cap_rate=formatC(yes/tot*100, format = "f", digits = 1) ) %>% 
  dplyr::filter(! tis %in% c("HL3","HRT3","LV3","LZ-LB1","LZ-LB2","NT2","NT3")) 

supfig1a <- data1 %>% ggplot(aes(x=tis,y=yes/1000000)) + geom_bar(stat="identity", position = "dodge") + geom_text(aes(label=paste0(cap_rate,"%")),position=position_stack(1),angle = 45,vjust=-.8,hjust=0,size=2) + theme_bw() + scale_y_continuous(limits = c(0,18)) + labs(y="Number of valid on-bait ditags (millions)",x="") + theme(axis.text.x = element_text(angle = 45,vjust=0.5,hjust=0.5,size=6),
                     axis.title=element_text(size=7),axis.text.y = element_text(size=6))
supfig1a

data1 %>% ggplot(aes(x=tis,y=tot/1000000)) + geom_bar(stat="identity", position = "dodge") + theme_bw() + labs(y="Number of valid ditags (millions)",x="", title="All valid ditags") + theme(axis.text.x = element_text(angle = 45,vjust=0.5,hjust=0.5,size=6), axis.title=element_text(size=7),axis.text.y = element_text(size=6))
```

supfig1b&c PCA for replicates

```{r}
library(factoextra)  #for PCA visualization
library(embed)
library(tidymodels)

files <- list.files(path="~/enh_capC/track_hub/chicago_round3/test/adj_ditag/replicates", pattern="_adjcent_count.txt$")
rm(dataset)
for (path in files){
  name=str_split(path,"_",simplify = T)
  print(name[1])
  temp_dataset <- read_tsv(paste0("~/enh_capC/track_hub/chicago_round3/test/adj_ditag/replicates/",path)) %>% add_count(bait_coord,oe_coord) %>% dplyr::filter(n>1) %>% distinct(bait_name,bait_coord,oe_coord,score)
  cn <- colnames(temp_dataset) %>% str_replace("score",name[1])
  temp_dataset <- magrittr::set_colnames(temp_dataset,cn)
  if (!exists("dataset")) {
    dataset <- temp_dataset 
  } else {
    dataset <- full_join(dataset,temp_dataset,by=c("bait_coord","bait_name","oe_coord"))
  }
}

#mat chicago score
mat <- dataset %>% mutate(across(CF3:TL4, ~tidyr::replace_na(.,0))) %>% 
  mutate(across(CF3:TL4, ~replace(.,.<=5,0 ))) %>% mutate(across(CF3:TL4, ~replace(.,.>5,1))) %>% 
  separate(bait_coord, c("bc", "bs","be"),sep=",",convert = TRUE) %>% separate(oe_coord, c("oc", "os","oe"),sep=",",convert = TRUE) %>% mutate(dist=abs((bs+be)/2-(os+oe)/2)) %>% dplyr::filter(dist<=2000000 & bc==oc) %>% dplyr::select(-c(bc:oe,dist)) %>% #only include cis interaction within 2mb
  dplyr::filter(rowSums(.)>=2 & rowSums(.)<=18) %>% #filter peaks that unique to 1 tissue
  rownames_to_column() %>% pivot_longer(!rowname,"tissue","value") %>% pivot_wider(names_from = rowname,values_from = value) %>% column_to_rownames("tissue")

#first 2 components
#can not use 
pca_mat <- mat %>% prcomp()  #no need to scale since is a 0,1 matrix
p1 <- fviz_pca_ind(pca_mat,
             col.ind = "cos2", # Color by the quality of representation
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
             )

test <- pca_mat$x %>% as_tibble() %>% mutate(name=rownames(pca_mat$x)) %>% dplyr::select(name, paste0("PC",seq(1,3)))  #try just use top n components(top8,), 21 comps in total


umap_rec <- recipe(~., data = test) %>%
  update_role(name, new_role = "id") %>%
  #step_normalize(all_predictors()) %>% after PCA analysis, no normalization is needed
  step_umap(all_predictors(),neighbors = 3)

umap_prep <- prep(umap_rec)
supfig1b <- juice(umap_prep) %>% mutate(tissue=str_remove(name,"[0-9]")) %>%
  ggplot(aes(x=UMAP1, y=UMAP2, label = name,color=tissue)) +
  geom_point(alpha = 0.7, size = 1) +
  geom_text_repel(size = 2,max.overlaps = Inf) + theme_bw() +
  theme(legend.position = "none",axis.text = element_text(size=6),
                     axis.title=element_text(size=7))

supfig1b
```

use peaks present or not and merge peaks less than 5kb away

```{r}
test <- dataset %>% separate(oe_coord, c("oc", "os","oe"),sep=",",convert = TRUE) %>% distinct(bait_name,oc,os,oe) %>% arrange(bait_name,oc,os,oe) %>% group_by(bait_name,oc) %>% mutate(dist=lead(os)-oe,num=if_else(dist>5000 | is.na(dist) | row_number()==1 ,1,0)) %>% ungroup() %>% mutate(gr=cumsum(num)) %>% distinct(bait_name,oc,os,oe,gr) #make group

mat <- dataset %>% row_count(., count = NA) %>% dplyr::filter(rowcount<19) %>% #remove peaks only appear in one samples
  separate(bait_coord, c("bc", "bs","be"),sep=",",convert = TRUE) %>% separate(oe_coord, c("oc", "os","oe"),sep=",",convert = TRUE) %>% 
  mutate(dist=abs((bs+be)/2-(os+oe)/2)) %>% dplyr::filter(dist<=2000000 & bc==oc) %>% #only include cis interaction within 2mb interactions
  mutate(across(CF3:TL4, ~tidyr::replace_na(.,0))) %>% 
  left_join(test,by=c("bait_name","oc", "os","oe")) %>% group_by(gr) %>% summarise(across(CF3:TL4, sum)) %>%
  mutate(across(CF3:TL4, ~replace(.,.<=5,0 ))) %>% mutate(across(CF3:TL4, ~replace(.,.>5,1))) %>%
  pivot_longer(!gr,"tissue","value") %>% pivot_wider(names_from = gr,values_from = value) %>% column_to_rownames("tissue")

M<-cor(t(mat) %>% as.tibble() %>% dplyr::select(FB2,FB4,MB2,MB4,HB5,HB6,NT6,NT7,
                                                HR5,HR6,
                                                FL2,FL3,HL4,HL5,CF3,CF4,TL3,TL4,TK1,TK2))

library(corrplot)

pdf(file = "../manuscript/Supp_fig1c.pdf",width=9/2.54,height = 6/2.54)
corrplot(M, method="color",tl.col="black",#addCoef.col = "black",
         diag=FALSE,#type="upper",
         col = rev(COL2('RdBu')),
         order="original", 
         #order="hclust", 
         addrect = 3,
         #col=COL1('YlOrRd'),
         #col.lim = c(0, 0.8),
         tl.cex = 0.5,cl.cex = 0.4,
         is.corr = FALSE
         ) 

dev.off()

```

fraction of replicatd interactions

```{r}
mat %>% rownames_to_column() %>% pivot_longer(!rowname, names_to = "peak", values_to = "score") %>% mutate(tis=str_extract(rowname, "[[:alpha:]]+"),score=replace(score,score>0, 1)) %>% group_by(tis,peak) %>% summarise(sum=sum(score)) %>% ungroup() %>% arrange(desc(sum)) %>% dplyr::filter(sum>0) %>% mutate(sum=factor(sum)) %>% dplyr::count(tis,sum) %>% ggplot(aes(x=tis,y=n,fill=sum)) + geom_bar(stat="identity") + scale_fill_manual(values=c("#4d77b1e8","#ec7f26e3"),labels = c("1"="unique", "2"="both"),name="replicates") + labs(title = ">=2 samples, 2mb cis, merge peaks within 5kb")

```

Fig1D. calculate the proportion of ep, ctcf, ee, polycomb interactions that within the same TAD or not For the first counting, just count the fragment-to-fragment interactions

```{r}
#just keep the pos_enh_bait and expressing genes
#account for alternative promoters
ep_exp <- read_tsv("~/ensembl_regulatory/chicago3_ditag_promoter_target_gene_preliminary_result.tsv") %>% mutate(target_mid=as.integer((target_start+target_end)/2)) %>% left_join(rmap,by=c("target_chr"="oe_chr","target_mid"="oeID_mid")) %>% group_by(bait_name,tissue,otherEndID) %>% mutate(dist=abs((promoter_start+promoter_end)/2-target_mid), overlap_len=if_else(target_end>=promoter_start & target_end<=promoter_end, target_end-promoter_start+1, if_else(target_start>=promoter_start & target_start<=promoter_end, promoter_end-target_start+1,0))) %>% dplyr::filter(overlap_len==max(overlap_len)) %>% dplyr::filter(dist==min(dist)) %>% ungroup() %>% ###add this step to filter out targets that overlap with two promoters and choose the one that closest to the target_mid
  unite("promoter",starts_with("promoter")) %>% #group them by promoter coordinate
  mutate(bait_type=if_else(bait_name %in% pos_enh_bait$bait_name, "enh", if_else(bait_name %in% prom_bait$bait_name, "pro", if_else(bait_name %in% neg_bait$bait_name, "neg","non"))))

ep_exp_5kb <- read_tsv("~/ensembl_regulatory/chicago3_ditag_promoter_target_gene_preliminary_result_5kb.tsv") %>% mutate(target_mid=as.integer((target_start+target_end)/2)) %>% left_join(rmap,by=c("target_chr"="oe_chr","target_mid"="oeID_mid")) %>% group_by(bait_name,tissue,otherEndID) %>% mutate(promoter_start=promoter_start+5000,promoter_end=promoter_end-5000) %>%
  mutate(dist=abs((promoter_start+promoter_end)/2-target_mid), overlap_len=if_else(target_end>=promoter_start & target_end<=promoter_end, target_end-promoter_start+1, if_else(target_start>=promoter_start & target_start<=promoter_end, promoter_end-target_start+1,0))) %>% dplyr::filter(overlap_len==max(overlap_len)) %>% dplyr::filter(dist==min(dist)) %>% ungroup() %>% ###add this step to filter out targets that overlap with two promoters and choose the one that closest to the target_mid
  mutate(promoter_start=promoter_start-5000,promoter_end=promoter_end+5000) %>%
  unite("promoter",starts_with("promoter")) %>% #group them by promoter coordinate
  mutate(bait_type=if_else(bait_name %in% pos_enh_bait$bait_name, "enh", if_else(bait_name %in% prom_bait$bait_name, "pro", if_else(bait_name %in% neg_bait$bait_name, "neg","non"))))

chic3_CTCF <- read_tsv("~/enh_capC/encode/chicago3_ctcf.tsv",col_names = c("target_chr","target_start","target_end","bait_chr","bait_start","bait_end","bait_name","score","tissue")) %>%   mutate(bait_type=if_else(bait_name %in% pos_enh_bait$bait_name, "enh", if_else(bait_name %in% prom_bait$bait_name, "pro", if_else(bait_name %in% neg_bait$bait_name, "neg","non"))))

chic3_CTCF_5kb <- read_tsv("~/enh_capC/encode/chicago3_ctcf_5kb.tsv",col_names = c("target_chr","target_start","target_end","bait_chr","bait_start","bait_end","bait_name","score","tissue")) %>%   mutate(bait_type=if_else(bait_name %in% pos_enh_bait$bait_name, "enh", if_else(bait_name %in% prom_bait$bait_name, "pro", if_else(bait_name %in% neg_bait$bait_name, "neg","non"))))


#through E10.5~E12.5
ee_CTCF_e10_12 <- read_tsv("~/enh_capC/encode/e10-12_chicago3_ditag_putative_enhancer_CTCF_removal.tsv") %>% unite(target,starts_with("target"),remove = F) %>% distinct(target,bait_name,score,tissue,.keep_all = T) %>% 
  mutate(bait_type=if_else(bait_name %in% pos_enh_bait$bait_name, "enh", if_else(bait_name %in% prom_bait$bait_name, "pro", if_else(bait_name %in% neg_bait$bait_name, "neg","non"))))

ee_CTCF_e10_12_5kb <- read_tsv("~/enh_capC/encode/e10-12_chicago3_ditag_putative_enhancer_CTCF_removal_5kb.tsv") %>% unite(target,starts_with("target"),remove = F) %>% distinct(target,bait_name,score,tissue,tissue,.keep_all = T) %>% 
  mutate(bait_type=if_else(bait_name %in% pos_enh_bait$bait_name, "enh", if_else(bait_name %in% prom_bait$bait_name, "pro", if_else(bait_name %in% neg_bait$bait_name, "neg","non"))))


#through E10.5~E12.5
chic3_pc_e10_12 <- read_tsv("~/enh_capC/encode/e10-12_chicago3_ditag_polycomb.tsv",col_names = c("target_chr","target_start","target_end","bait_chr","bait_start","bait_end","bait_name","score","tissue"))  %>%   mutate(bait_type=if_else(bait_name %in% pos_enh_bait$bait_name, "enh", if_else(bait_name %in% prom_bait$bait_name, "pro", if_else(bait_name %in% neg_bait$bait_name, "neg","non"))))
chic3_pc_e10_12_5kb  <- read_tsv("~/enh_capC/encode/e10-12_chicago3_ditag_polycomb_5kb.tsv",col_names = c("target_chr","target_start","target_end","bait_chr","bait_start","bait_end","bait_name","score","tissue"))  %>%   mutate(bait_type=if_else(bait_name %in% pos_enh_bait$bait_name, "enh", if_else(bait_name %in% prom_bait$bait_name, "pro", if_else(bait_name %in% neg_bait$bait_name, "neg","non"))))


#count number of interactions of frag-to-frag across or within TAD, here doesn't account for promoter or h3k27ac peaks
data1 <- ep_exp %>% distinct(bait_name,promoter,target_chr,target_start,target_end,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% left_join(dplyr::select(bmap,target_chr=chr,b_start=start,b_end=end,bait_name=name),by=c("target_chr","bait_name")) %>% foverlaps(tad, by.x=c("target_chr","b_start", "b_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>%
  dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>%  #first remove interactions that not have tad annotation and preferentially select within_tad using filter(row_number()==n())
  dplyr::count(tad,bait_type) %>% add_column(type="ep_1bp")

data1 <- ep_exp_5kb %>% distinct(bait_name,promoter,target_chr,target_start,target_end,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% left_join(dplyr::select(bmap,target_chr=chr,b_start=start,b_end=end,bait_name=name),by=c("target_chr","bait_name")) %>% foverlaps(tad, by.x=c("target_chr","b_start", "b_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>%
  dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>%  #first remove interactions that not have tad annotation and preferentially select within_tad using filter(row_number()==n())
  dplyr::count(tad,bait_type) %>% add_column(type="ep_5kb") %>% bind_rows(data1)

data1 <- ee_CTCF_e10_12 %>% distinct(target_chr,target_start,target_end,bait_chr,bait_start,bait_end,bait_name,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% foverlaps(tad, by.x=c("bait_chr","bait_start", "bait_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>% 
    dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>%  #first remove interactions that not have tad annotation and get unique tad annotation for each ep pair
dplyr::count(tad,bait_type) %>% add_column(type="ee_1bp") %>% bind_rows(data1)

data1 <- ee_CTCF_e10_12_5kb %>% distinct(target_chr,target_start,target_end,bait_chr,bait_start,bait_end,bait_name,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% foverlaps(tad, by.x=c("bait_chr","bait_start", "bait_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>% 
    dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>%  #first remove interactions that not have tad annotation and get unique tad annotation for each ep pair
dplyr::count(tad,bait_type) %>% add_column(type="ee_5kb") %>% bind_rows(data1)

data1 <- chic3_CTCF %>% distinct(target_chr,target_start,target_end,bait_chr,bait_start,bait_end,bait_name,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% foverlaps(tad, by.x=c("bait_chr","bait_start", "bait_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>%
  dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>%  #first remove interactions that not have tad annotation and preferentially select within_tad using filter(row_number()==n())
  mutate(bound=if_else(abs(i.start - (target_start+target_end)/2) < 10000, "yes", if_else(abs(i.end - (target_start+target_end)/2) < 10000,"yes","no"))) %>% dplyr::count(tad,bait_type,bound) %>% mutate(type=if_else(bound=="yes","boundary_ctcf_1bp","within_ctcf_1bp")) %>% dplyr::select(-bound) %>% bind_rows(data1)

data1 <- chic3_CTCF_5kb %>% distinct(target_chr,target_start,target_end,bait_chr,bait_start,bait_end,bait_name,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% foverlaps(tad, by.x=c("bait_chr","bait_start", "bait_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>%
  dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>% #first remove interactions that not have tad annotation and preferentially select within_tad using filter(row_number()==n())
  mutate(bound=if_else(abs(i.start - (target_start+target_end)/2) < 10000, "yes", if_else(abs(i.end - (target_start+target_end)/2) < 10000,"yes","no"))) %>% dplyr::count(tad,bait_type,bound) %>% mutate(type=if_else(bound=="yes","boundary_ctcf_5kb","within_ctcf_5kb")) %>% dplyr::select(-bound) %>% bind_rows(data1)

data1 <- chic3_pc_e10_12 %>% distinct(target_chr,target_start,target_end,bait_chr,bait_start,bait_end,bait_name,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% foverlaps(tad, by.x=c("bait_chr","bait_start", "bait_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>% 
    dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>%  #first remove interactions that not have tad annotation and get unique tad annotation for each ep pair
dplyr::count(tad,bait_type) %>% add_column(type="pc_1bp") %>% bind_rows(data1)

data1 <- chic3_pc_e10_12_5kb %>% distinct(target_chr,target_start,target_end,bait_chr,bait_start,bait_end,bait_name,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% foverlaps(tad, by.x=c("bait_chr","bait_start", "bait_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>% 
    dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>%  #first remove interactions that not have tad annotation and get unique tad annotation for each ep pair
dplyr::count(tad,bait_type) %>% add_column(type="pc_5kb") %>% bind_rows(data1)

supfig1_data <- data1
ord <- distinct(supfig1_data,type) %>% dplyr::filter(grepl("5kb",type)) %>% .$type %>% rev()
supfig1_data %>% dplyr::filter(grepl("5kb",type)) %>% mutate(type=factor(type,levels = ord)) %>% ggplot(aes(x=type,y=n,fill=tad,group=bait_type)) + geom_bar(stat="identity") + facet_wrap(~ bait_type, scales="free_y") + scale_x_discrete(labels=c("Prom","Enh","boundary_CTCF","within_tad_CTCF","Polycomb","left")) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) #actual number

p1 <- supfig1_data %>% dplyr::filter(grepl("5kb",type)) %>% dplyr::filter(bait_type=="enh") %>% dplyr::count(tad,wt=n) %>% mutate(perc=n/sum(n)) %>% ggplot(aes(x="",y=perc,fill=tad)) + geom_bar(width = 1, stat = "identity", color = "white") + geom_text(aes(label=n),position=position_stack(0.5),size=3) + coord_polar("y", start=0) + theme_void() + scale_fill_manual(values=c("#ec7f26e3","#4d77b1e8"))
ggsave("../manuscript/version_4_figure/fig1d.pdf",width = 3, height = 2)

supfig1_data %>% dplyr::filter(grepl("5kb",type)) %>% mutate(type=factor(type,levels = ord)) %>% group_by(bait_type,type) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% ggplot(aes(x=type,y=perc,fill=tad,group=bait_type)) + geom_bar(stat="identity") + facet_wrap(~ bait_type, scales="free_y") + scale_x_discrete(labels=c("Prom","Enh","boundary_CTCF","within_tad_CTCF","Polycomb","left")) + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) #percent

supfig1_data %>% dplyr::filter(grepl("5kb",type)) %>% mutate(type=factor(type,levels = ord)) %>% group_by(bait_type,tad) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% arrange(tad,bait_type) %>%
  ggplot(aes(x=tad,y=perc,fill=type,group=bait_type)) + geom_bar(stat="identity") + facet_wrap(~ bait_type, scales="free_y") + scale_fill_discrete(labels=c("Prom","Enh","boundary_CTCF","within_tad_CTCF","Polycomb","left"))  #percent

supfig1_data %>% dplyr::filter(grepl("5kb",type)) %>% mutate(type=factor(type,levels = ord)) %>% group_by(bait_type,tad) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% arrange(tad,bait_type) %>% dplyr::filter(tad=="within" & bait_type=="enh")

supfig1_data %>% dplyr::filter(grepl("5kb",type)) %>% mutate(type=factor(type,levels = ord)) %>% group_by(bait_type,tad) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% arrange(tad,bait_type) %>% dplyr::filter(bait_type=="enh") %>% group_by(tad) %>% summarise(sum=sum(n)) %>% mutate(perc=sum/sum(sum))

ep_exp_5kb %>% distinct(promoter) #count promoter number

supfig1_data %>% dplyr::filter(grepl("5kb",type)) %>% mutate(type=factor(type,levels = ord)) %>% group_by(bait_type,tad) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% arrange(tad,bait_type) %>% dplyr::filter(tad=="within" & bait_type=="pro")

#write_tsv(ep_exp_5kb %>% left_join((separate(pos_pos_int,pair,into=c("bait_name","promoter"),extra = "merge") %>% mutate(golden_set="yes")), by=c("bait_name","promoter")),"ep_exp_5kb.tsv")
#write_tsv(chic3_CTCF_5kb,"chic3_CTCF_5kb.tsv")
#write_tsv(ee_CTCF_e10_12_5kb,"ee_CTCF_e10_12_5kb.tsv")
#write_tsv(chic3_pc_e10_12_5kb,"chic3_pc_e10_12_5kb.tsv")
```


count average interactions per bait #176 prom baits, 5kb around promoters #after removing those overlaping with h3k27ac is 87 neg baits #931 pos enh bait

```{r}

data2 <- ep_exp_5kb %>% distinct(bait_name,promoter,target_chr,target_start,target_end,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% left_join(dplyr::select(bmap,target_chr=chr,b_start=start,b_end=end,bait_name=name),by=c("target_chr","bait_name")) %>% foverlaps(tad, by.x=c("target_chr","b_start", "b_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>%
  dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>% dplyr::select(bait_name,bait_type,tad,starts_with("target")) %>% add_column(type="ep_5kb") 

data2 <- ee_CTCF_e10_12_5kb %>% distinct(target_chr,target_start,target_end,bait_chr,bait_start,bait_end,bait_name,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% foverlaps(tad, by.x=c("bait_chr","bait_start", "bait_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>% 
    dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>% dplyr::select(bait_name,bait_type,tad,starts_with("target")) %>% add_column(type="ee_5kb") %>% bind_rows(data2)

data2 <- chic3_CTCF_5kb %>% distinct(target_chr,target_start,target_end,bait_chr,bait_start,bait_end,bait_name,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% foverlaps(tad, by.x=c("bait_chr","bait_start", "bait_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>%
  dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>% #first remove interactions that not have tad annotation and preferentially select within_tad using filter(row_number()==n())
  mutate(bound=if_else(abs(i.start - (target_start+target_end)/2) < 10000, "yes", if_else(abs(i.end - (target_start+target_end)/2) < 10000,"yes","no"))) %>% dplyr::select(bait_name,bait_type,tad,bound,starts_with("target")) %>% mutate(type=if_else(bound=="yes","boundary_ctcf_5kb","within_ctcf_5kb")) %>% dplyr::select(-bound) %>% bind_rows(data2)

data2 <- chic3_pc_e10_12_5kb %>% distinct(target_chr,target_start,target_end,bait_chr,bait_start,bait_end,bait_name,bait_type) %>% as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% foverlaps(tad, by.x=c("bait_chr","bait_start", "bait_end"), type="any") %>% mutate(tad=if_else(id==i.id,"within","across")) %>% 
    dplyr::filter(!is.na(tad)) %>% group_by(bait_name,target_chr,target_start,target_end,bait_type) %>% arrange(tad) %>% dplyr::filter(row_number()==n()) %>% ungroup() %>% dplyr::select(bait_name,bait_type,tad,starts_with("target"))  %>% add_column(type="pc_5kb") %>% bind_rows(data2)


ord=c("ep_5kb","ee_5kb","within_ctcf_5kb","boundary_ctcf_5kb","pc_5kb")
supfig2_data <- dplyr::select(bmap,bait_name=name) %>% mutate(bait_type=if_else(bait_name %in% pos_enh_bait$bait_name, "enh", if_else(bait_name %in% prom_bait$bait_name, "pro", "neg"))) %>% bind_rows(dplyr::count(data2,bait_name,bait_type,tad,type)) %>% 
  complete(nesting(bait_name,bait_type),nesting(tad,type), fill = list(n = 0)) %>% dplyr::filter(tad=="within") %>% 
  dplyr::filter(bait_type != "non") %>%  
  mutate(type=factor(type,levels = ord), bait_type=factor(bait_type,levels = c("pro","enh","neg"))) %>% dplyr::select(bait_type,type,n) 

stat.test <- supfig2_data %>% mutate(bait_type2=if_else(bait_type=="neg","neg","active")) %>% group_by(type) %>% rstatix::wilcox_test(n ~ bait_type) %>% 
  adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>% add_x_position(x = "type")  %>% #separate the p values to x.axis
  mutate(p.adj=if_else(p.adj.signif=="ns","ns",formatC(p.adj, format = "g", digits = 3))) %>% dplyr::filter(p.adj != "ns") 

supfig1d <- supfig2_data %>%  ggbarplot(x = "type", y = "n", add = c("mean_se"),add.params = list(size = 0.3), size = 0.3, fill="bait_type", position = position_dodge(0.7),error.plot = "upper_errorbar") + 
  stat_pvalue_manual(stat.test, label = "p.adj",y.position=c(6.5,7.5,5.5,8.5,10,10.5,11.5,9),tip.length = 0.002,size=2) + #,10,11,8.5
  scale_fill_manual(values=c("#dd8581ff","#e9b86fff","#68abdaff"),name = "Bait type", labels = c("Promoter", "Positive enhancer", "Control element")) + scale_x_discrete(labels=c("Promoter","Enhancer","Within TAD CTCF","Boundary CTCF","Polycomb")) +
  labs(y="Average number of interactions per bait",x="Type of interacting regions") +
  theme_bw() + theme(axis.text.x = element_text(angle = 15, vjust = 1, hjust = 1,size=6),
                     axis.title=element_text(size=7),axis.text.y = element_text(size=6),
                     legend.key.size = unit(2, 'mm'), legend.title = element_text(size=6), legend.text = element_text(size=6), legend.position = "top") 
supfig1d
```

how many interactions are with 87 neg elements

```{r}
ord=c("ep_5kb","ee_5kb","within_ctcf_5kb","boundary_ctcf_5kb","pc_5kb")
supfig1h <- supfig1_data %>% dplyr::filter(bait_type != "non") %>%  
  mutate(type=factor(type,levels = ord), bait_type=factor(bait_type,levels = c("pro","enh","neg"))) %>% dplyr::filter(tad != "across") %>% 
  dplyr::filter(grepl("5kb",type)) %>% group_by(bait_type) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% ggplot(aes(x=bait_type,y=perc,fill=type)) + geom_bar(stat="identity",width = .5) + 
  scale_fill_manual(values=c("#dd8581ff","#e9b86fff","#77b27bff","#68abdaff","#9f9f9fff"),labels=c("Promoter","Enhancer","Within TAD CTCF","Boundary CTCF","Polycomb")) +
  scale_x_discrete(labels=c("Promoter", "Positive enhancer", "Control element"))  + 
  scale_y_continuous(labels = scales::percent) +
  labs(y="",x="Bait type") + geom_text(aes(label=n),position = position_stack(vjust = 0.5),size=2) +
  theme_bw() + theme(axis.title=element_text(size=7),axis.text = element_text(size=6), legend.key.size = unit(2, 'mm'),legend.title = element_text(size=6), legend.text = element_text(size=6)) #actual number
supfig1h
```

rna_seq_stage

```{r}
files <- list.files(path="~/enh_capC/encode/RNA-seq/", pattern="tsv$", recursive=T) 
rm(dataset)
for (path in files){
  name=str_split(path, "_",simplify = T)
  if (length(name)==3) {
    temp_dataset <- read_tsv(paste0("~/enh_capC/encode/RNA-seq/",path),col_types = "ccddddddddddddddd") %>% mutate(tissue = name[1,1], rep=name[1,3], stage=name[1,2]) %>% dplyr::filter(grepl("ENS",gene_id)) %>% dplyr::select(gene_id,expected_count,TPM,FPKM,tissue,stage,rep)
    if (!exists("dataset")) {
      dataset <- temp_dataset 
    } else {
      dataset <- temp_dataset %>% bind_rows(dataset)
    }
    rm(temp_dataset)
  }
}
rna_seq_stage <- dataset %>% group_by(gene_id,tissue,stage) %>% partition(new_cluster(10)) %>% summarise(mean_TPM=mean(TPM)) %>% collect() %>% separate(gene_id,c("gene",NULL),sep="[^[:alnum:]]+") %>% left_join(read_tsv("gene_ensID_vs_name.tsv",col_names = c("gene","gene_name","gene_chr","gene_start","gene_end")), by="gene") %>% group_by(gene_name,gene_chr,stage,tissue) %>% summarise(mean_TPM=mean(mean_TPM),gene_start=min(gene_start),gene_end=max(gene_end)) %>% ungroup()
rna_seq_stage %>% dplyr::filter(gene_name=="Dlx6os1")
```

how many of the control elements interact with the most proximal expressing genes

```{r}
#same tad??
#2.5kb up- and down-stream of TSS to overlap with target fragment 
refgene_tss <- read_tsv("~/ensembl_regulatory/refGene_all_TSS_2kb.txt",col_names = c("chr","tss_start","tss_end","strand","gene_name")) %>% as.data.table()
setkey(refgene_tss,chr,tss_start,tss_end)

tis_corr <- data.frame(tissue=c("embryonicfacialprominence","forebrain","heart","hindbrain","limb","limb","liver","midbrain","neuraltube"),new=c("CF","FB","HR","HB","FL","HL","LV","MB","NT"))
rna_seq_e115 <- left_join(rna_seq_stage,tis_corr,by="tissue") %>% dplyr::filter(stage=="E115") %>% dplyr::select(gene_name,mean_TPM,tissue=new)

neg_tss_exp <- ep_exp_5kb %>% dplyr::filter(bait_type %in% c("neg","enh")) %>% 
  dplyr::select(starts_with("target"),starts_with("bait"),promoter,gene,score) %>% as.data.table() %>%
  foverlaps(refgene_tss, by.x=c("target_chr","target_start", "target_end"), type="any") %>% dplyr::filter(!is.na(gene_name)) %>%
  left_join(rna_seq_e115,by="gene_name") %>% tidyr::replace_na(list(mean_TPM = 0)) %>%
  group_by(across(starts_with("tss")),across(starts_with("bait")),gene_name) %>% summarise(max_TPM=max(mean_TPM)) %>% ungroup() %>%
  dplyr::filter(max_TPM>0.5) %>%
  as.data.table()


intra_gene <- read_tsv("~/ensembl_regulatory/within_gene_enh.tsv",col_names = c("chr","bait_start","bait_end","bait_name","gene_chr","gene_start","gene_end","gene_name","bp")) %>% group_by(bait_name,gene_name) %>% summarise(gene_chr=last(gene_chr),gene_start=min(gene_start),gene_end=max(gene_end)) %>% ungroup() %>% distinct()

data <- neg_tss_exp %>% #dplyr::filter(bait_name=="hs654") %>%
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1)) %>%
  dplyr::select(-c(tss_start,tss_end)) %>% 
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>% distinct(bait_name,bait_type,i.gene_name,gene_name) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% #group_by(bait_name,i.gene_name) %>% 
  add_count(bait_name,i.gene_name) %>% mutate(n=replace(n, is.na(gene_name), 0)) %>% distinct(bait_name,n,i.gene_name,bait_type) %>%
  left_join(dplyr::filter(intra_gene, gene_name %in% refgene_tss$gene_name),by="bait_name") %>% 
  mutate(gr=if_else(n>=1,">=1 genes",if_else((i.gene_name!=gene_name | is.na(gene_name)),"nei", "intra"))) %>%
  group_by(bait_name) %>% dplyr::filter(n==min(n)) %>% #closest genes
  arrange(bait_name) %>% distinct(bait_name,gr,bait_type) %>% add_count(bait_name,name="num") %>%  dplyr::filter((num>1 & gr=="intra") | num==1) %>% ungroup() %>% #enhancer-centric
  mutate(gr=if_else(gr=="intra","nei",gr)) %>% #consider intragenic as neighboring
  dplyr::count(gr,bait_type)


#chi-square test 
pval <- data %>% pivot_wider(names_from = bait_type, values_from = n) %>% column_to_rownames("gr") %>% chisq.test() %>%  .$p.value %>% formatC(., format = "g", digits = 2)

supfig1k <- data %>% group_by(bait_type) %>% mutate(gr = factor(gr, levels = c("nei", ">=1 genes")), frac=round(n/sum(n), 1)) %>% ungroup() %>%
  ggplot(aes(x=bait_type,y=frac,group=gr,fill=gr)) + geom_bar(stat="identity", width = .4) +
  annotate("segment", x=1, xend=2, y=1.05, yend=1.05, colour="black", size=0.5) +
  annotate("segment", x=1, xend=1, y=1.02, yend=1.051, colour="black", size=0.5) +
  annotate("segment", x=2, xend=2, y=1.02, yend=1.051, colour="black", size=0.5) +
  scale_x_discrete(labels=c("enh"="Positive enhancer","neg"="Control element")) +
  scale_y_continuous(labels = scales::percent) +
  annotate("text", x=1.5, y=1.07, label=paste0("P=",pval),color="black",size=2.5) +
  geom_text(aes(label = n), position = position_stack(vjust = .5),size=2) + labs(x="Bait type",y="Proportion") + 
  scale_fill_manual(values=c("#dd8581ff","#e9b86fff","#77b27bff","#68abdaff"),name="", labels=c("Neighboring gene",expression("Skipping ">="1 gene"))) +
  theme_bw() + theme(axis.text.x = element_text(size=6),
                     axis.title=element_text(size=7),axis.text.y = element_text(size=6),
                     legend.position = "top",legend.key.size = unit(3, 'mm'),legend.text = element_text(size=6))
supfig1k
```

how control region interacting active or inactive genes regardless of their own activity

```{r}
neg_tss <- ep_exp_5kb %>% dplyr::filter(bait_type != "non") %>%  
  dplyr::select(starts_with("target"),starts_with("bait"),promoter,gene,score,tissue,otherEndID) %>%
  complete(nesting(target_chr,target_start,target_end,target_mid,bait_chr,bait_start,bait_end,bait_name,bait_type,promoter,gene,otherEndID),tissue,fill=list(score=0)) %>% #get full tissue list of interacting neg-gene
  as.data.table() %>%
  foverlaps(refgene_tss, by.x=c("target_chr","target_start", "target_end"), type="any") %>% dplyr::filter(!is.na(gene_name)) %>% 
  distinct(gene_name,bait_name,bait_type,otherEndID,tissue,score,target_chr,target_mid) %>%
  left_join(rna_seq_e115,by=c("gene_name","tissue")) %>% tidyr::replace_na(list(mean_TPM = 0)) %>%
  mutate(exp=if_else(mean_TPM>0.5,"yes","no"),int=if_else(score>0,"yes","no")) %>% #even though neg baits don't have active tissues, but genes have expressing tissues
  mutate(exp_int=if_else(exp=="yes" & int=="yes","yes","no")) %>%
  #dplyr::filter(max_TPM>0.5) %>%
  as.data.table()

new_rmap <- rmap %>% mutate(oeID_mid2=oeID_mid) %>% as.data.table()
setkey(new_rmap,oe_chr,oeID_mid,oeID_mid2)

#get +/- 2.5kb reads
data1 <- neg_tss %>% mutate(start=target_mid-2500,end=target_mid+2500) %>% as.data.table() %>% foverlaps(., new_rmap, by.x=c("target_chr","start", "end"), type="any", nomatch=NULL) %>%
  left_join(dplyr::select(count_scaled,bait_name=name,otherEndID,scaled_Nav,tissue=condition), by=c("bait_name","i.otherEndID"="otherEndID","tissue")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>% 
  group_by(gene_name,bait_name,bait_type,exp_int,tissue) %>% summarise(sum=sum(scaled_Nav)) %>%
  group_by(gene_name,bait_name,bait_type,exp_int) %>% summarise(mean_Nav=mean(sum)) %>% ungroup() %>% add_count(gene_name,bait_name,bait_type) %>% dplyr::filter(n==2)

num=data1 %>% distinct(gene_name,bait_name,bait_type) %>% dplyr::count(bait_type)
data1 %>% pivot_wider(names_from = "exp_int",values_from = "mean_Nav") %>% mutate(ratio=log2(yes/no)) %>% 
  ggviolin(x="bait_type",y="ratio",color="bait_type", add = "boxplot") + 
  annotate(geom = "text", x = num$bait_type, y = -2, label=paste0("n=",num$n) , size=1.5) +
  stat_compare_means(comparisons = list( c("enh", "pro"), c("enh", "neg"), c("pro", "neg") ), label.y = c(4.2, 3.5, 3),size=1.5) +
  labs(x="Bait type",y="interaction frequency when genes in active vs. inactive state (log2)") +
  scale_x_discrete(labels=c("enh"="Positive enhancer","neg"="Control element","pro"="Promoter")) +
  theme(axis.title=element_text(size=7),axis.text = element_text(size=6),legend.position = "none") + scale_fill_manual(values=c("#dd8581ff","#e9b86fff","#77b27bff","#68abdaff"))

#side by side
supfig1h <- data1 %>% dplyr::filter(exp_int=="yes") %>%
  ggviolin(x="bait_type",y="mean_Nav",color="bait_type") + 
  geom_boxplot(aes(color=bait_type),width=0.2,outlier.size = 0.5) +
  scale_y_continuous(trans=pseudo_log_trans(base = 2),breaks = c(0,1,4,16,25,64,256,512)) +
  stat_compare_means(comparisons = list( c("enh", "pro"), c("enh", "neg"), c("pro", "neg") ), label.y = c(11, 11.5, 10.5),size=2) +
  labs(x="",y="interaction frequency",title="only expressing genes") +
  geom_hline(yintercept = 25,color="black",linetype="dashed") +
  scale_x_discrete(labels=c("enh"="Positive enhancer","neg"="Control element","pro"="Promoter")) +
  theme(axis.title=element_text(size=7),axis.text = element_text(size=6),legend.position = "none") 
supfig1h
```

```{r}
#exact number of ditags
rm(chinputs)
for(file in c("CF","FL","HL","FB","MB","HB","NT","HR","TL","TK")) {
  temp_dataset <- read_tsv(file=paste0("~/enh_capC/track_hub/chicago_round3/chinputs/",file,"/",file,".chinput"),skip=1) %>% add_column(tissue=file)
  if (!exists("chinputs")) {
    chinputs <- temp_dataset 
  } else {
    chinputs <- bind_rows(chinputs,temp_dataset)
  }
}
chinputs


data <- neg_tss %>% mutate(start=target_mid-2500,end=target_mid+2500) %>% as.data.table() %>% foverlaps(., new_rmap, by.x=c("target_chr","start", "end"), type="any", nomatch=NULL) %>% left_join(dplyr::select(bmap,baitID,name),by=c("bait_name"="name")) %>%
  left_join(chinputs,by=c("baitID","otherEndID","tissue")) %>% tidyr::replace_na(list(N=0)) %>% 
  group_by(gene_name,bait_name,bait_type,exp_int,tissue) %>% summarise(sum=sum(N)) %>%
  group_by(gene_name,bait_name,bait_type,exp_int) %>% summarise(mean_Nav=mean(sum)) %>% ungroup() %>% add_count(gene_name,bait_name,bait_type) %>% dplyr::filter(n==2)

supfig1h <- data %>% dplyr::filter(exp_int=="yes") %>%
  ggviolin(x="bait_type",y="mean_Nav",color="bait_type") + 
  geom_boxplot(aes(color=bait_type),width=0.2,outlier.size = 0.5) +
  scale_y_continuous(trans=pseudo_log_trans(base = 2),breaks = c(0,1,4,16,30,64,256,512)) +
  stat_compare_means(comparisons = list( c("enh", "pro"), c("enh", "neg"), c("pro", "neg") ), label.y = c(11, 11.5, 10.5),size=3) +
  labs(x="",y="Read counts on interacting expressing genes",title="only expressing genes") +
  geom_hline(yintercept = 30,color="black",linetype="dashed") +
  scale_x_discrete(labels=c("enh"="Positive enhancer","neg"="Control element","pro"="Promoter")) +
  theme(axis.title=element_text(size=9),axis.text = element_text(size=8),legend.position = "none") 
supfig1h

#mean and range for the number of unique read pairs that are determined for each viewpoint
chinputs %>% dplyr::count(baitID,tissue,wt=N,name="tot_ditag") %>% group_by(baitID) %>% summarise(mean=mean(tot_ditag)/2) %>% #for each replicate (tissue is pooled replicates)
  summarise(mean_tag=mean(mean),sd=sd(mean))
```

on average read count on called peaks

```{r}
supfig1_data %>% #dplyr::filter(bait_type=="enh") %>% 
  dplyr::filter(grepl("5kb",type)) %>% dplyr::count(wt=n)
#total number of enhancer bait interactions is 18208, slightly different from supfig1_data, because some of them don't have tad info
ep_exp_5kb %>% #dplyr::filter(bait_type=="enh") %>% 
  distinct(bait_name,target_chr,target_start,target_end) %>% nrow()
ee_CTCF_e10_12_5kb %>% #dplyr::filter(bait_type=="enh") %>% 
  distinct(bait_name,target_chr,target_start,target_end) %>% nrow()
chic3_CTCF_5kb %>% #dplyr::filter(bait_type=="enh") %>% 
  distinct(bait_name,target_chr,target_start,target_end) %>% nrow()
chic3_pc_e10_12_5kb %>% #dplyr::filter(bait_type=="enh") %>% 
  distinct(bait_name,target_chr,target_start,target_end) %>% nrow()

test <- distinct(ep_exp_5kb,bait_name,target_chr,target_start,target_end,tissue,bait_type) %>% bind_rows(distinct(ee_CTCF_e10_12_5kb,bait_name,target_chr,target_start,target_end,tissue,bait_type)) %>% bind_rows(distinct(chic3_CTCF_5kb,bait_name,target_chr,target_start,target_end,tissue,bait_type)) %>% bind_rows(distinct(chic3_pc_e10_12_5kb,bait_name,target_chr,target_start,target_end,tissue,bait_type)) %>% left_join(bmap,by=c("bait_name"="name")) %>% left_join(full_rmap,by=c("target_chr"="chr","target_start"="start","target_end"="end")) %>% left_join(chinputs,by=c("baitID", "id"="otherEndID","tissue")) 

test %>% group_by(baitID,id,bait_type) %>% summarise(mean=mean(N)) %>% dplyr::filter(bait_type != "non") %>% 
  ggviolin(x="bait_type",y="mean",color="bait_type") + geom_boxplot(aes(color=bait_type),width=0.2,outlier.size = 0.5) + scale_y_continuous(trans=pseudo_log_trans(base = 2),breaks = c(0,1,4,16,64,256,512,1024))+
  labs(x="",y="Unique read pairs for all called interactions averaging across tissues") +
  scale_x_discrete(labels=c("enh"="Positive enhancer","neg"="Control element","pro"="Promoter")) +
  theme(axis.title=element_text(size=9),axis.text = element_text(size=8),legend.position = "none")

test %>% group_by(baitID,id,bait_type) %>% summarise(mean=mean(N)) %>% dplyr::filter(bait_type != "non") %>% ungroup() %>% summarise(mean_ditag=mean(mean),sd=sd(mean))
```

how many promoter bait also call enhancer baits

```{r}
setDT(prom_bait)
setkey(prom_bait,chr,start,end)
data2 <- ep_exp_5kb %>% unite("pair",bait_name,promoter,remove = F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% #only golden set
  dplyr::filter(bait_type=="enh") %>% as.data.table() %>% dplyr::select(enh_name=bait_name,everything()) %>%
  #dplyr::filter(bait_name=="hs654") %>%
  mutate(target_start=target_start-5000,target_end=target_end+5000) %>% 
  foverlaps(prom_bait, by.x=c("target_chr","target_start", "target_end"), type="any") %>% dplyr::filter(!is.na(start)) %>% mutate(type="enh_bait") %>% dplyr::select(pro_name=bait_name,enh_name,tissue,type) %>% distinct(enh_name,pro_name,tissue,type)



setDT(pos_enh_bait)
setkey(pos_enh_bait,chr,start,end)
#promoter bait with enhancer bait
data1 <- ee_CTCF_e10_12_5kb %>% dplyr::filter(bait_type=="pro") %>% as.data.table() %>% dplyr::select(pro_name=bait_name,everything()) %>%
  mutate(target_start=target_start-5000,target_end=target_end+5000) %>% 
  foverlaps(pos_enh_bait, by.x=c("target_chr","target_start", "target_end"), type="any") %>% dplyr::filter(!is.na(start)) %>% mutate(type="pro_bait") %>% dplyr::select(pro_name,enh_name=bait_name,tissue,type) %>% distinct(enh_name,pro_name,tissue,type) 


test <- data1 %>% bind_rows(data2) %>% arrange(enh_name,pro_name) %>% add_count(enh_name,pro_name,tissue) %>%   
  dplyr::count(tissue,type,n) %>% mutate(type=if_else(n==2,"enh_pro",type)) %>% distinct() %>% complete(tissue, type,fill=list(nn=0)) %>% 
  mutate(nn=replace(nn, type=="pro_bait", 0))

tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")

library(ggforce)   
df.venn <- data.frame(x = c(0, 0.35),
                      y = c(1, 1),
                      labels = c('enhancer_baits','promoter_baits'))
j=0
for (i in tis) {
  j=j+1
  data <- ggplot(df.venn, aes(x0 = x, y0 = y, r = c(1,1), fill = labels)) + 
    geom_circle(alpha = .3, size = 1, colour = 'grey') + coord_fixed() + theme_void() + labs(title=i) +
    annotate("text", x = c(-0.5,0.5,1.5), y = c(1,1,1), label = dplyr::filter(test,tissue==i)$nn, size = 5) +
    theme(legend.position = "top", legend.direction='horizontal',
                     legend.key.size = unit(4, 'mm'),
                     legend.title = element_blank(), 
                     legend.text = element_text(size=6))
  assign(paste0("p",j),data)
}

plot_grid(p1,p2,p3,p4,p5,p6,p7,p8,p9,p10,nrow=3,label_x = 2, label_y = -1,labels = "10kb")

#integrate them, accounting for tissue
ggplot(df.venn, aes(x0 = x, y0 = y, r = c(1,0.65), fill = labels)) + 
    geom_circle(alpha = .3, size = 1, colour = 'grey') + coord_fixed() + theme_void() + labs(title="all") +
    annotate("text", x = c(-0.5,0.5), y = c(1,1), label = dplyr::count(test,type,wt=nn)$n[1:2], size = 5) +
    theme(legend.position = "top", legend.direction='horizontal',
                     #legend.key.size = unit(4, 'mm'),
                     legend.title = element_blank(), 
                     legend.text = element_text(size=10))

#integrate them, ignore tissues
test1 <- data1 %>% bind_rows(data2) %>% distinct(enh_name,pro_name,type) %>% arrange(enh_name,pro_name) %>% add_count(enh_name,pro_name) %>%   
  dplyr::count(type,n) %>% mutate(type=if_else(n==2,"enh_pro",type)) %>% distinct() %>%  
  mutate(nn=replace(nn, type=="pro_bait", 0)) 

supfig1g <- ggplot(df.venn, aes(x0 = x, y0 = y, r = c(1,0.65), fill = labels)) + 
    geom_circle(alpha = .3, size = 1, colour = 'grey') + coord_fixed() + theme_void() + 
    annotate("text", x = c(-0.5,0.5), y = c(1,1), label = dplyr::count(test1,type,wt=nn)$n[1:2], size = 3) +
    theme(legend.position = "top", legend.direction='horizontal',
                     legend.key.size = unit(4, 'mm'),
                     legend.title = element_blank(), 
                     legend.text = element_text(size=8)) + guides(fill=guide_legend(nrow=2))
supfig1g

```


Fig1B. How the heatmap like plot came from.. just midbrain for hs654
(hs1473-pitx1, hs1166-pax9, hs271-Nr2f1)

```{r}
rm_sd <- function(df, k) {
  df %>%
    mutate(
      rmean = rollapply(rescaled_Nav, k, FUN = mean, partial=T, align="center"),
      rstd = rollapply(rescaled_Nav, width = k, FUN = sd, partial=T, align="center")
    )  #add the prefix to calculate sd successfully
}

tis="MB"

targ="hs654"
dplyr::filter(bmap,name==targ)

chrom="chr9"
gene_limit=c(89500000,92500000)

bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid
enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav) %>% distinct(name,condition,.keep_all = T)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

p0 <- data.frame(oeID_mid=rep(gene_limit[2]-50000,2), condition=c("scale_bar","text"), width=c(100000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.5) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-50000, y = "text", label="100kb",size=2) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

tis_col=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TK","TL"),color=c("#629fd8ff","#716bbdff","#c881bbff","#c1d17dff","#c99516ff","#c75b7cff","#c83737ff","#cfcf3cff","#5b863dff","#73d758ff"))
#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "any")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)

        
p1 <- readcount_bait %>% fill(midpoint,.direction = "downup") %>%
  mutate(rescaled_Nav=if_else(is.na(rescaled_Nav),0,rescaled_Nav)) %>%
  group_by(condition,name) %>% rm_sd(.,5) %>% ungroup() %>%
  mutate(rmean=replace(rmean,oeID_mid==bait_mid,NA)) %>% #remove bait read count
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20",size=0.3) + theme_bw() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = 76, xend = oeID_mid, yend = 76), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "point", x = enh$baitID_mid, y = 85,fill = "red2",color = "red2", size = 1,shape=25) +
  annotate(geom = "text", x = enh$baitID_mid, y = 100, label="enhancer bait",size=2) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = 110,forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 110,label=unique(gid$symbol),size = 2) +
  annotate(geom = "tile", x = enh$baitID_mid, y = 110,fill = "#448838",width=enh$wid,height=8) +
  annotate(geom = "text", x=enh$baitID_mid+50000, y= 110,label=enh$name,size = 2) +
  scale_y_continuous(limits = c(0,125)) +

  labs(y = "MB normalized\ninteraction\nfrequencies") +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.title.y=element_text(size=7,angle = 0, vjust = 0.5),
        axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

p2 <- readcount_bait %>% tidyr::replace_na(list(rescaled_Nav=0, score=0)) %>% 
  ggplot(aes(oeID_mid, condition, fill= rescaled_Nav, width = width)) + geom_tile() +
  scale_fill_gradient(low="white", high="#552492",name="normalized\nread count") +
  labs(y = "") +
  scale_x_continuous(limits = gene_limit) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#-------------------------------------------------------------------------------------
targ="hs1043"

bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid
enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep1 <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav) %>% distinct(name,condition,.keep_all = T)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>% #here, it doesn't matter that midpoint is NA, because midpoint is used to determine how to plot the curve, if midpoint is NA, that means there's no reads on it, so no interactions, no need to plot.
  left_join(max_dep1,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

p3 <- readcount_bait %>% fill(midpoint,.direction = "downup") %>%   mutate(rescaled_Nav=if_else(is.na(rescaled_Nav),0,rescaled_Nav)) %>%
  group_by(condition,name) %>% rm_sd(.,5) %>% ungroup() %>%
  mutate(rmean=replace(rmean,oeID_mid==bait_mid,NA)) %>% #remove bait read count
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20",size=0.3) + theme_bw() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = 76, xend = oeID_mid, yend = 76), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "point", x = enh$baitID_mid, y = 85,fill = "red2",color = "red2", size = 1,shape=25) +
  annotate(geom = "text", x = enh$baitID_mid, y = 100, label="promoter bait",size=2) +
  scale_y_continuous(limits = c(0,125)) +
  
  labs(y = "MB normalized\ninteraction\nfrequencies") +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.title.y=element_text(size=7,angle = 0, vjust = 0.5),
        axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))


p4 <- readcount_bait %>% tidyr::replace_na(list(rescaled_Nav=0, score=0)) %>% 
  ggplot(aes(oeID_mid, condition, fill= rescaled_Nav, width = width)) + geom_tile() +
  scale_fill_gradient(low="white", high="#552492",name="normalized\nread count") +
  labs(y = "") +
  scale_x_continuous(limits = gene_limit) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))


#------------------------------------------------------------------------------------
#reduce size of tracks below
#h3k27ac get scaled to 10
p5 <- read_tsv("~/enh_capC/encode/bigwig/bg/hs654_midbrain_H3K27ac_e11_5.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>20,10,norm_rc/2)) %>%
  
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#c99516ff", colour="#c99516ff", size=0.1) +
  labs(y = "MB H3K27ac") +
  scale_x_continuous(limits = gene_limit) +
  #scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

p6 <- read_tsv("~/enh_capC/encode/bigwig/bg/hs654_midbrain_H3K4me3_e11_5.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>20,10,norm_rc/2)) %>%
  
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#599231ff", colour="#599231ff", size=0.1) +
  labs(y = "MB H3K4me3") +
  scale_x_continuous(limits = gene_limit) +
  #scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#DNase looks better
p7 <- read_tsv("~/enh_capC/encode/bigwig/bg/hs654_midbrain_DNase_e11_5.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>2,10,norm_rc/0.2)) %>%
  
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#c881bbff", colour="#c881bbff", size=0.1) +
  labs(y = "MB DNase") +
  scale_x_continuous(limits = gene_limit) +
  #scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

p8 <- read_tsv("~/enh_capC/encode/bigwig/bg/hs654_GSM5501396_E12_brain_ChIP_CTCF.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>1,10,norm_rc/0.1)) %>%
  
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#678cccff", colour="#678cccff", size=0.1) +
  labs(y = "WB CTCF") +
  scale_x_continuous(limits = gene_limit) +
  #scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))


fig1b <- plot_grid(p0, p1, p2,p3,p4,p5,p6,p7,p8, ncol=1, align = "v", rel_heights = c(0.4,3,0.5,2.1,0.5,0.9,0.9,0.9,0.9))
ggsave("../manuscript/version_4_figure/fig1b.pdf",width = 100, height = 100,units="mm")
fig1b

```

Fig2A. example of results (zi1/zic4), use heatmap like plot to stack
different tissues

```{r}
tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")

targ="hs654"
dplyr::filter(bmap,name==targ)

chrom="chr9"
gene_limit=c(90390345,91695319)


bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid

exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

tis_col=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TK","TL"),color=c("#629fd8ff","#716bbdff","#c881bbff","#c1d17dff","#c99516ff","#c75b7cff","#c83737ff","#cfcf3cff","#5b863dff","#73d758ff"))
enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "any")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)

p0 <- data.frame(oeID_mid=rep(gene_limit[2]-50000,2), condition=c("scale_bar","text"), width=c(100000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-50000, y = "text", label="100kb",size=1) +
  annotate(geom = "text",x=gene_limit[2]-1000000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=1) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()
fig2a <- readcount_bait %>% 
  tidyr::replace_na(list(rescaled_Nav=0, score.y=0)) %>% 
  add_row(condition=c(paste0(tis,rep("-1",length(tis))),"gene","anno","text")) %>%
  mutate(condition=factor(condition, levels = ord)) %>%
  ggplot(aes(oeID_mid, condition, fill= rescaled_Nav, width = width)) + geom_tile() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = condition, xend = oeID_mid, yend = condition), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "tile", x = enh$baitID_mid, y = "gene",fill = "#448838",width=enh$wid,height=0.5) +
  annotate(geom = "text", x=enh$baitID_mid, y= "gene",label=enh$name,size = 2) +
  annotate(geom = "point", x = enh$baitID_mid, y = "anno", fill = "red2",color="red2", size = 1,shape=25) +
  annotate(geom = "text", x = enh$baitID_mid, y = "text", label="bait",size=3) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = "gene",forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= "gene",label=unique(gid$symbol),size = 2.5) +
  scale_fill_gradient(low="white", high="#552492",name="Normalized interaction frequencies") +
  scale_y_discrete(labels=c("FB-1"="","MB-1"="","HB-1"="","CF-1"="","FL-1"="","HL-1"="","HR-1"="","NT-1"="","TL-1"="","TK-1"="","anno"="","text"="","gene"="")) +
  theme_bw() + theme(axis.title=element_blank(),axis.ticks = element_blank(),
                     axis.text.x=element_blank(),panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                     axis.text.y=element_text(size=6),
                     plot.margin = margin(0, 0, 0, 0),
                     legend.position = "bottom",
                     legend.direction='horizontal',
                     legend.key.size = unit(2, 'mm'),
                     legend.title = element_text(size=6), 
                     legend.text = element_text(size=5))
fig2a <- plot_grid(p0,fig2a, ncol=1, align = "v", rel_heights = c(0.1,2))
#ggsave("../manuscript/fig1c.pdf",width = 100, height = 60, units = "mm")
fig2a


#zoom in

p1 <- readcount_bait %>% dplyr::filter(oeID_mid > 91310000 & oeID_mid < 91430000) %>% tidyr::replace_na(list(rescaled_Nav=0, score.y=0)) %>% #should fill NA or not?
  add_row(condition=c(paste0(tis,rep("-1",length(tis))),"gene","anno","text")) %>%
  mutate(condition=factor(condition, levels = ord)) %>%
  ggplot(aes(oeID_mid, condition, fill= rescaled_Nav, width = width)) + geom_tile() + 
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = "gene",forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= "gene",label=unique(gid$symbol),size = 2.5) +
  scale_fill_gradient(low="white", high="#552492",name="Normalized interaction frequencies") +
  scale_y_discrete(labels=c("FB-1"="","MB-1"="","HB-1"="","CF-1"="","FL-1"="","HL-1"="","HR-1"="","NT-1"="","TL-1"="","TK-1"="","anno"="","text"="","gene"="")) +
  scale_x_continuous(breaks= seq(91310000, 91430000, by = 10000)) +
  theme_bw() + theme(axis.title=element_blank(),axis.ticks.y = element_blank(),
                     panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                     axis.text.x=element_text(size=6),
                     plot.margin = margin(0, 0, 0, 0),
                     legend.position = "bottom",
                     legend.direction='horizontal',
                     legend.key.size = unit(2, 'mm'),
                     legend.title = element_text(size=6), 
                     legend.text = element_text(size=5))
p1

p2 <- readcount_bait %>% dplyr::filter(oeID_mid > 91310000 & oeID_mid < 91430000) %>% tidyr::replace_na(list(rescaled_Nav=0, score.y=0)) %>% 
  ggplot(aes(oeID_mid, rescaled_Nav, color= condition)) + 
  geom_line() +
  scale_x_continuous(breaks= seq(91310000, 91430000, by = 10000)) +
  theme_bw() + theme(axis.title=element_blank(),axis.ticks.y = element_blank(),
                     axis.text.x=element_text(size=6),
                     plot.margin = margin(0, 0, 0, 0),
                     legend.position = c(0.2,0.8),
                     legend.direction='horizontal',
                     legend.key.size = unit(2, 'mm'),
                     legend.title = element_text(size=6), 
                     legend.text = element_text(size=5))


supfig1l <- plot_grid(p2,p1,ncol=1, rel_heights = c(1,2), align = "v")
ggsave("../manuscript/fig1l.pdf",width = 180, height = 60, units = "mm")
supfig1l
```

How many genes(or promoters) are skipped, expression doesn't matter
#have alternative promoters

```{r}
gene_in_middle <- read_tsv("~/ensembl_regulatory/ep_pair_gene_in_middle.tsv")
gene_in_middle <- read_tsv("~/ensembl_regulatory/ep_pair_no_gene_in_middle.tsv") %>% bind_rows(gene_in_middle)

intra_gene <- read_tsv("~/ensembl_regulatory/within_gene_enh.tsv",col_names = c("chr","bait_start","bait_end","bait_name","gene_chr","gene_start","gene_end","gene_name","bp")) %>% group_by(bait_name,gene_name) %>% summarise(gene_chr=last(gene_chr),gene_start=min(gene_start),gene_end=max(gene_end)) %>% ungroup() %>% distinct()

#in ep_exp_5kb, promoter coordinates are expanded by 5kb up and down, so subtract it
p1 <- ep_exp_5kb %>% dplyr::select(starts_with("target"),starts_with("bait"),promoter,gene) %>% dplyr::filter(bait_type=="enh") %>%
  separate(promoter,paste("pro",c("chr","start","end"),sep="_"),sep="_") %>%
  mutate(pro_start=as.numeric(pro_start),pro_end=as.numeric(pro_end), up=if_else(target_start>bait_start,bait_end,pro_end),down=if_else(target_start>bait_start,pro_start,bait_start)) %>%
  left_join(gene_in_middle,by=c("pro_chr"="chr","up"="ep_start","down"="ep_end","bait_name")) %>%
  distinct(bait_name,across(starts_with("pro")),gene,mid_gene) %>% group_by(bait_name,pro_chr,pro_start,pro_end) %>% mutate(num=n()) %>% 
  dplyr::filter(gene != mid_gene | is.na(mid_gene)) %>% 
  mutate(pro_start=pro_start+5000,pro_end=pro_end-5000, num=replace(num,is.na(mid_gene),0)) %>% distinct(gene,num) %>% ungroup() %>%
  left_join(intra_gene,by="bait_name") %>% mutate(intra=str_remove_all(gene,gene_name)) %>%
  mutate(gr=if_else(num>1,">=2 pro",if_else(num==1,"1 pro",if_else((str_count(intra,".")==str_count(gene,".") | is.na(gene_name)),"nei", "intra")))) %>%
  group_by(bait_name) %>% dplyr::filter(num==min(num)) %>% arrange(bait_name) %>% distinct(bait_name,gr) %>% add_count(bait_name,name="num") %>%  dplyr::filter((num>1 & gr=="intra") | num==1) %>% ungroup() %>% #enhancer-centric
  dplyr::count(gr) %>% mutate(gr = factor(gr, levels = c("intra","nei", "1 pro",">=2 pro")), frac=paste0(formatC(n/sum(n)*100, format = "f",digits=1),"%")) %>%
  ggplot(aes(x=gr,y=n,fill=gr)) + geom_bar(stat="identity", position = "dodge") +
  geom_text(aes(label = frac),position = position_dodge(width = 0.9), vjust = -0.5) + labs(x="",y="Number of enhancers",title="all ep pairs") +
  scale_x_discrete(labels=c("Intragenic","Neighboring promoter","Skipping 1 promoter",expression("Skipping ">="2 promoters"))) + theme_bw() + theme(legend.position="none") + scale_fill_manual(values=c("#dd8581ff","#e9b86fff","#77b27bff","#68abdaff"))


#2.5kb up- and down-stream of TSS to overlap with target fragment 
refgene_tss <- read_tsv("~/ensembl_regulatory/refGene_all_TSS_2kb.txt",col_names = c("chr","tss_start","tss_end","strand","gene_name")) %>% as.data.table()
setkey(refgene_tss,chr,tss_start,tss_end)

p2 <- ep_exp_5kb %>% dplyr::select(starts_with("target"),starts_with("bait"),promoter,gene) %>% dplyr::filter(bait_type=="enh") %>% as.data.table() %>%
  foverlaps(refgene_tss, by.x=c("target_chr","target_start", "target_end"), type="any") %>% dplyr::filter(!is.na(gene_name)) %>% distinct(across(starts_with("tss")),across(starts_with("bait")),gene_name) %>% #dplyr::filter(bait_name=="hs654") %>%
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1)) %>%
  #dplyr::select(-c(tss_start,tss_end)) %>% 
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>% distinct(bait_name,bait_type,i.gene_name,gene_name) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% #group_by(bait_name,i.gene_name) %>% 
  add_count(bait_name,i.gene_name) %>% mutate(n=replace(n, is.na(gene_name), 0)) %>% distinct(bait_name,n,i.gene_name) %>%
  left_join(dplyr::filter(intra_gene, gene_name %in% refgene_tss$gene_name),by="bait_name") %>% 
  mutate(gr=if_else(n>1,">=2 genes",if_else(n==1,"1 gene",if_else((i.gene_name!=gene_name | is.na(gene_name)),"nei", "intra")))) %>%
  group_by(bait_name) %>% dplyr::filter(n==min(n)) %>% arrange(bait_name) %>% distinct(bait_name,gr) %>% add_count(bait_name,name="num") %>%  dplyr::filter((num>1 & gr=="intra") | num==1) %>% ungroup() %>% #enhancer-centric
  dplyr::count(gr) %>% mutate(gr = factor(gr, levels = c("intra","nei", "1 gene",">=2 genes")), frac=paste0(formatC(n/sum(n)*100, format = "f",digits=1),"%")) %>%
  ggplot(aes(x=gr,y=n,fill=gr)) + geom_bar(stat="identity", position = "dodge") +
  geom_text(aes(label = frac),position = position_dodge(width = 0.9), vjust = -0.5) + labs(x="",y="Number of enhancers") +
  scale_x_discrete(labels=c("Intragenic","Neighboring gene","Skipping 1 gene",expression("Skipping ">="2 promoters"))) + labs(title="all enh-gene pairs") + theme_bw() + theme(legend.position="none") + scale_fill_manual(values=c("#dd8581ff","#e9b86fff","#77b27bff","#68abdaff"))


#use TSS instead of promoters to defined the genes skipped, use refseq genes only
#enhancer-centric, which means if one enhancer interacts with 2 genes, then just count the closest one, the other one doesn't mean skipping
#2.5kb up- and down-stream of TSS to overlap with target fragment 
#only golden set
 
files <- list.files(path="~/enh_capC/encode/RNA-seq/", pattern="tsv$", recursive=T) 
rm(dataset)
for (path in files){
  name=str_split(path, "_",simplify = T)
  if (length(name)==3) {
    temp_dataset <- read_tsv(paste0("~/enh_capC/encode/RNA-seq/",path),col_types = "ccddddddddddddddd") %>% mutate(tissue = name[1,1], rep=name[1,3], stage=name[1,2]) %>% dplyr::filter(grepl("ENS",gene_id)) %>% dplyr::select(gene_id,expected_count,TPM,FPKM,tissue,stage,rep)
    if (!exists("dataset")) {
      dataset <- temp_dataset 
    } else {
      dataset <- temp_dataset %>% bind_rows(dataset)
    }
    rm(temp_dataset)
  }
}

tis_corr <- data.frame(tissue=c("embryonicfacialprominence","forebrain","heart","hindbrain","limb","limb","liver","midbrain","neuraltube"),new=c("CF","FB","HR","HB","FL","HL","LV","MB","NT"))

rna_seq_exp <- dataset %>% group_by(gene_id,tissue,stage) %>% partition(new_cluster(10)) %>% summarise(mean_TPM=mean(TPM)) %>% collect() %>% #dplyr::filter(mean_TPM>0.5) %>%
  separate(gene_id,c("gene",NULL),sep="[^[:alnum:]]+") %>% left_join(read_tsv("gene_ensID_vs_name.tsv",col_names = c("gene","gene_name")), by="gene") %>% ungroup() %>% dplyr::filter(stage=="E115") %>% left_join(tis_corr,by="tissue") %>% dplyr::filter(! is.na(new)) %>% dplyr::select(-tissue) %>% dplyr::select(tissue=new,gene_name,stage,mean_TPM) %>% mutate(exp=if_else(mean_TPM>0.5,"yes","no")) 

pos_pos_int <- ep_exp_5kb %>% dplyr::filter(bait_type=="enh") %>% separate(gene,paste0("gene_",seq(1:15)),sep=",") %>% pivot_longer(!c(target_chr, target_start,target_end,bait_chr,bait_start,bait_end,bait_name,score,tissue,promoter,overlapping_bp, target_mid, otherEndID,dist,overlap_len,bait_type), names_to = "name", values_to = "gene_name") %>% mutate(gene_name=gsub(" ","",gene_name)) %>% dplyr::filter(!is.na(gene_name)) %>% left_join(rna_seq_exp,by=c("gene_name","tissue")) %>% left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% dplyr::filter(exp=="yes" & class=="POS") %>% distinct(bait_name,promoter) %>% unite("pair",c("bait_name","promoter"))  #969

tis_corr <- data.frame(tissue=c("embryonicfacialprominence","forebrain","heart","hindbrain","limb","limb","liver","midbrain","neuraltube"),new=c("CF","FB","HR","HB","FL","HL","LV","MB","NT"))
rna_seq_e115 <- left_join(rna_seq_stage,tis_corr,by="tissue") %>% dplyr::filter(stage=="E115") %>% dplyr::select(gene_name,mean_TPM,tissue=new)

enh_tss_exp <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>%
  dplyr::select(starts_with("target"),starts_with("bait"),promoter,gene,score) %>% dplyr::filter(bait_type=="enh") %>% as.data.table() %>%
  foverlaps(refgene_tss, by.x=c("target_chr","target_start", "target_end"), type="any") %>% dplyr::filter(!is.na(gene_name)) %>%
  left_join(rna_seq_e115,by="gene_name") %>% tidyr::replace_na(list(mean_TPM = 0)) %>%
  group_by(across(starts_with("tss")),across(starts_with("bait")),gene_name) %>% summarise(max_TPM=max(mean_TPM)) %>% ungroup() %>%
  dplyr::filter(max_TPM>0.5) %>%
  as.data.table()

test <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% separate(promoter,into=c("promoter_chr","promoter_start","promoter_end")) %>% dplyr::select(starts_with("bait"),gene,starts_with("promoter"),tissue) %>% distinct()
write_tsv(test,"golden_set_ep_interaction.tsv")
test <- stars %>% left_join(bmap,by=c("elementID"="name")) %>% dplyr::select(bait_name=elementID,stars,tissue=tis_group,chr,start,end) %>% distinct() %>% dplyr::filter(stars==3)
write_tsv(test,"vista_enhancer_3_star.tsv")


test <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% dplyr::select(starts_with("target"),starts_with("bait"),promoter,gene,score,tissue) %>% dplyr::filter(bait_type=="enh") %>% separate(gene,paste0("gene_",seq(1:15)),sep=",") %>% pivot_longer(!c(target_chr, target_start,target_end,target_mid,bait_chr,bait_start,bait_end,bait_name,score,tissue,promoter,bait_type), names_to = "name", values_to = "gene_name") %>% dplyr::filter(!is.na(gene_name)) %>% unite("bait",c("bait_chr","bait_start","bait_end")) %>% dplyr::select(bait,bait_name,promoter,gene_name,tissue) %>% distinct()
write_tsv(test,"golden_set_enh_gene_interactions.tsv")


enh_tss_exp %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1)) %>%
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>% distinct(bait_name,bait_type,i.gene_name,gene_name) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% #group_by(bait_name,i.gene_name) %>% 
  add_count(bait_name,i.gene_name) %>% mutate(n=replace(n, is.na(gene_name), 0)) %>% distinct(bait_name,n,i.gene_name) %>%
  left_join(dplyr::filter(intra_gene, gene_name %in% refgene_tss$gene_name),by="bait_name") %>% 
  mutate(gr=if_else(n>1,">=2 genes",if_else(n==1,"1 gene",if_else((i.gene_name!=gene_name | is.na(gene_name)),"nei", "intra")))) %>%
  group_by(bait_name) %>% dplyr::filter(n==min(n)) %>% #closest genes
  arrange(bait_name) %>% distinct(bait_name,gr) %>% add_count(bait_name,name="num") %>%  dplyr::filter((num>1 & gr=="intra") | num==1) %>% ungroup() %>% #enhancer-centric
  mutate(gr=if_else(gr=="intra","nei",gr)) %>% #consider intragenic as neighboring
  dplyr::count(gr) %>% mutate(gr = factor(gr, levels = c("nei", "1 gene",">=2 genes")), frac=paste0(formatC(n/sum(n)*100, format = "f",digits=1),"%")) 

```

```{r}
library(ggpattern)
#closest genes
fig4a <- enh_tss_exp %>% #dplyr::filter(bait_name=="hs654") %>%
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1)) %>%
  #dplyr::select(-c(tss_start,tss_end)) %>% 
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>% distinct(bait_name,bait_type,i.gene_name,gene_name) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% #group_by(bait_name,i.gene_name) %>% 
  add_count(bait_name,i.gene_name) %>% mutate(n=replace(n, is.na(gene_name), 0)) %>% distinct(bait_name,n,i.gene_name) %>%
  left_join(dplyr::filter(intra_gene, gene_name %in% refgene_tss$gene_name),by="bait_name") %>% 
  mutate(gr=if_else(n>1,">=2 genes",if_else(n==1,"1 gene",if_else((i.gene_name!=gene_name | is.na(gene_name)),"nei", "intra")))) %>%
  group_by(bait_name) %>% dplyr::filter(n==min(n)) %>% #closest genes
  arrange(bait_name) %>% distinct(bait_name,gr) %>% add_count(bait_name,name="num") %>%  dplyr::filter((num>1 & gr=="intra") | num==1) %>% ungroup() %>% #enhancer-centric
  mutate(gr=if_else(gr=="intra","nei",gr)) %>% #consider intragenic as neighboring
  dplyr::count(gr) %>% mutate(gr = factor(gr, levels = c("nei", "1 gene",">=2 genes")), frac=paste0(formatC(n/sum(n)*100, format = "f",digits=1),"%")) %>%
  
  ggplot(aes(x=gr,y=n)) + 
  #geom_bar(stat="identity", position = "dodge",width = .3, fill="#68abdaff") +
  geom_col_pattern(aes(pattern_angle=gr), position = position_dodge(preserve = "single"),width = .3,
                   color = "black", 
                   fill="#68abdaff",
                   pattern_color = "white",
                   pattern_fill = "white",
                   pattern_density = 0.1,
                   pattern_spacing = 0.1,
                   pattern_key_scale_factor = 0.6) +
  geom_text(aes(label = frac),position = position_dodge(width = 0.9), vjust = -0.5,size=2) + labs(x="",y="Number of enhancers") +
  scale_x_discrete(labels=c("Neighboring gene","Skipping 1 gene",expression("Skipping ">="2 genes"))) + #labs(title="golden set enh-gene pairs") + 
  theme_bw() + theme(legend.position="none",
                     axis.text.x = element_text(angle = 15, vjust = 1, hjust = 1,size=6),
                     axis.title.y=element_text(size=7),axis.text.y = element_text(size=6)) + scale_pattern_angle_manual(values = c(90,45,-45)) 
#+ guides(pattern = guide_legend(override.aes = list(fill = "white")), fill = guide_legend(override.aes = list(pattern = "none")))
fig4a
```

```{r}
#regulate the closest TSS(gene) or not, based on exact distance
p4 <- enh_tss_exp %>% #dplyr::filter(bait_name=="hs654") %>%
  mutate(up=bait_start-2000000,down=bait_end+2000000) %>% 
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>% group_by(bait_name,i.gene_name) %>% mutate(dist=abs((bait_start+bait_end)/2-(tss_start+tss_end)/2)) %>% dplyr::filter(dist==min(dist)) %>% 
  mutate(closest_gene=if_else(i.gene_name==gene_name,1,0)) %>% 
  group_by(bait_name) %>% dplyr::filter(closest_gene==max(closest_gene)) %>% arrange(bait_name) %>% distinct(bait_name,closest_gene) %>% ungroup() %>% #enhancer-centric
  dplyr::count(closest_gene) %>% mutate(closest_gene=as.factor(closest_gene), frac=paste0(formatC(n/sum(n)*100, format = "f",digits=1),"%")) %>%
  ggplot(aes(x="",y=n,fill=closest_gene)) + geom_bar(stat="identity", color = "white") + geom_text(aes(label=paste0("n=",n,"\n(",frac,")"), x = 1.1), position=position_stack(0.5), size=4) + coord_polar("y", start=0) + theme_void() + scale_fill_discrete(labels=c("no","yes")) + scale_fill_manual(values=c("gray80","#68abdaff"))

#for those intragenic enhancers, how many of them regulate host genes
p5 <- enh_tss_exp %>% mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1)) %>%
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>% distinct(bait_name,bait_type,i.gene_name,gene_name) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% #group_by(bait_name,i.gene_name) %>% 
  add_count(bait_name,i.gene_name) %>% mutate(n=replace(n, is.na(gene_name), 0)) %>% distinct(bait_name,n,i.gene_name) %>%
  right_join(dplyr::filter(intra_gene, gene_name %in% refgene_tss$gene_name),by="bait_name") %>% dplyr::filter(!is.na(i.gene_name)) %>% #remove those don't have target genes
  mutate(intra=if_else(i.gene_name==gene_name , "yes","no"),len=nchar(intra)) %>%
  group_by(bait_name) %>% dplyr::filter(len==max(len)) %>%  distinct(bait_name,intra) %>% ungroup() %>% #enhancer-centric
  dplyr::count(intra) %>% mutate(frac=paste0(formatC(n/sum(n)*100, format = "f",digits=1),"%")) %>%
  ggplot(aes(x="",y=n,fill=intra)) + geom_bar(stat="identity", color = "white") + geom_text(aes(label=paste0(c("Not host gene", "Host gene"),"\n",frac), x = 1.05), position=position_stack(0.5), size=2) + 
  coord_polar("y", start=0) + theme_void() + #labs(title="Intragenic enhancer interact with host gene or not") +
  scale_fill_manual(values=c("gray80","#68abdaff")) + theme(legend.position = "none")
#how many target genes for each enhancer

p6 <- enh_tss_exp %>% dplyr::count(bait_name,name="num") %>% mutate(num=if_else(num>4, ">=5",paste0("=",num))) %>% dplyr::count(num) %>% mutate(num = factor(num, levels = c("=1","=2", "=3","=4",">=5")), frac=paste0(formatC(n/sum(n)*100, format = "f",digits=1),"%")) %>%
  ggplot(aes(x="",y=n,fill=num)) + geom_bar(stat="identity", color = "white") + geom_text(aes(label=paste0("n", num,"\n",frac), x = 1.05), position=position_stack(0.5), size=2) + 
  coord_polar("y", start=0) + theme_void() + #labs(title="Number of genes that enhancers interact with") + 
  theme(legend.position="none") + scale_fill_manual(values=c("#dd8581ff","#e9b86fff","#77b27bff","#68abdaff","#9f9f9fff")) 


p4
p5
p6
```

```{r}
fig1de <- plot_grid(p6,p5,nrow=1,labels = c("D","E"),label_size = 10) 
ggsave("../manuscript/version_4_figure/fig1e.pdf",width = 3, height = 1.5)
fig4a
fig1de
```

Supp fig1f enh-tss distance median

```{r}
supfig1f <- enh_tss_exp %>% mutate(dist_kb=abs((tss_start+tss_end)/2-(bait_start+bait_end)/2)/1000) %>% arrange(dist_kb) %>% 
  dplyr::count(gr=cut(dist_kb, breaks= seq(0, 2010, by = 20),labels=FALSE)) %>% mutate(cumsum=cumsum(n)) %>% mutate(cumperc=cumsum / sum(n) *100/2, label=gr*20-20) %>% 
  ggplot(aes(x=label,y=n)) + geom_freqpoly(stat="identity") + geom_path(aes(y=cumperc),colour="red", size=0.7) +
  scale_y_continuous(
    name = "Number",
    sec.axis = sec_axis(~.*2, name="Cumulative (%)") )  +
  labs(x="Enhancer to target TSS distance (kb)") + theme_classic() +
  theme( axis.line.y.right = element_line(color = "red"), 
       axis.ticks.y.right = element_line(color = "red"),
       axis.text.y.right = element_text(color = "red",size=6),
       axis.title.y.right = element_text(color = "red",size=7),
       axis.title.y.left = element_text(size=7),
       axis.text.y.left = element_text(size=6),
       axis.text.x = element_text(size=6), axis.title.x=element_text(size=7),
       plot.margin = margin(.8, .5, 0.2, .5, "cm")) +
  geom_segment(aes(x = 410, y = 100/2/2, xend = Inf, yend = 100/2/2),linetype=2,color="red") +
  geom_segment(aes(x = 410, y = -Inf, xend = 410, yend = 100/2/2),linetype=2,color="red")

enh_tss_exp %>% mutate(dist_kb=abs((tss_start+tss_end)/2-(bait_start+bait_end)/2)/1000) %>% summarise(median=median(dist_kb))

supfig1abc <- plot_grid(supfig1a,supfig1b,supfig1h,labels = c("A","B","C"),nrow=1, rel_widths = c(1,1,1),label_size = 10)
supfig1gk <- plot_grid(supfig1g,supfig1k,labels = c("F","G"),ncol=1,rel_heights = c(1,3), label_size = 10)
supfig1def <- plot_grid(supfig1d,supfig1f,supfig1gk,labels = c("D","E",""),nrow=1,rel_widths = c(1,1,0.8),label_size = 10)
supfig1 <- plot_grid(supfig1abc,supfig1def,labels = c("",""),ncol=1, rel_heights = c(1,1.1))
ggsave("../manuscript/Supp_fig1.pdf",width=183,height = 120,units="mm")
supfig1
```

whether genes more active when enhancer active? get all the target genes

```{r}
tis=c("FB","MB","HB","NT","CF","LB","HR")
#limb should be one
tis_corr <- data.frame(tissue=c("embryonicfacialprominence","forebrain","heart","hindbrain","limb","liver","midbrain","neuraltube"),new=c("CF","FB","HR","HB","LB","LV","MB","NT"))
rna_seq_e115 <- left_join(rna_seq_stage,tis_corr,by="tissue") %>% dplyr::filter(stage=="E115") %>% dplyr::select(gene_name,mean_TPM,tissue=new)

enh_tss_exp_tis_LB <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>%
  dplyr::select(starts_with("target"),starts_with("bait"),promoter,gene,int_tis=tissue,score) %>% dplyr::filter(bait_type=="enh") %>% as.data.table() %>%
  foverlaps(refgene_tss, by.x=c("target_chr","target_start", "target_end"), type="any") %>% dplyr::filter(!is.na(gene_name)) %>% mutate(int_tis=if_else(int_tis %in% c("FL","HL"),"LB",int_tis)) %>%
  left_join(rna_seq_e115,by="gene_name") %>% tidyr::replace_na(list(mean_TPM = 0)) %>%
  group_by(across(starts_with("tss")),across(starts_with("bait")),gene_name) %>% mutate(max_TPM=max(mean_TPM)) %>% ungroup() %>% dplyr::filter(max_TPM>0.5) %>% #get the expressing ones
  distinct(bait_chr,tss_start,tss_end,across(starts_with("bait")),gene_name,max_TPM,int_tis) %>% as.data.table()

#whether target genes expression are different in active and inactive tissues
#include all the target genes (not just closest ones)
data <- enh_tss_exp_tis_LB %>% left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% #this would only include pos enh and pos interactions in the same tissue
  complete(nesting(gene_name,bait_name),int_tis=tis,fill=list(class="NEG")) %>% #get both positive and negative tissues
  left_join(rna_seq_e115,by=c("gene_name","int_tis"="tissue")) %>% distinct(gene_name,bait_name,int_tis,class,mean_TPM) %>% dplyr::filter(!is.na(mean_TPM)) %>%
  group_by(gene_name,int_tis) %>% mutate(n=if_else(length(unique(class))==2,1,0)) %>% ungroup() %>% mutate(class=if_else(n==0,class,"POS")) %>% distinct(gene_name,int_tis,mean_TPM,class) %>% ##it's gene TSS centric, because in the same tissue, there would be multiple active enhancers 
  group_by(gene_name,class) %>% summarise(TPM=mean(mean_TPM)) %>% ungroup() %>% add_count(gene_name) %>% dplyr::filter(n==2) #only enh_gene pairs with pos and neg

stat.test <- data %>% wilcox_test(TPM ~ class, paired = T) %>% #paired samples, because expression of the same gene in different tissues, the expression have the samb background
  mutate(p=formatC(p, format = "g", digits = 3)) #actual p-value

data %>% ggviolin(x="class",y="TPM",fill="class",add = "boxplot",add.params = list(fill = "white",width=.2),width=1) + 
  scale_y_continuous(trans = log2_trans(), breaks = trans_breaks("log2", function(x) 2^x, n=20), labels = trans_format("log2", math_format(2^.x))) +
  stat_pvalue_manual(stat.test, label = "p",y.position=15,size=2) + #use this to add precise p-value
  labs(x="") +
  theme(legend.position="none",axis.text.x = element_text(size=6),
        axis.title.y=element_text(size=7),axis.text.y = element_text(size=6),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7),
        legend.text = element_text(size=6)) +
  scale_x_discrete(labels=c("NEG"="no interaction or\ninactive tissues","POS"="have interactions\nin active tissues")) +
  scale_fill_manual(values=c("#e9b86fff","#68abdaff"))

num=data %>% distinct(gene_name) %>% nrow()
#plot fold-change, because directly plot would smooth out the fold change
supfig2j <- data %>% pivot_wider(names_from = class, values_from = TPM) %>% mutate(fold=POS/NEG) %>% dplyr::filter(NEG != 0) %>% arrange(desc(fold)) %>%
  ggbarplot(y = "fold", add = c("mean_se")) + labs(x="",y="Fold-change") + scale_x_discrete(labels="interact in active tissues vs.\nno interaction or inactive tissues") + annotate(geom = "text", x= 1, y = 18, label=paste0("n=",num,"\np=", stat.test$p) , size=2.5) +
  theme(axis.text.x = element_text(size=6),
        axis.title.y=element_text(size=7),axis.text.y = element_text(size=6))

data %>% pivot_wider(names_from = class, values_from = TPM) %>% mutate(fold=POS/NEG) %>% dplyr::filter(NEG != 0) %>% arrange(desc(fold)) %>% summarise(mean_fold=mean(fold))
supfig2j
```

```{r}
#whether skipping genes expression are different in active and inactive tissues
#only closest enh-gene pairs that have skipping genes inside

#longest interaction
#multiple skipping genes, so collapse by gene

enh_tss_exp_tis_LB %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>% distinct(bait_name,bait_type,i.gene_name,gene_name,int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,target_gene=i.gene_name,bait_name,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs?????????
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% complete(nesting(skipping_gene,target_gene,bait_name),int_tis=tis,fill=list(class="NEG")) %>% #get both positive and negative tissues, now NEG represent no interaction or no enh activity
  dplyr::select(-bait_name) %>% pivot_longer(!c(int_tis,class),names_to = "type",values_to = "gene_name") %>% distinct() %>% arrange(int_tis,gene_name) %>%
  group_by(gene_name,int_tis) %>% mutate(n=if_else(length(unique(type))==2,1,0)) %>% ungroup() %>% mutate(type=if_else(n==1,"target_gene",type)) %>% distinct() %>% dplyr::filter(type!="target_gene") %>% #remove all possible target genes
  group_by(int_tis,gene_name) %>% mutate(n=if_else(length(unique(class))==2,1,0)) %>% ungroup() %>% mutate(class=if_else(n==1,"POS",class)) %>% #since skipping genes can be target gene in another interaction, so if POS in one interaction mean POS in that tissue
  
  left_join(rna_seq_e115,by=c("gene_name","int_tis"="tissue")) %>% dplyr::filter(!is.na(mean_TPM)) %>% distinct(gene_name,int_tis,mean_TPM,class) %>% 
  group_by(gene_name,class) %>% summarise(TPM=mean(mean_TPM)) %>% ungroup() %>% add_count(gene_name) %>% dplyr::filter(n==2) %>% #only enh_gene pairs with pos and neg
  ggviolin(x="class",y="TPM",fill="class",add = "boxplot",add.params = list(fill = "white",width=.1),width=.5) + yscale("log2", .format = TRUE) +
  stat_compare_means(label = "p.format", comparisons = list(c("NEG","POS")),method = "wilcox.test",paired = T,tip.length = .01,label.y = 14,size=4) + 
  labs(x="",title = "skipping genes") +
  theme(legend.position="none") + 
  scale_x_discrete(labels=c("NEG"="inactive","POS"="active")) +
  scale_fill_manual(values=c("#e9b86fff","#68abdaff"))
```

whether skipped genes express much less than target genes in active tissues? 
```{r}
#only closest enh-gene pairs that have skipping genes inside
tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL")

tis_corr <- data.frame(tissue=c("embryonicfacialprominence","forebrain","heart","hindbrain","limb","limb","liver","midbrain","neuraltube"),new=c("CF","FB","HR","HB","FL","HL","LV","MB","NT"))
rna_seq_e115 <- left_join(rna_seq_stage,tis_corr,by="tissue") %>% dplyr::filter(stage=="E115") %>% dplyr::select(gene_name,mean_TPM,tissue=new)

#get all the skipped genes and target genes

data <- enh_tss_exp_tis %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>% distinct(bait_name,bait_type,i.gene_name,gene_name,int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,target_gene=i.gene_name,bait_name,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>%
  dplyr::select(-bait_name) %>% pivot_longer(!c(int_tis,class),names_to = "type",values_to = "gene_name") %>% distinct() %>% dplyr::filter(class=="POS") %>%
  group_by(gene_name,int_tis) %>% mutate(n=if_else(length(unique(type))==2,1,0)) %>% ungroup() %>% mutate(type=if_else(n==1,"target_gene",type)) %>% distinct() %>% #since skipping genes can be target gene in another interaction, so if POS in one interaction mean POS in that tissue
  left_join(rna_seq_e115,by=c("gene_name","int_tis"="tissue")) %>% dplyr::filter(!is.na(mean_TPM)) %>% distinct(gene_name,int_tis,mean_TPM,class,type) %>%
  mutate(int_tis=factor(int_tis,levels=c("FB","MB","HB","NT","CF","FL","HL","HR"))) %>% dplyr::filter(int_tis != "HR")
  
stat.test <- data %>% group_by(int_tis) %>% wilcox_test(mean_TPM ~ type, paired = F) %>% 
  add_x_position(x="int_tis") %>% adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 3)) #actual p-value
 
fig4d <- data %>% 
  ggviolin(x="int_tis",y="mean_TPM", fill="type",  trim=T) +  #,alpha=0.8
  geom_boxplot(aes(dodge=type),position = position_dodge(width = 0.8), outlier.shape = NA,width=0.02) +
  scale_y_continuous(trans=pseudo_log_trans(base = 2), breaks = scales::trans_breaks("log2", function(x) 2^x) (c(1, 1000)), labels = scales::trans_format("log2", scales::math_format(2^.x))) +
  stat_pvalue_manual(stat.test, label = "p.adj",y.position=11,tip.length = .01,size=2.5,angle=30,vjust=-1,hjust=0.5) + #???paired or not
  labs(x="",y="TPM") +
  scale_fill_manual(values=c("#e9b86fff","#68abdaff"), name="", labels=c("skipping_gene"="Skipped genes","target_gene"="Target genes")) +
      theme(axis.text.x = element_text(size=6),
        axis.title.y=element_text(size=7),axis.text.y = element_text(size=6),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7),
        legend.text = element_text(size=6))


num=enh_tss_exp_tis %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>% distinct(bait_name,bait_type,i.gene_name,gene_name,int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,target_gene=i.gene_name,bait_name,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>% distinct(target_gene,bait_name)

exp_gene <- rna_seq %>% ungroup() %>% distinct(gene_name) %>% add_column(exp="yes")
data1 <- enh_tss_exp_tis %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>% distinct(bait_name,bait_type,i.gene_name,gene_name,int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,target_gene=i.gene_name,bait_name,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(exp_gene,by=c("skipping_gene"="gene_name")) %>% 
  dplyr::filter(!is.na(exp)) %>% #filter out non-expressed genes
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>%
  dplyr::select(-bait_name) %>% pivot_longer(!c(int_tis,class),names_to = "type",values_to = "gene_name") %>% distinct() %>% dplyr::filter(class=="POS") %>%
  group_by(gene_name,int_tis) %>% mutate(n=if_else(length(unique(type))==2,1,0)) %>% ungroup() %>% mutate(type=if_else(n==1,"target_gene",type)) %>% distinct() %>% #since skipping genes can be target gene in another interaction, so if POS in one interaction mean POS in that tissue
  left_join(rna_seq_e115,by=c("gene_name","int_tis"="tissue")) %>% dplyr::filter(!is.na(mean_TPM)) %>% distinct(gene_name,int_tis,mean_TPM,class,type) %>%
  mutate(int_tis=factor(int_tis,levels=c("FB","MB","HB","NT","CF","FL","HL","HR"))) 
  
stat.test <- data1 %>% group_by(int_tis) %>% wilcox_test(mean_TPM ~ type, paired = F) %>% 
  add_x_position(x="int_tis") %>% adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 3)) #actual p-value
 
data1 %>% ggboxplot(x="int_tis",y="mean_TPM",color="type",outlier.size = 0.3)  + 
  scale_y_continuous(trans=pseudo_log_trans(base = 2), breaks = scales::trans_breaks("log2", function(x) 2^x) (c(1, 1000)), labels = scales::trans_format("log2", scales::math_format(2^.x))) +
  #yscale("log2", .format = TRUE) + 
  stat_pvalue_manual(stat.test, label = "p.adj",y.position=11,tip.length = .01,size=2.5,angle=30,vjust=-1,hjust=0.5) + 
  labs(x="",y="log2 (TPM)",title = "Exp. skipped genes") +
  scale_color_manual(values=c("#ec7f26e3","#4d77b1e8"),name="",labels=c("skipping_gene"="Skipped genes","target_gene"="Target genes")) +
  theme(axis.text.x = element_text(size=6),
        axis.title.y=element_text(size=7),axis.text.y = element_text(size=6),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7),
        legend.text = element_text(size=6))

fig4d
num #total interaction number
data %>% group_by(type,int_tis) %>% summarise(mean_tis=mean(mean_TPM)) %>% pivot_wider(names_from = "type",values_from = "mean_tis") %>% mutate(fold=skipping_gene/target_gene) %>% arrange(fold) %>% summarise(meanfold=mean(fold)) #mean fold for interactions
```

is it because skipping genes are methylated or polycombed H3K27me3? no..

```{r}
tis_corr <- data.frame(tissue=c("embryonicfacialprominence","forebrain","heart","hindbrain","limb","limb","liver","midbrain","neuraltube"),new=c("CF","FB","HR","HB","FL","HL","LV","MB","NT"))

files <- list.files(path="~/enh_capC/encode/h3k27me3", pattern="gz$", recursive=T) 
rm(h3k27me3)
for (path in files){
  name=str_split(path, "_",simplify = T)
    temp_dataset <- read_tsv(gzfile(paste0("~/enh_capC/encode/h3k27me3/",path)),col_names=c("chr","start","end","name","no1","strand","signal","logp","qvalue","no2")) %>% add_column(tissue = name[1,1], stage=name[1,3]) %>% left_join(tis_corr) %>% dplyr::select(chr,start,end,signal,qvalue,tissue=new,stage) %>% distinct()
    if (!exists("h3k27me3")) {
      h3k27me3 <- temp_dataset 
    } else {
      h3k27me3 <- temp_dataset %>% bind_rows(h3k27me3)
    }
    rm(temp_dataset)
}

setDT(h3k27me3)
setkey(h3k27me3,chr,start,end)

refgene_tss_h3k27me3 <- refgene_tss %>% 
  foverlaps(h3k27me3, by.x=c("chr","tss_start", "tss_end"), type="any") %>% group_by(chr,tss_start,tss_end,gene_name,tissue,stage) %>% summarise(qvalue=max(qvalue),signal=max(signal)) %>% ungroup() %>% distinct(chr,across(starts_with("tss")),gene_name,tissue,stage,signal,qvalue) 

enh_tss_exp_tis <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>%
  dplyr::select(starts_with("target"),starts_with("bait"),promoter,gene,int_tis=tissue,score) %>% dplyr::filter(bait_type=="enh") %>% as.data.table() %>%
  foverlaps(refgene_tss, by.x=c("target_chr","target_start", "target_end"), type="any") %>% dplyr::filter(!is.na(gene_name)) %>%
  left_join(rna_seq_e115,by="gene_name") %>% tidyr::replace_na(list(mean_TPM = 0)) %>%
  group_by(across(starts_with("tss")),across(starts_with("bait")),gene_name) %>% mutate(max_TPM=max(mean_TPM)) %>% ungroup() %>% dplyr::filter(max_TPM>0.5) %>% #get the expressing ones
  distinct(bait_chr,tss_start,tss_end,across(starts_with("bait")),gene_name,max_TPM,int_tis) %>% as.data.table()

data1 <- enh_tss_exp_tis %>% 
  group_by(bait_name,gene_name) %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  #dplyr::filter(len==min(len)) %>% #leave all tss for the target gene and take the mean for downstream analysis #get the closest tss for each gene, no
  ungroup() %>% as.data.table() %>%  
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>%
  distinct(bait_name,bait_chr,i.gene_name,gene_name,across(contains("tss")),int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,skipping_tss_start=tss_start,skipping_tss_end=tss_end, target_gene=i.gene_name,target_tss_start=i.tss_start,target_tss_end=i.tss_end,bait_name,bait_chr,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% dplyr::filter(class=="POS") %>% #only positive tissues
  pivot_longer(!c(bait_name,bait_chr,int_tis,skipping_tss_start,skipping_tss_end,target_tss_start,target_tss_end,class),names_to = "type", values_to = "gene_name") %>% #also filter those are target genes in skipping gene list
  mutate(tss_start=if_else(type=="skipping_gene",skipping_tss_start,target_tss_start),tss_end=if_else(type=="skipping_gene",skipping_tss_end,target_tss_end)) %>% dplyr::select(-c(skipping_tss_start,target_tss_start,skipping_tss_end,target_tss_end)) %>% distinct() %>%
  group_by(gene_name,int_tis) %>% mutate(type2=if_else(length(unique(type))==2,1,0)) %>% #after this step, it's gene TSS centric
  ungroup() %>% mutate(type=if_else(type2==1,"target_gene",type)) %>% distinct(bait_chr,type,gene_name,tss_start,tss_end,int_tis) %>%
  left_join(dplyr::filter(refgene_tss_h3k27me3,stage=="e11"),by=c("bait_chr"="chr","tss_start","tss_end","int_tis"="tissue","gene_name")) %>% 
  tidyr::replace_na(list(qvalue = 0, signal = 0)) %>% 
  group_by(gene_name,type,int_tis) %>% summarise(signal=max(signal)) %>% ungroup() %>% #one gene has multiple TSSs, so take a max
  mutate(int_tis=factor(int_tis,levels=c("FB","MB","HB","NT","CF","FL","HL","HR"))) %>% dplyr::filter(int_tis %in% c("FB","MB","HB","NT","CF","FL","HL","HR"))
  
stat.test <- data1 %>% group_by(int_tis) %>% wilcox_test(signal ~ type, paired = F) %>% 
  add_x_position(x="int_tis") %>% adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 3)) %>% mutate(p.adj=if_else(p.adj<0.05,p.adj,"ns"))#actual p-value
 
supfig7a <- data1 %>% 
  ggboxplot(x="int_tis",y="signal",color="type",outlier.shape = NA,add = "jitter",add.params = list(size=0.3,binwidth=5)) +
  stat_pvalue_manual(stat.test, label = "p.adj.signif",y.position=4.2,tip.length = .01,size=2.5) + 
  labs(x="",y="Signal of H3K27me3 peaks at ±2.5 kb\naround TSS in active tissues") +
  scale_y_continuous(trans=pseudo_log_trans(base = 2),limits = c(0,25),breaks = c(0,2^0,2^1,2^2,2^3,2^4)) +
  scale_color_manual(values=c("grey70","#68abdaff"), name="",labels=c("skipping_gene"="Skipped genes","target_gene"="Interacting genes")) +
  theme(axis.text = element_text(size=6),
        axis.title.y=element_text(size=6),
        legend.key.size = unit(2, 'mm'),
        legend.title = element_text(size=6),
        legend.text = element_text(size=5.5))

#chi-square test on pie chart
test <- data1 %>% mutate(pos=if_else(signal>0,"Yes","No")) %>% 
  group_by(int_tis) %>%
  summarise(pval = chisq.test(pos,type)$p.value) %>% adjust_pvalue(method = "fdr") %>% add_significance("pval.adj") 
labs <- test$pval.adj.signif
names(labs) <- test$int_tis
  
xpie <- data1 %>% mutate(pos=if_else(signal>0,"Yes","No")) %>% dplyr::count(int_tis,pos,type) %>% group_by(int_tis,type) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% ggplot( aes(x="", y=perc, fill=pos)) + geom_bar(width = 1, stat = "identity") +
  coord_polar("y", start=0) + theme_void() + 
  facet_wrap(int_tis ~ type, nrow=1, labeller = labeller(int_tis = labs,type=c("skipping_gene"="S","target_gene"="T"), .multi_line=F)) + 
  scale_fill_manual(values=c("#c6ccc2","#dd8581ff"), name="") +
  theme(legend.key.size = unit(2.5, 'mm'),
        legend.text = element_text(size=8),
        strip.background = element_blank())

supfig7a <- ggdraw(insert_xaxis_grob(supfig7a,wrap_plots(xpie), position = "bottom")) #use wrap_plots() to make the facet plot as a whole
supfig7a


```

is it because skipping genes are methylated or polycombed H3K9me3? no..

```{r}
tis_corr <- data.frame(tissue=c("embryonicfacialprominence","forebrain","heart","hindbrain","limb","limb","liver","midbrain","neuraltube"),new=c("CF","FB","HR","HB","FL","HL","LV","MB","NT"))

files <- list.files(path="~/enh_capC/encode/h3k9me3", pattern="gz$", recursive=T) 
rm(h3k9me3)
for (path in files){
  name=str_split(path, "_",simplify = T)
    temp_dataset <- read_tsv(gzfile(paste0("~/enh_capC/encode/h3k9me3/",path)),col_names=c("chr","start","end","name","no1","strand","signal","logp","qvalue","no2")) %>% add_column(tissue = name[1,1], stage=name[1,3]) %>% left_join(tis_corr) %>% dplyr::select(chr,start,end,signal,qvalue,tissue=new,stage) %>% distinct()
    if (!exists("h3k9me3")) {
      h3k9me3 <- temp_dataset 
    } else {
      h3k9me3 <- temp_dataset %>% bind_rows(h3k9me3)
    }
    rm(temp_dataset)
}

setDT(h3k9me3)
setkey(h3k9me3,chr,start,end)

refgene_tss_h3k9me3 <- refgene_tss %>% 
  foverlaps(h3k9me3, by.x=c("chr","tss_start", "tss_end"), type="any") %>% group_by(chr,tss_start,tss_end,gene_name,tissue,stage) %>% summarise(qvalue=max(qvalue),signal=max(signal)) %>% ungroup() %>% distinct(chr,across(starts_with("tss")),gene_name,tissue,stage,signal,qvalue) 

enh_tss_exp_tis <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>%
  dplyr::select(starts_with("target"),starts_with("bait"),promoter,gene,int_tis=tissue,score) %>% dplyr::filter(bait_type=="enh") %>% as.data.table() %>%
  foverlaps(refgene_tss, by.x=c("target_chr","target_start", "target_end"), type="any") %>% dplyr::filter(!is.na(gene_name)) %>%
  left_join(rna_seq_e115,by="gene_name") %>% tidyr::replace_na(list(mean_TPM = 0)) %>%
  group_by(across(starts_with("tss")),across(starts_with("bait")),gene_name) %>% mutate(max_TPM=max(mean_TPM)) %>% ungroup() %>% dplyr::filter(max_TPM>0.5) %>% #get the expressing ones
  distinct(bait_chr,tss_start,tss_end,across(starts_with("bait")),gene_name,max_TPM,int_tis) %>% as.data.table()

data2 <- enh_tss_exp_tis %>% 
  group_by(bait_name,gene_name) %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  ungroup() %>% as.data.table() %>%  
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>%
  distinct(bait_name,bait_chr,i.gene_name,gene_name,across(contains("tss")),int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,skipping_tss_start=tss_start,skipping_tss_end=tss_end, target_gene=i.gene_name,target_tss_start=i.tss_start,target_tss_end=i.tss_end,bait_name,bait_chr,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% dplyr::filter(class=="POS") %>% #only positive tissues
  pivot_longer(!c(bait_name,bait_chr,int_tis,skipping_tss_start,skipping_tss_end,target_tss_start,target_tss_end,class),names_to = "type", values_to = "gene_name") %>% #also filter those are target genes in skipping gene list
  mutate(tss_start=if_else(type=="skipping_gene",skipping_tss_start,target_tss_start),tss_end=if_else(type=="skipping_gene",skipping_tss_end,target_tss_end)) %>% dplyr::select(-c(skipping_tss_start,target_tss_start,skipping_tss_end,target_tss_end)) %>% distinct() %>%
  group_by(gene_name,int_tis) %>% mutate(type2=if_else(length(unique(type))==2,1,0)) %>% 
  ungroup() %>% mutate(type=if_else(type2==1,"target_gene",type)) %>% distinct(bait_chr,type,gene_name,tss_start,tss_end,int_tis) %>%
  left_join(dplyr::filter(refgene_tss_h3k9me3,stage=="e11"),by=c("bait_chr"="chr","tss_start","tss_end","int_tis"="tissue","gene_name")) %>% 
  tidyr::replace_na(list(qvalue = 0, signal = 0)) %>%
  group_by(gene_name,type,int_tis) %>% summarise(signal=max(signal)) %>% ungroup() %>% #one gene has multiple TSSs, so take a max
  mutate(int_tis=factor(int_tis,levels=c("FB","MB","HB","NT","CF","FL","HL","HR"))) %>% dplyr::filter(int_tis %in% c("FB","MB","HB","NT","CF","FL","HL","HR"))
  
supfig7b <- data2 %>% 
  ggboxplot(x="int_tis",y="signal",color="type",outlier.shape = NA,add = "jitter",add.params = list(size=0.3,binwidth=5)) +
   labs(x="",y="Signal of H3k9me3 peaks at ±2.5 kb\naround TSS in active tissues") +
  scale_color_manual(values=c("grey70","#68abdaff"), name="",labels=c("skipping_gene"="Skipped genes","target_gene"="Interacting genes")) +
  theme(axis.text = element_text(size=6),
        axis.title.y=element_text(size=6),
        legend.key.size = unit(2, 'mm'),
        legend.title = element_text(size=6),
        legend.text = element_text(size=5.5))

#there's no way to do chi-square on this, since just <=3 genes have H3K9me3 marks in each tissue
xpie <- data2 %>% mutate(pos=if_else(signal>0,"Yes","No")) %>% dplyr::count(int_tis,pos,type) %>% group_by(int_tis,type) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% ggplot( aes(x="", y=perc, fill=pos)) + geom_bar(width = 1, stat = "identity") + coord_polar("y", start=0) + theme_void() + facet_wrap(int_tis ~ type, nrow=1) + 
  scale_fill_manual(values=c("#c6ccc2","#dd8581ff"), name="") +
  theme(legend.key.size = unit(2.5, 'mm'),
        legend.text = element_text(size=8),
        strip.background = element_blank(),
        strip.text.x = element_blank())

supfig7b <- ggdraw(insert_xaxis_grob(supfig7b,wrap_plots(xpie), position = "bottom")) 

plot_grid(supfig7a,supfig7b,labels = c("A","B"),nrow=1,rel_widths = c(1,1))

library("colorspace") 
#pie chart for proportion 
supfig7i <- dplyr::select(data1,h3k27me3=signal,everything()) %>% left_join(dplyr::select(data2,h3k9me3=signal,everything()), by=c("gene_name","type","int_tis")) %>% left_join(methyl_group,by=c("gene_name","type","int_tis")) %>% dplyr::filter(int_tis != "HR") %>% dplyr::filter(type != "target_gene") %>% mutate(h3k27me3=if_else(h3k27me3==0,0,1),h3k9me3=if_else(h3k9me3==0,0,1),me_cat=if_else(me_cat=="low",0,1)) %>% mutate(sum=h3k27me3+h3k9me3+me_cat) %>% mutate(mark=if_else((sum==1 & h3k27me3==1),"h3k27me3",if_else((sum==1 & h3k9me3==1),"h3k9me3",if_else((sum==1 & me_cat==1),"methyl",if_else(sum==2,"2",if_else(sum==3,"3","no_mark")))))) %>% dplyr::count(mark,int_tis) %>% group_by(int_tis) %>% mutate(perc=n/sum(n), mark=factor(mark,levels = c("no_mark","methyl","h3k27me3","h3k9me3","2","3")),int_tis=factor(int_tis,levels=c("FB","MB","HB","NT","CF","FL","HL"))) %>% ungroup() %>%
  ggplot(aes(x="",y=perc,fill=mark)) + 
  geom_bar(width = 1, stat = "identity", color = "white") + 
  geom_text(aes(label=n, x = 1.2), #change the in or out of the label
    position=position_stack(0.5), #align the label in the middle of each pie
    size=2.5) +   #label size
  coord_polar("y", start=0) + theme_void() +
  facet_wrap(~ int_tis,nrow=2) +
  scale_fill_discrete_qualitative(palette="Set3",name = "",labels = c("No Mark","Methylated","H3K27me3","h3k9me3","2 Marks","3 Marks")) +
  theme(legend.key.size = unit(2.5, 'mm'),legend.position = "top",
        legend.text = element_text(size=6))

supfig7i
```

methylation

```{r}
tis_corr <- data.frame(tissue=c("embryonicfacialprominence","forebrain","heart","hindbrain","limb","limb","liver","midbrain","neuraltube"),new=c("CF","FB","HR","HB","FL","HL","LV","MB","NT"))

methyl <- read_tsv("~/enh_capC/encode/WGBS/ref_tss_methyl_sum_1kb.tsv") %>% dplyr::filter(stage=="E115") %>% group_by(tissue,chr,tss_start,tss_end) %>% summarise(mean_methyl=mean(mean_methyl)) %>% left_join(tis_corr,by="tissue") %>% dplyr::select(tissue=new,chr,tss_start,tss_end,mean_methyl) %>% mutate(tss_start=tss_start-1500,tss_end=tss_end+1500) 

enh_tss_exp %>% group_by(bait_name) %>%
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  dplyr::filter(len==min(len)) %>% 
  ungroup() %>% as.data.table() %>%  #only get closest enh-gene pairs for each enhancer
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>%
  distinct(bait_name,bait_chr,i.gene_name,gene_name,across(contains("tss"))) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,skipping_tss_start=tss_start,skipping_tss_end=tss_end, target_gene=i.gene_name,target_tss_start=i.tss_start,target_tss_end=i.tss_end,bait_name,bait_chr) %>% dplyr::filter(!is.na(skipping_gene)) %>%  
  left_join(bait_enh_act,by=c("bait_name"="elementID")) %>% complete(nesting(skipping_gene,skipping_tss_start,skipping_tss_end, target_gene,target_tss_start,target_tss_end,bait_name,bait_chr),tissue=tis,fill=list(class="NEG")) %>% #get both positive and negative tissues
  left_join(methyl,by=c("bait_chr"="chr","target_tss_start"="tss_start","target_tss_end"="tss_end","tissue")) %>% dplyr::select(target_me=mean_methyl,everything()) %>%
  left_join(methyl,by=c("bait_chr"="chr","skipping_tss_start"="tss_start","skipping_tss_end"="tss_end","tissue")) %>% dplyr::select(skip_me=mean_methyl,everything()) %>%
  dplyr::filter(!is.na(skip_me)) %>% dplyr::filter(!is.na(target_me)) %>% #some of the tissues or genes weren't detected in encode methyl data, so remove them

  group_by(bait_name,target_gene,skipping_gene,class) %>% #no need to 
  summarise(mean_skipping_me=mean(skip_me),mean_target_me=mean(target_me)) %>% ungroup() %>% #group by bait_gene_skip pairs
  pivot_longer(!c(bait_name,target_gene,skipping_gene,class),names_to = "group", values_to = "methyl") %>%
  unite("code",c("group","class"),remove = F) %>% 
  ungroup() %>% mutate(code=factor(code,levels=c("mean_skipping_me_NEG","mean_skipping_me_POS","mean_target_me_NEG","mean_target_me_POS"))) %>%
  #distinct(bait_name,target_gene) %>% nrow() #217
  ggboxplot(x="code",y="methyl",color="class",outlier.size = 0.5) +  
  stat_compare_means(label = "p.format", comparisons = list(c("mean_skipping_me_NEG","mean_target_me_NEG"), c("mean_skipping_me_POS","mean_target_me_POS")),method = "wilcox.test",paired = F,tip.length = .01,label.y = c(105,115),size=4) + #???paired or not
  stat_compare_means(label = "p.format", comparisons = list(c("mean_skipping_me_NEG","mean_skipping_me_POS"), c("mean_target_me_NEG","mean_target_me_POS")),method = "wilcox.test",paired = F,tip.length = .01,label.y = 95,size=4) + 
  labs(x="",y="CpG methylation (%)",title = "CpG methylation 1kb around TSS") +
  scale_x_discrete(labels=c("Skipping_genes","Skipping_genes","Target_genes","Target_genes")) +
  scale_color_manual(values=c("#ec7f26e3","#4d77b1e8"),labels=c("NEG"="inactive_tissues","POS"="active_tissues"))
```

plot methylation with different window size

```{r}
tis_corr <- data.frame(tissue=c("embryonicfacialprominence","forebrain","heart","hindbrain","limb","limb","liver","midbrain","neuraltube"),new=c("CF","FB","HR","HB","FL","HL","LV","MB","NT"))

methyl <- read_tsv("~/enh_capC/encode/WGBS/ref_tss_methyl_E115_2.5kb.tsv",col_names = c("tissue","stage","rep","chr","tss_start","tss_end","gene_name","cpg_start","cpg_end","read_num","methyl_perc")) %>%
  #dplyr::filter(stage=="E115") %>% 
  mutate(tss_mid=(tss_start+tss_end)/2) %>% mutate(dist=if_else((cpg_start>tss_mid-250 & cpg_start<tss_mid+250), 0.25, if_else((cpg_start>tss_mid-500 & cpg_start<tss_mid+500), 0.5, if_else((cpg_start>tss_mid-1000 & cpg_start<tss_mid+1000), 1, 2.5))))

methyl_025 <- methyl %>% dplyr::filter(dist<=0.25) %>%
  group_by(tissue,rep,chr,tss_start,tss_end,gene_name) %>% summarise(mean_methyl=mean(methyl_perc)) %>% group_by(tissue,chr,tss_start,tss_end) %>% summarise(mean_methyl=mean(mean_methyl)) %>% left_join(tis_corr,by="tissue") %>% dplyr::select(tissue=new,chr,tss_start,tss_end,mean_methyl) 

methyl_05 <- methyl %>% dplyr::filter(dist<=0.5) %>%
  group_by(tissue,rep,chr,tss_start,tss_end,gene_name) %>% summarise(mean_methyl=mean(methyl_perc)) %>% group_by(tissue,chr,tss_start,tss_end) %>% summarise(mean_methyl=mean(mean_methyl)) %>% left_join(tis_corr,by="tissue") %>% dplyr::select(tissue=new,chr,tss_start,tss_end,mean_methyl) 

methyl_1k <- methyl %>% dplyr::filter(dist<=1) %>%
  group_by(tissue,rep,chr,tss_start,tss_end,gene_name) %>% summarise(mean_methyl=mean(methyl_perc)) %>% group_by(tissue,chr,tss_start,tss_end) %>% summarise(mean_methyl=mean(mean_methyl)) %>% left_join(tis_corr,by="tissue") %>% dplyr::select(tissue=new,chr,tss_start,tss_end,mean_methyl) 

methyl_2k <- methyl %>% dplyr::filter(dist<=2.5) %>%
  group_by(tissue,rep,chr,tss_start,tss_end,gene_name) %>% summarise(mean_methyl=mean(methyl_perc)) %>% group_by(tissue,chr,tss_start,tss_end) %>% summarise(mean_methyl=mean(mean_methyl)) %>% left_join(tis_corr,by="tissue") %>% dplyr::select(tissue=new,chr,tss_start,tss_end,mean_methyl) 
```

```{r}
#plot 250bp and 2kb
data1 <- enh_tss_exp_tis %>% 
  group_by(bait_name,gene_name) %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  ungroup() %>% as.data.table() %>%  
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>%
  distinct(bait_name,bait_chr,i.gene_name,gene_name,across(contains("tss")),int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,skipping_tss_start=tss_start,skipping_tss_end=tss_end, target_gene=i.gene_name,target_tss_start=i.tss_start,target_tss_end=i.tss_end,bait_name,bait_chr,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% dplyr::filter(class=="POS") %>% #only positive tissues
  pivot_longer(!c(bait_name,bait_chr,int_tis,skipping_tss_start,skipping_tss_end,target_tss_start,target_tss_end,class),names_to = "type", values_to = "gene_name") %>% #also filter those are target genes in skipping gene list
  mutate(tss_start=if_else(type=="skipping_gene",skipping_tss_start,target_tss_start),tss_end=if_else(type=="skipping_gene",skipping_tss_end,target_tss_end)) %>% dplyr::select(-c(skipping_tss_start,target_tss_start,skipping_tss_end,target_tss_end)) %>% distinct() %>%
  group_by(gene_name,int_tis) %>% mutate(type2=if_else(length(unique(type))==2,1,0)) %>%
  ungroup() %>% mutate(type=if_else(type2==1,"target_gene",type)) %>% distinct(bait_chr,type,gene_name,tss_start,tss_end,int_tis) %>%
  left_join(methyl_025,by=c("bait_chr"="chr","tss_start","tss_end","int_tis"="tissue")) %>%  #methyl_025,methyl_05,methyl_1k,methyl_2k
  
  dplyr::filter(!is.na(mean_methyl)) %>% #some of the tissues or genes weren't detected in encode methyl data, so remove them
  group_by(gene_name,type,int_tis) %>% summarise(methyl=mean(mean_methyl)) %>% ungroup() %>% #one gene has multiple TSSs, so take a mean
  mutate(int_tis=factor(int_tis,levels=c("FB","MB","HB","NT","CF","FL","HL","HR"))) %>% dplyr::filter(int_tis != "HR")
  

data2 <- enh_tss_exp_tis %>% 
  group_by(bait_name,gene_name) %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  #dplyr::filter(len==min(len)) %>% #leave all tss for the target gene and take the mean for downstream analysis #get the closest tss for each gene, no
  ungroup() %>% as.data.table() %>%  
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>%
  distinct(bait_name,bait_chr,i.gene_name,gene_name,across(contains("tss")),int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,skipping_tss_start=tss_start,skipping_tss_end=tss_end, target_gene=i.gene_name,target_tss_start=i.tss_start,target_tss_end=i.tss_end,bait_name,bait_chr,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% dplyr::filter(class=="POS") %>% #only positive tissues
  pivot_longer(!c(bait_name,bait_chr,int_tis,skipping_tss_start,skipping_tss_end,target_tss_start,target_tss_end,class),names_to = "type", values_to = "gene_name") %>% 
  mutate(tss_start=if_else(type=="skipping_gene",skipping_tss_start,target_tss_start),tss_end=if_else(type=="skipping_gene",skipping_tss_end,target_tss_end)) %>% dplyr::select(-c(skipping_tss_start,target_tss_start,skipping_tss_end,target_tss_end)) %>% distinct() %>%
  group_by(gene_name,int_tis) %>% mutate(type2=if_else(length(unique(type))==2,1,0)) %>%
  ungroup() %>% mutate(type=if_else(type2==1,"target_gene",type)) %>% distinct(bait_chr,type,gene_name,tss_start,tss_end,int_tis) %>%
  left_join(methyl_2k,by=c("bait_chr"="chr","tss_start","tss_end","int_tis"="tissue")) %>%  #methyl_025,methyl_05,methyl_1k,methyl_2k
  dplyr::filter(!is.na(mean_methyl)) %>% #some of the tissues or genes weren't detected in encode methyl data, so remove them
  group_by(gene_name,type,int_tis) %>% summarise(methyl=mean(mean_methyl)) %>% ungroup() %>% #one gene has multiple TSSs, so take a mean
  mutate(int_tis=factor(int_tis,levels=c("FB","MB","HB","NT","CF","FL","HL","HR"))) %>% dplyr::filter(int_tis != "HR")
  
#250bp
stat.test <- data1 %>% left_join(dplyr::select(methyl_group,-methyl),by=c("gene_name","type","int_tis")) %>% group_by(int_tis) %>% wilcox_test(methyl ~ me_cat, paired = F) %>% 
  add_x_position() %>% add_y_position(step.increase = 0.02) %>%
  dplyr::filter(group2=="int_gene") %>%
  adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 2)) #actual p-value

supfig7g <- data1 %>% left_join(dplyr::select(methyl_group,-methyl),by=c("gene_name","type","int_tis")) %>% ggviolin(x="me_cat",y="methyl", fill="me_cat", trim=T,scale="width", width = .5) +  
  geom_boxplot(outlier.shape = NA,width=0.05) +
  facet_wrap( ~ int_tis,nrow=1) +
  labs(x="",y="CpG methylation (%) at TSSs (±250bp)") +
  scale_fill_manual(values=c("grey70","grey70","#68abdaff")) +
  scale_x_discrete(labels=c("high"="methylated\nSkipped","low"="ummethylated\nSkipped","int_gene"="Interacting")) +
  stat_pvalue_manual(stat.test, label = "p.adj.signif", tip.length = .01,size=2.5,vjust=-1,hjust=0.5) + 
  theme_bw() +
  theme(axis.text.y = element_text(size=6),axis.text.x = element_text(size=6,angle = 30, vjust = 1, hjust = 1),
        axis.title.y=element_text(size=6),
        legend.position = "none",
        strip.text.x = element_text(size = 6.5))

xpie <- data1 %>% left_join(dplyr::select(methyl_group,-methyl),by=c("gene_name","type","int_tis")) %>% mutate(pos=if_else(methyl>50,"Yes","No")) %>% dplyr::count(int_tis,pos,me_cat) %>% group_by(int_tis,me_cat) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% ggplot( aes(x="", y=perc, fill=pos)) + geom_bar(width = 1, stat = "identity") + coord_polar("y", start=0) + theme_void() + facet_wrap(int_tis ~ me_cat, nrow=1) + 
  scale_fill_manual(values=c("#c6ccc2","#dd8581ff"), name="") +
  theme(legend.key.size = unit(2.5, 'mm'),
        legend.text = element_text(size=6),legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank())
supfig7g <- plot_grid(supfig7g,xpie,ncol=1,rel_heights = c(10,1),align = "v")
supfig7g
  


#2kb
stat.test <- data2 %>% left_join(dplyr::select(methyl_group,-methyl),by=c("gene_name","type","int_tis")) %>% group_by(int_tis) %>% wilcox_test(methyl ~ me_cat, paired = F) %>% 
  add_x_position() %>% add_y_position(step.increase = 0.02) %>%
  dplyr::filter(group2=="int_gene") %>%
  adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 2)) #actual p-value

supfig7h <- data2 %>% left_join(dplyr::select(methyl_group,-methyl),by=c("gene_name","type","int_tis")) %>% ggviolin(x="me_cat",y="methyl", fill="me_cat", trim=T,scale="width", width = .5) +
  geom_boxplot(outlier.shape = NA,width=0.05) +
  facet_wrap( ~ int_tis,nrow=1) +
  labs(x="",y="CpG methylation (%) at TSSs (±2kb)") +
  scale_fill_manual(values=c("grey70","grey70","#68abdaff")) +
  scale_x_discrete(labels=c("high"="methylated\nSkipped","low"="ummethylated\nSkipped","int_gene"="Interacting")) +
  stat_pvalue_manual(stat.test, label = "p.adj.signif", tip.length = .01,size=2.5,vjust=-1,hjust=0.5) + 
  theme_bw() +
  theme(axis.text.y = element_text(size=6),axis.text.x = element_text(size=6,angle = 30, vjust = 1, hjust = 1),
        axis.title.y=element_text(size=6),
        legend.position = "none",
        strip.text.x = element_text(size = 6.5))

xpie <- data2 %>% left_join(dplyr::select(methyl_group,-methyl),by=c("gene_name","type","int_tis")) %>% mutate(pos=if_else(methyl>50,"Yes","No")) %>% dplyr::count(int_tis,pos,me_cat) %>% group_by(int_tis,me_cat) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% ggplot( aes(x="", y=perc, fill=pos)) + geom_bar(width = 1, stat = "identity") + coord_polar("y", start=0) + theme_void() + facet_wrap(int_tis ~ me_cat, nrow=1) + 
  scale_fill_manual(values=c("#c6ccc2","#dd8581ff"), name="") +
  theme(legend.key.size = unit(2.5, 'mm'),
        legend.text = element_text(size=6),legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank())
supfig7h <- plot_grid(supfig7h,xpie,ncol=1,rel_heights = c(10,1),align = "v")

supfig7gh <- plot_grid(supfig7g,supfig7h,nrow=1,rel_widths = c(1,1),align = "h",labels = c("H","I"),label_size = 12)
#ggsave("../manuscript/Supp_fig9_2.pdf",width = 183, height = 60, units = "mm")
supfig7gh

```

split skipping genes into high and low methylated classes and plot
dnase/rnaseq, broadness of expression

```{r}
methyl_group <- enh_tss_exp_tis %>% #dplyr::filter(int_tis=="FB") %>% 
  group_by(bait_name,gene_name) %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  ungroup() %>% as.data.table() %>%  
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>%
  distinct(bait_name,bait_chr,i.gene_name,gene_name,across(contains("tss")),int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,skipping_tss_start=tss_start,skipping_tss_end=tss_end, target_gene=i.gene_name,target_tss_start=i.tss_start,target_tss_end=i.tss_end,bait_name,bait_chr,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% dplyr::filter(class=="POS") %>% #only positive tissues
  pivot_longer(!c(bait_name,bait_chr,int_tis,skipping_tss_start,skipping_tss_end,target_tss_start,target_tss_end,class),names_to = "type", values_to = "gene_name") %>% #also filter those are target genes in skipping gene list
  mutate(tss_start=if_else(type=="skipping_gene",skipping_tss_start,target_tss_start),tss_end=if_else(type=="skipping_gene",skipping_tss_end,target_tss_end)) %>% dplyr::select(-c(skipping_tss_start,target_tss_start,skipping_tss_end,target_tss_end)) %>% distinct() %>%
  group_by(gene_name,int_tis) %>%  mutate(type2=if_else(length(unique(type))==2,1,0)) %>% 
  ungroup() %>% mutate(type=if_else(type2==1,"target_gene",type)) %>% distinct(bait_chr,type,gene_name,tss_start,tss_end,int_tis) %>%
  left_join(methyl_1k,by=c("bait_chr"="chr","tss_start","tss_end","int_tis"="tissue")) %>%  #methyl_025,methyl_05,methyl_1k,methyl_2k
  dplyr::filter(!is.na(mean_methyl)) %>% 
  group_by(gene_name,type,int_tis) %>% summarise(methyl=mean(mean_methyl)) %>% ungroup() %>% 
  dplyr::filter(int_tis != "HR") %>% mutate(me_cat=if_else(type=="target_gene","int_gene",if_else(methyl>50,"high","low"))) %>% mutate(me_cat=factor(me_cat,levels = c("high","low","int_gene")))

#pie chart, methylation state is different among tissues
methyl_group %>% dplyr::count(int_tis,me_cat) %>% dplyr::filter(me_cat != "int_gene") %>% group_by(int_tis) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% ggplot( aes(x="", y=perc, fill=me_cat)) + geom_bar(width = 1, stat = "identity", color = "white") + coord_polar("y", start=0) + theme_void() + facet_wrap(~ int_tis)  

#dnase
data <- methyl_group %>% left_join(dnase,by=c("int_tis"="tissue","gene_name")) %>% mutate(int_tis=factor(int_tis,levels=c("FB","MB","HB","NT","CF","FL","HL"))) %>%  group_by(gene_name,int_tis,methyl,me_cat) %>% summarise(dnase=mean(ave_depth)) %>% ungroup()

data1 <- data %>% dplyr::filter(int_tis %in% c("FB","CF","FL")) %>% mutate(int_tis=factor(int_tis,levels=c("FB","CF","FL")))

stat.test <- data1 %>% group_by(int_tis) %>% wilcox_test(dnase ~ me_cat, paired = F) %>% 
  add_x_position() %>% 
  add_y_position(step.increase = 0.02) %>%
  adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>% #dplyr::filter(p.adj<0.05) %>%
  dplyr::filter(group2=="int_gene") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 2)) #actual p-value

fig4k <- data1 %>% ggviolin(x="me_cat",y="dnase",fill="me_cat",trim=T,#draw_quantiles = 0.5,
                            scale="width",width = .5) + 
  geom_boxplot(outlier.shape = NA,width=0.05) +
  facet_wrap( ~ int_tis,nrow=1) +
  scale_fill_manual(values=c("grey70","grey70","#68abdaff")) + 
  scale_x_discrete(labels=c("high"="methylated\nSkipped","low"="ummethylated\nSkipped","int_gene"="Interacting")) +
  stat_pvalue_manual(stat.test, label = "p.adj", tip.length = .01,size=2.5,vjust=-1,hjust=0.5) + 
  labs(x="",y="DNase I cleavge at TSSs") +
  theme_bw() +
  theme(axis.text.y = element_text(size=6),axis.text.x = element_text(size=6,angle = 30, vjust = 1, hjust = 1),
        axis.title.y=element_text(size=6),
        legend.position = "none",
        strip.text.x = element_text(size = 6.5))
fig4k

data2 <- data %>% dplyr::filter(int_tis %in% c("MB","HB","NT","HL")) %>% mutate(int_tis=factor(int_tis,levels=c("MB","HB","NT","HL")))

stat.test <- data2 %>% group_by(int_tis) %>% wilcox_test(dnase ~ me_cat, paired = F) %>% 
  add_x_position() %>% 
  add_y_position(step.increase = 0.02) %>%
  adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>% 
  dplyr::filter(group2=="int_gene") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 2)) 

supfig7e <- data2 %>% ggviolin(x="me_cat",y="dnase",fill="me_cat",trim=T,
                            scale="width",width = .5) + 
  geom_boxplot(outlier.shape = NA,width=0.05) +
  facet_wrap( ~ int_tis,nrow=1) +
  scale_fill_manual(values=c("grey70","grey70","#68abdaff")) + 
  scale_x_discrete(labels=c("high"="methylated\nSkipped","low"="ummethylated\nSkipped","int_gene"="Interacting")) +
  stat_pvalue_manual(stat.test, label = "p.adj", tip.length = .01,size=2.5,vjust=-1,hjust=0.5) + 
  labs(x="",y="DNase I cleavge at TSSs") +
  theme_bw() +
  theme(axis.text.y = element_text(size=6),axis.text.x = element_text(size=6,angle = 30, vjust = 1, hjust = 1),
        axis.title.y=element_text(size=6),
        legend.position = "none",
        strip.text.x = element_text(size = 6.5))
supfig7e
```

```{r}
#rnaseq
data <- methyl_group %>% left_join(rna_seq_e115,by=c("gene_name","int_tis"="tissue")) %>% dplyr::filter(!is.na(mean_TPM)) %>% distinct(gene_name,int_tis,mean_TPM,me_cat) 
  
data1 <- data %>% dplyr::filter(int_tis %in% c("FB","CF","FL")) %>% mutate(int_tis=factor(int_tis,levels=c("FB","CF","FL")))
#count forebrain skipped gene fraction
data1 %>% dplyr::filter(int_tis=="FB") %>% dplyr::filter(me_cat != "int_gene") %>% mutate(exp=if_else(mean_TPM<0.5, "NO","YES")) %>% dplyr::count(exp,me_cat) %>% mutate(perc=n/sum(n))

stat.test <- data1 %>% group_by(int_tis) %>% wilcox_test(mean_TPM ~ me_cat, paired = F) %>% 
  add_x_position() %>% 
  add_y_position(y.trans=log2) %>%
  adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>% #dplyr::filter(p.adj<0.05) %>%
  dplyr::filter(group2=="int_gene") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 2)) #actual p-value

fig4j <- data1 %>% ggviolin(x="me_cat",y="mean_TPM",fill="me_cat",trim=T,#draw_quantiles = 0.5,
                            scale="width",width = .5) + 
  geom_boxplot(outlier.shape = NA,width=0.05) +
  facet_wrap( ~ int_tis,nrow=1) +
  scale_fill_manual(values=c("grey70","grey70","#68abdaff")) +   
  scale_x_discrete(labels=c("high"="methylated\nSkipped","low"="ummethylated\nSkipped","int_gene"="Interacting")) +
  stat_pvalue_manual(stat.test, label = "p.adj.signif", tip.length = .01,size=2.5,vjust=-1,hjust=0.5) +
  scale_y_continuous(trans=pseudo_log_trans(base = 2), breaks = scales::trans_breaks("log2", function(x) 2^x) (c(1, 1000)), labels = scales::trans_format("log2", scales::math_format(2^.x))) + #use pseudo scale to preserve 0 values
  labs(x="",y="TPM") + theme_bw() +
  theme(axis.text.y = element_text(size=6),axis.text.x = element_text(size=6,angle = 30, vjust = 1, hjust = 1),
        axis.title.y=element_text(size=6),
        legend.position = "none",
        strip.text.x = element_text(size = 6.5))
fig4j

data2 <- data %>% dplyr::filter(int_tis %in% c("MB","HB","NT","HL")) %>% mutate(int_tis=factor(int_tis,levels=c("MB","HB","NT","HL")))

stat.test <- data2 %>% group_by(int_tis) %>% wilcox_test(mean_TPM ~ me_cat, paired = F) %>% 
  add_x_position() %>% 
  add_y_position(y.trans=log2) %>%
  adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>% #dplyr::filter(p.adj<0.05) %>%
  dplyr::filter(group2=="int_gene") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 2)) #actual p-value

supfig7d <- data2 %>% ggviolin(x="me_cat",y="mean_TPM",fill="me_cat",trim=T,#draw_quantiles = 0.5,
                            scale="width",width = .5) + 
  geom_boxplot(outlier.shape = NA,width=0.05) +
  facet_wrap( ~ int_tis,nrow=1) +
  scale_fill_manual(values=c("grey70","grey70","#68abdaff")) +   
  scale_x_discrete(labels=c("high"="methylated\nSkipped","low"="ummethylated\nSkipped","int_gene"="Interacting")) +
  stat_pvalue_manual(stat.test, label = "p.adj.signif", tip.length = .01,size=2.5,vjust=-1,hjust=0.5) +
  scale_y_continuous(trans=pseudo_log_trans(base = 2), breaks = scales::trans_breaks("log2", function(x) 2^x) (c(1, 1000)), labels = scales::trans_format("log2", scales::math_format(2^.x))) + #use pseudo scale to preserve 0 values
  labs(x="",y="TPM") + theme_bw() +
  theme(axis.text.y = element_text(size=6),axis.text.x = element_text(size=6,angle = 30, vjust = 1, hjust = 1),
        axis.title.y=element_text(size=6),
        legend.position = "none",
        strip.text.x = element_text(size = 6.5))
supfig7d

#broadness of expression
methyl_group %>% mutate(exp_tis=sample(c("FB","MB","HB","NT","CF","FL","HL"),nrow(.), replace = TRUE)) %>% complete(nesting(gene_name,me_cat,int_tis), exp_tis) %>% left_join(rna_seq_e115,by=c("gene_name","exp_tis"="tissue")) %>% dplyr::filter(!is.na(mean_TPM)) %>% mutate(exp=if_else(mean_TPM>0.5,1,0)) %>% distinct(gene_name,int_tis,exp,me_cat,exp_tis) %>% dplyr::count(gene_name,int_tis,me_cat,wt=exp,name="broad") %>% dplyr::count(int_tis,me_cat,broad) %>% group_by(int_tis,me_cat) %>% mutate(perc=n/sum(n)) %>% ungroup() %>%
  ggplot( aes(x="", y=perc, fill=broad)) + geom_bar(width = 1, stat = "identity", color = "white") + coord_polar("y", start=0) + theme_void() + facet_wrap(me_cat~ int_tis,nrow=3)  + scale_fill_gradient("Tissues No.",low = "#deebf7ff", high = "#2171b5ff") +
   theme(axis.text = element_text(size=6),
        axis.title.y=element_text(size=6),
        legend.key.size = unit(2, 'mm'),
        legend.title = element_text(size=6),
        legend.text = element_text(size=5.5))
```


```{r}
cpg <- read_tsv("~/enh_capC/encode/mm10_cpg_island.tsv") %>% dplyr::select(chrom,chromStart,chromEnd,length,cpgNum,perCpg) %>% as.data.table()
setkey(cpg,chrom,chromStart,chromEnd)

#cpg island
data <- enh_tss_exp_tis %>% 
  group_by(bait_name,gene_name) %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  ungroup() %>% as.data.table() %>%  
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>%
  distinct(bait_name,bait_chr,i.gene_name,gene_name,across(contains("tss")),int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,skipping_tss_start=tss_start,skipping_tss_end=tss_end, target_gene=i.gene_name,target_tss_start=i.tss_start,target_tss_end=i.tss_end,bait_name,bait_chr,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% dplyr::filter(class=="POS") %>% #only positive tissues
  pivot_longer(!c(bait_name,bait_chr,int_tis,skipping_tss_start,skipping_tss_end,target_tss_start,target_tss_end,class),names_to = "type", values_to = "gene_name") %>% #also filter those are target genes in skipping gene list
  mutate(tss_start=if_else(type=="skipping_gene",skipping_tss_start,target_tss_start),tss_end=if_else(type=="skipping_gene",skipping_tss_end,target_tss_end)) %>% dplyr::select(-c(skipping_tss_start,target_tss_start,skipping_tss_end,target_tss_end)) %>% distinct() %>%
  group_by(gene_name,int_tis) %>% mutate(type2=if_else(length(unique(type))==2,1,0)) %>%
  ungroup() %>% mutate(type=if_else(type2==1,"target_gene",type)) %>% distinct(bait_chr,type,gene_name,tss_start,tss_end,int_tis) %>%
  mutate(tss_start=tss_start-1000,tss_end=tss_end+1000) %>% #extend 5kb
  as.data.table() %>% #join cpg island with different distance
  foverlaps(cpg, by.x=c("bait_chr","tss_start", "tss_end"), type="any") %>%
  mutate(across(length:perCpg, ~if_else(is.na(.),0,.))) %>% 
  group_by(type,gene_name) %>% summarise(perCpg=max(perCpg),length=max(length),cpgNum=sum(cpgNum)) %>% ungroup() %>% right_join(methyl_group,by=c("gene_name","type")) %>% 
  mutate(int_tis=factor(int_tis,levels=c("FB","MB","HB","NT","CF","FL","HL")))

stat.test <- data %>% group_by(int_tis) %>% wilcox_test(length ~ me_cat, paired = F) %>% 
  add_x_position() %>% 
  add_y_position(y.trans=log2) %>%
  adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>% #dplyr::filter(p.adj<0.05) %>%
  dplyr::filter(group2=="int_gene") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 2)) 

supfig7f <- data %>% ggviolin(x="me_cat",y="length",fill="me_cat",trim=T,#draw_quantiles = 0.5,
                              scale="width",width = .5) + 
  geom_boxplot(outlier.shape = NA,width=0.05) +
  facet_wrap( ~ int_tis,nrow=1) +
  scale_fill_manual(values=c("grey70","grey70","#68abdaff")) + 
  scale_x_discrete(labels=c("high"="methylated\nSkipped","low"="ummethylated\nSkipped","int_gene"="Interacting")) +
  stat_pvalue_manual(stat.test, label = "p.adj.signif", tip.length = .01,size=2.5,vjust=-1,hjust=0.5) +
  scale_y_continuous(trans=pseudo_log_trans(base = 2), breaks = scales::trans_breaks("log2", function(x) 2^x) (c(1, 10000)), labels = scales::trans_format("log2", scales::math_format(2^.x))) + #use pseudo scale to preserve 0 values
  labs(x="",y="CpG length (bp)") +
  theme_bw() +
  theme(axis.text.y = element_text(size=6),axis.text.x = element_text(size=6,angle = 30, vjust = 1, hjust = 1),
        axis.title.y=element_text(size=6),
        legend.position = "none",
        strip.text.x = element_text(size = 6.5))


xpie <- data %>% mutate(perCpg=if_else(perCpg==0,"No","Yes")) %>% #as 0 and 1
  dplyr::count(int_tis,me_cat,perCpg) %>%
  group_by(int_tis,me_cat) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% ggplot( aes(x="int_tis", y=perc, fill=perCpg)) + geom_bar(width = 1, stat = "identity") + coord_polar("y", start=0) + theme_void() + facet_wrap(int_tis ~ me_cat,nrow=1) + 
  scale_fill_manual(values=c("#c6ccc2","#dd8581ff"), name="Overlap with CpG") +
  theme(legend.key.size = unit(2.5, 'mm'),
        legend.text = element_text(size=6),legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank(),
        legend.position = "bottom")

supfig7f <- plot_grid(supfig7f,xpie,ncol=1,rel_heights = c(10,1),align = "v") 
supfig7f

data %>% mutate(perCpg=if_else(perCpg==0,"No","Yes")) %>% #as 0 and 1
  dplyr::count(int_tis,me_cat,perCpg) %>% group_by(int_tis,me_cat) %>% mutate(perc=n/sum(n)) %>% ungroup() %>%
   dplyr::filter(me_cat=="low" & perCpg=="Yes") %>% arrange(perc)
```

DNA methylation

```{r}
data <- enh_tss_exp_tis %>% #dplyr::filter(int_tis=="FB") %>% 
  group_by(bait_name,gene_name) %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  ungroup() %>% as.data.table() %>%  
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>%
  distinct(bait_name,bait_chr,i.gene_name,gene_name,across(contains("tss")),int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,skipping_tss_start=tss_start,skipping_tss_end=tss_end, target_gene=i.gene_name,target_tss_start=i.tss_start,target_tss_end=i.tss_end,bait_name,bait_chr,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% dplyr::filter(class=="POS") %>% #only positive tissues
  pivot_longer(!c(bait_name,bait_chr,int_tis,skipping_tss_start,skipping_tss_end,target_tss_start,target_tss_end,class),names_to = "type", values_to = "gene_name") %>% #also filter those are target genes in skipping gene list
  mutate(tss_start=if_else(type=="skipping_gene",skipping_tss_start,target_tss_start),tss_end=if_else(type=="skipping_gene",skipping_tss_end,target_tss_end)) %>% dplyr::select(-c(skipping_tss_start,target_tss_start,skipping_tss_end,target_tss_end)) %>% distinct() %>%
  group_by(gene_name,int_tis) %>% mutate(type2=if_else(length(unique(type))==2,1,0)) %>% #after this step, it's gene TSS centric, because in the same tissue, there would be multiple active enhancers and some skipping genes are the target genes for other active enhancers
  ungroup() %>% mutate(type=if_else(type2==1,"target_gene",type)) %>% distinct(bait_chr,type,gene_name,tss_start,tss_end,int_tis) %>%
  left_join(methyl_1k,by=c("bait_chr"="chr","tss_start","tss_end","int_tis"="tissue")) %>%  #methyl_025,methyl_05,methyl_1k,methyl_2k
  
  dplyr::filter(!is.na(mean_methyl)) %>% #some of the tissues or genes weren't detected in encode methyl data, so remove them
  group_by(gene_name,type,int_tis) %>% summarise(methyl=mean(mean_methyl)) %>% ungroup() %>% #one gene has multiple TSSs, so take a mean
  dplyr::filter(int_tis != "HR")

data1 <- data %>% dplyr::filter(int_tis %in% c("FB","CF","FL")) %>% mutate(int_tis=factor(int_tis,levels=c("FB","CF","FL")), type=factor(type,levels=c("skipping_gene","target_gene")))

stat.test <- data1 %>% group_by(int_tis) %>% wilcox_test(methyl ~ type, paired = F) %>% 
  add_x_position() %>% adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 2)) #actual p-value
 
fig4i <- data1 %>% 
  ggviolin(x="type",y="methyl", fill="type", trim=T,scale="width", width = .5) +  #,alpha=0.8
  geom_boxplot(outlier.shape = NA,width=0.05) +
  facet_wrap( ~ int_tis,nrow=1) +
  stat_pvalue_manual(stat.test, label = "p.adj.signif",y.position=100,tip.length = .01,size=2.5) + 
  labs(x="",y="CpG methylation (%) at TSSs") +
  scale_fill_manual(values=c("grey70","#68abdaff")) +
  scale_x_discrete(labels=c("skipping_gene"="Skipped","target_gene"="Interacting")) +
  coord_cartesian(ylim= c(0,110)) +
  theme_bw() +
  theme(axis.text.y = element_text(size=6),axis.text.x = element_text(size=6,angle = 30, vjust = 1, hjust = 1),
        axis.title.y=element_text(size=6),
        legend.position = "none",
        strip.text.x = element_text(size = 6.5))


xpie <- data1 %>% mutate(pos=if_else(methyl>50,"Yes","No")) %>% dplyr::count(int_tis,pos,type) %>% group_by(int_tis,type) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% ggplot( aes(x="", y=perc, fill=pos)) + geom_bar(width = 1, stat = "identity") + coord_polar("y", start=0) + theme_void() + facet_wrap(int_tis ~ type, nrow=1) + 
  scale_fill_manual(values=c("#c6ccc2","#dd8581ff"), name="") +
  theme(legend.key.size = unit(2.5, 'mm'),
        legend.text = element_text(size=6),legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank())

fig4i <- plot_grid(fig4i,xpie,ncol=1,rel_heights = c(10,1),align = "v")


#how many fold to skipped genes on average
#data %>% group_by(type,int_tis) %>% summarise(mean_tis=mean(methyl)) %>% pivot_wider(names_from = "type",values_from = "mean_tis") %>% mutate(fold=skipping_gene/target_gene) %>% arrange(fold) #%>% summarise(meanfold=mean(fold))

#gene number
#data %>% distinct(gene_name,int_tis,type) %>% dplyr::count(int_tis,type)

data2 <- data %>% dplyr::filter(int_tis %in% c("MB","HB","NT","HL")) %>% mutate(int_tis=factor(int_tis,levels=c("MB","HB","NT","HL")), type=factor(type,levels=c("skipping_gene","target_gene")))

stat.test <- data2 %>% group_by(int_tis) %>% wilcox_test(methyl ~ type, paired = F) %>% 
  add_x_position() %>% adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 2)) #actual p-value

supfig7c <- data2 %>% 
  
  ggviolin(x="type",y="methyl", fill="type", trim=T,scale="width", width = .5) +  #,alpha=0.8
  geom_boxplot(outlier.shape = NA,width=0.05) +
  facet_wrap( ~ int_tis,nrow=1) +
  stat_pvalue_manual(stat.test, label = "p.adj.signif",y.position=100,tip.length = .01,size=2.5) + 
  labs(x="",y="CpG methylation (%) at TSSs") +
  scale_fill_manual(values=c("grey70","#68abdaff")) +
  scale_x_discrete(labels=c("skipping_gene"="Skipped","target_gene"="Interacting")) +
  coord_cartesian(ylim= c(0,110)) +
  theme_bw() +
  theme(axis.text.y = element_text(size=6),axis.text.x = element_text(size=6,angle = 30, vjust = 1, hjust = 1),
        axis.title.y=element_text(size=6),
        legend.position = "none",
        strip.text.x = element_text(size = 6.5))

xpie <- data2 %>% mutate(pos=if_else(methyl>50,"Yes","No")) %>% dplyr::count(int_tis,pos,type) %>% group_by(int_tis,type) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% ggplot( aes(x="", y=perc, fill=pos)) + geom_bar(width = 1, stat = "identity") + coord_polar("y", start=0) + theme_void() + facet_wrap(int_tis ~ type, nrow=1) + 
  scale_fill_manual(values=c("#c6ccc2","#dd8581ff"), name="") +
  theme(legend.key.size = unit(2.5, 'mm'),
        legend.text = element_text(size=6),legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank())

supfig7c <- plot_grid(supfig7c,xpie,ncol=1,rel_heights = c(10,1),align = "v")

fig4i
supfig7c

methyl_group %>% group_by(me_cat,int_tis) %>% summarise(mean_methyl=mean(methyl)) %>% group_by(me_cat) %>% summarise(methyl=mean(mean_methyl))

methyl_group %>% dplyr::filter(me_cat!="int_gene") %>% dplyr::count(int_tis,me_cat) %>% group_by(int_tis) %>% mutate(frac=n/sum(n)) #80.8/9.2? fold?

methyl_group %>% dplyr::filter(me_cat!="int_gene") %>% dplyr::count(int_tis,me_cat) %>% group_by(int_tis) %>% mutate(frac=n/sum(n)) %>% dplyr::filter(me_cat=="low")

methyl_group %>% left_join(rna_seq_e115,by=c("gene_name","int_tis"="tissue")) %>% dplyr::filter(!is.na(mean_TPM)) %>%  group_by(me_cat,int_tis) %>% summarise(mean=mean(mean_TPM)) # 44.8/0.8

```

DNaseI

```{r}
tis_corr <- data.frame(tissue=c("embryonicfacialprominence","forebrain","hindbrain","forelimb","hindlimb","midbrain","neuraltube"),new=c("CF","FB","HB","FL","HL","MB","NT"))

files <- list.files(path="~/enh_capC/encode/DNase/", pattern="bg$", recursive=T) 
rm(dataset)
for (path in files){
  name=strsplit(path, "_") %>% unlist()
  if (name[1] != "limb") {
    temp_dataset <- read_tsv(paste0("~/enh_capC/encode/DNase/",path),col_names = c("chr","start","end","read_depth")) %>% mutate(tissue = name[1], stage=name[3])
    if (!exists("dataset")) {
      dataset <- temp_dataset 
    } else {
      dataset <- temp_dataset %>% bind_rows(dataset)
    }
    rm(temp_dataset)
  }
}
dnase <- dataset %>% as.data.table() %>% foverlaps(mutate(refgene_tss,tss_start=tss_start+1500,tss_end=tss_end-1500), by.x=c("chr","start","end"),type="any") %>% group_by(chr,tss_start,tss_end,gene_name,tissue,stage) %>% partition(new_cluster(10)) %>% summarise(sum_depth=sum(read_depth*(end-start)),len=sum((end-start))) %>% collect() %>% ungroup() %>% mutate(ave_depth=sum_depth/len) %>% left_join(tis_corr,by="tissue") %>% dplyr::select(chr,tss_start,tss_end,gene_name,tissue=new,ave_depth) %>% mutate(tss_start=tss_start-1500,tss_end=tss_end+1500)

data <- enh_tss_exp_tis %>% 
  group_by(bait_name,gene_name) %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  ungroup() %>% as.data.table() %>%  
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>%
  distinct(bait_name,bait_chr,i.gene_name,gene_name,across(contains("tss")),int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,skipping_tss_start=tss_start,skipping_tss_end=tss_end, target_gene=i.gene_name,target_tss_start=i.tss_start,target_tss_end=i.tss_end,bait_name,bait_chr,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% dplyr::filter(class=="POS") %>% #only positive tissues
  pivot_longer(!c(bait_name,bait_chr,int_tis,skipping_tss_start,skipping_tss_end,target_tss_start,target_tss_end,class),names_to = "type", values_to = "gene_name") %>% #also filter those are target genes in skipping gene list
  mutate(tss_start=if_else(type=="skipping_gene",skipping_tss_start,target_tss_start),tss_end=if_else(type=="skipping_gene",skipping_tss_end,target_tss_end)) %>% dplyr::select(-c(skipping_tss_start,target_tss_start,skipping_tss_end,target_tss_end)) %>% distinct() %>%
  group_by(gene_name,int_tis) %>% mutate(type2=if_else(length(unique(type))==2,1,0)) %>% 
  ungroup() %>% mutate(type=if_else(type2==1,"target_gene",type)) %>% distinct(bait_chr,type,gene_name,tss_start,tss_end,int_tis) %>%
  left_join(dnase,by=c("bait_chr"="chr","tss_start","tss_end","int_tis"="tissue","gene_name")) %>% 
  dplyr::filter(!is.na(ave_depth)) %>% #some of the tissues or genes weren't detected in encode methyl data, so remove them
  group_by(gene_name,type,int_tis) %>% summarise(dnase=mean(ave_depth)) %>% ungroup() %>% #one gene has multiple TSSs, so take a mean
  mutate(int_tis=factor(int_tis,levels=c("FB","MB","HB","NT","CF","FL","HL")))

stat.test <- data %>% group_by(int_tis) %>% wilcox_test(dnase ~ type, paired = F) %>% 
  add_x_position(x="int_tis") %>% adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>%
  mutate(p.adj=formatC(p.adj, format = "g", digits = 3)) #actual p-value
 
fig4f <- data %>% 
  ggviolin(x="int_tis",y="dnase", fill="type",  trim=T) +  
  geom_boxplot(aes(dodge=type),position = position_dodge(width = 0.8), outlier.shape = NA,width=0.02) +
  stat_pvalue_manual(stat.test, label = "p.adj",y.position=4.4,tip.length = .01,size=2.5,angle=30,vjust=-1,hjust=0.5) + #???paired or not
  labs(x="",y="Average DNase I cleavge\n±1 kb around TSS in active tissues") +
  scale_y_continuous(limits = c(0,4.7)) +
  scale_fill_manual(values=c("#e9b86fff","#68abdaff"),name="",labels=c("skipping_gene"="Skipped genes","target_gene"="Target genes")) +
  theme(axis.text.x = element_text(size=6),
        axis.title.y=element_text(size=7),axis.text.y = element_text(size=6),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7),
        legend.text = element_text(size=6))

fig4f
```

GO for expressed skipped genes 

```{r}
data1 <- enh_tss_exp_tis %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>% distinct(bait_name,bait_type,i.gene_name,gene_name,int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,target_gene=i.gene_name,bait_name,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>% 
  left_join(exp_gene,by=c("skipping_gene"="gene_name")) %>% 
  dplyr::filter(!is.na(exp)) %>% #filter out non-expressed genes
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>%
  dplyr::select(-bait_name) %>% pivot_longer(!c(int_tis,class),names_to = "type",values_to = "gene_name") %>% distinct() %>% dplyr::filter(class=="POS") %>%
  group_by(gene_name,int_tis) %>% mutate(n=if_else(length(unique(type))==2,1,0)) %>% ungroup() %>% mutate(type=if_else(n==1,"target_gene",type)) %>% distinct() %>%
  left_join(rna_seq_e115,by=c("gene_name","int_tis"="tissue")) %>% dplyr::filter(!is.na(mean_TPM)) %>% distinct(gene_name,int_tis,mean_TPM,class,type) %>%
  mutate(int_tis=factor(int_tis,levels=c("FB","MB","HB","NT","CF","FL","HL","HR"))) 
  
gene <- data1 %>% group_by(type,gene_name) %>% summarise(mean_TPM=min(mean_TPM)) %>% ungroup() %>% add_count(gene_name) %>% dplyr::filter(n==1) %>% #only in one category
  dplyr::filter(mean_TPM>1) %>% dplyr::filter(type=="skipping_gene") %>% #expressing skipped genes
  distinct(gene_name) %>% .$gene_name
anno <- AnnotationDbi::select(org.Mm.eg.db,keys=gene,columns=c("ENTREZID","ENSEMBL","GENENAME"),keytype="SYMBOL") %>% dplyr::filter(!is.na(ENSEMBL))
  
ego <- enrichGO(gene          = anno$ENTREZID,
                OrgDb         = org.Mm.eg.db,
                ont           = "MF",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
        readable      = TRUE)
df <- head(ego,n=20) %>% as_tibble() %>% separate(GeneRatio,into = c("on_gene","on_tot"),sep="/",remove = F) %>% separate(BgRatio,into = c("bg_gene","bg_tot"),sep="/") %>% mutate(across(on_gene:bg_tot, as.numeric)) %>% mutate(FoldEnrich=formatC( (on_gene/on_tot)/(bg_gene/bg_tot), format = "g", digits = 2), qvalue=formatC(qvalue, format = "g", digits = 2)) %>% 
  dplyr::select(ID,MolecularFunction=Description,GeneRatio,FoldEnrich,qvalue)
df2 <- head(ego,n=20)

ego <- enrichGO(gene          = anno$ENTREZID,
                OrgDb         = org.Mm.eg.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
        readable      = TRUE)
df1 <- head(ego,n=5) %>% as_tibble() %>% separate(GeneRatio,into = c("on_gene","on_tot"),sep="/",remove = F) %>% separate(BgRatio,into = c("bg_gene","bg_tot"),sep="/") %>% mutate(across(on_gene:bg_tot, as.numeric)) %>% mutate(FoldEnrich=formatC( (on_gene/on_tot)/(bg_gene/bg_tot), format = "g", digits = 2), qvalue=formatC(qvalue, format = "g", digits = 2)) %>% 
  dplyr::select(ID,BiologicalProcess=Description,GeneRatio,FoldEnrich,qvalue)

ggplot() + theme_void() +                                              # Add table to ggplot2 plot
  annotate(geom = "table", x = 0,  y = 0, label = list(df),size = 3) +
  annotate(geom = "table", x = 0,  y = 1, label = list(df1),size = 3) 

```

is there cpg island difference between target and skipped genes

```{r}
cpg <- read_tsv("~/enh_capC/encode/mm10_cpg_island.tsv") %>% dplyr::select(chrom,chromStart,chromEnd,name,length,cpgNum,perCpg) %>% as.data.table()
setkey(cpg,chrom,chromStart,chromEnd)

enh_tss_exp_tis %>% 
  group_by(bait_name,gene_name) %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  ungroup() %>% as.data.table() %>%  
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>%
  distinct(bait_name,bait_chr,i.gene_name,gene_name,across(contains("tss")),int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,skipping_tss_start=tss_start,skipping_tss_end=tss_end, target_gene=i.gene_name,target_tss_start=i.tss_start,target_tss_end=i.tss_end,bait_name,bait_chr,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% dplyr::filter(class=="POS") %>% #only positive tissues
  pivot_longer(!c(bait_name,bait_chr,int_tis,skipping_tss_start,skipping_tss_end,target_tss_start,target_tss_end,class),names_to = "type", values_to = "gene_name") %>% 
  mutate(tss_start=if_else(type=="skipping_gene",skipping_tss_start,target_tss_start),tss_end=if_else(type=="skipping_gene",skipping_tss_end,target_tss_end)) %>% dplyr::select(-c(skipping_tss_start,target_tss_start,skipping_tss_end,target_tss_end)) %>% distinct() %>%
  group_by(gene_name,int_tis) %>% mutate(type2=if_else(length(unique(type))==2,1,0)) %>% 
  ungroup() %>% mutate(type=if_else(type2==1,"target_gene",type)) %>% distinct(bait_chr,type,gene_name,tss_start,tss_end,int_tis) %>%
  mutate(tss_start=tss_start-5000,tss_end=tss_end+5000) %>% #extend 5kb
  
  as.data.table() %>% #join cpg island with different distance
  foverlaps(cpg, by.x=c("bait_chr","tss_start", "tss_end"), type="any") %>% mutate(perCpg=if_else(is.na(perCpg),0,1)) %>% #as 0 and 1
  group_by(type,gene_name) %>% summarise(perCpg=max(perCpg)) %>% ungroup() %>% add_count(type) %>% 
  ggboxplot(x="type",y="perCpg",color="type",outlier.size = 0.3,add = "jitter",add.params = list(width = 0.1,size = 0.3)) +
  geom_text(aes(type,y=1.2,label=paste0("n=",n)),stat = "unique",size=2.5) +
  stat_compare_means(label = "p.format", comparisons = list(c("skipping_gene","target_gene")),method = "wilcox.test",paired = FALSE,tip.length = .02,label.y = 1.05,size=2.5) + 
  labs(x="",y="Cpg or not",title="extended 5kb") +
  scale_x_discrete(labels=c("skipping_gene"="Skipped genes","target_gene"="Target genes")) +
  scale_color_manual(values=c("#ec7f26e3","#4d77b1e8"),name="",labels=c("skipping_gene"="Skipped genes","target_gene"="Target genes")) +
  theme(axis.text.x = element_text(size=6),
        axis.title.y=element_text(size=7),axis.text.y = element_text(size=6),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7),
        legend.text = element_text(size=6))

```

```{r}
data <- enh_tss_exp_tis %>% 
  group_by(bait_name,gene_name) %>% 
  mutate(up=if_else(tss_start>bait_start,bait_end+1,tss_end+1),down=if_else(tss_start>bait_start,tss_start-1,bait_start-1),len=down-up) %>%
  ungroup() %>% as.data.table() %>%  
  foverlaps(refgene_tss, by.x=c("bait_chr","up", "down"), type="any") %>%
  distinct(bait_name,bait_chr,i.gene_name,gene_name,across(contains("tss")),int_tis) %>% dplyr::filter((i.gene_name != gene_name) | is.na(gene_name)) %>% dplyr::select(skipping_gene=gene_name,skipping_tss_start=tss_start,skipping_tss_end=tss_end, target_gene=i.gene_name,target_tss_start=i.tss_start,target_tss_end=i.tss_end,bait_name,bait_chr,int_tis) %>% dplyr::filter(!is.na(skipping_gene)) %>%  #remove no skipping gene pairs
  left_join(bait_enh_act,by=c("bait_name"="elementID","int_tis"="tissue")) %>% dplyr::filter(class=="POS") %>% #only positive tissues
  pivot_longer(!c(bait_name,bait_chr,int_tis,skipping_tss_start,skipping_tss_end,target_tss_start,target_tss_end,class),names_to = "type", values_to = "gene_name") %>% 
  mutate(tss_start=if_else(type=="skipping_gene",skipping_tss_start,target_tss_start),tss_end=if_else(type=="skipping_gene",skipping_tss_end,target_tss_end)) %>% dplyr::select(-c(skipping_tss_start,target_tss_start,skipping_tss_end,target_tss_end)) %>% distinct() %>%
  group_by(gene_name,int_tis) %>% mutate(type2=if_else(length(unique(type))==2,1,0)) %>%
  ungroup() %>% mutate(type=if_else(type2==1,"target_gene",type)) %>% distinct(bait_chr,type,gene_name,tss_start,tss_end,int_tis) %>%
  as.data.table() %>% #join cpg island with different distance
  foverlaps(cpg, by.x=c("bait_chr","tss_start", "tss_end"), type="any") %>% mutate(perCpg=if_else(is.na(perCpg),0,1)) %>% #as 0 and 1
  group_by(type,gene_name) %>% summarise(perCpg=max(perCpg)) %>% ungroup() %>% add_count(type) %>% mutate(perCpg=factor(perCpg)) %>% count(type,perCpg) 

test <- data %>% pivot_wider(names_from = perCpg, values_from = n) %>% column_to_rownames("type") %>% fisher_test(detailed = TRUE, alternative = "greater")

data %>% group_by(type) %>% mutate(perc=n/sum(n)) %>% ungroup() %>%
  ggbarplot(x="type",y="perc",fill="perCpg",color ="perCpg", palette = "Paired") +
  geom_text(aes(type,y=c(0.75,0.25,0.92,0.45),label=paste0("n=",n)),stat = "unique",size=4) +
  geom_text(aes(x=1.5, y=1.1,label=paste0("Fisher's exact test, p=",formatC(test$p, format = "g", digits = 2))),size=4,stat = "unique") +
  labs(x="",y="Cpg or not",subtitle="") 
```



plot #hs271-Nr2f1, Pou5f2. FB as an example 
```{r}
rm_sd <- function(df, k) {
  df %>%
    mutate(
      rmean = rollapply(rescaled_Nav, k, FUN = mean, partial=T, align="center"),
      rstd = rollapply(rescaled_Nav, width = k, FUN = sd, partial=T, align="center")
    )  #add the prefix to calculate sd successfully
}

tis="FB"
targ="hs271"
dplyr::filter(bmap,name==targ)

chrom="chr13"
gene_limit=c(77500000,78500000)

exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100


tis_col=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TK","TL"),color=c("#629fd8ff","#716bbdff","#c881bbff","#c1d17dff","#c99516ff","#c75b7cff","#c83737ff","#cfcf3cff","#5b863dff","#73d758ff"))
enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "within")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)


#scale bar
p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=2) +
  annotate(geom = "text",x=gene_limit[2]-500000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=2) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()
p1 <- readcount_bait %>% fill(midpoint,.direction = "downup") %>% 
  mutate(rescaled_Nav=if_else(is.na(rescaled_Nav),0,rescaled_Nav)) %>%
  group_by(condition,name) %>% rm_sd(.,3) %>% ungroup() %>%
  mutate(rmean=replace(rmean,oeID_mid==bait_mid,NA)) %>% #remove bait read count
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20",size=0.3) + theme_bw() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = 100, xend = oeID_mid, yend = 100), 
                 curvature = -dat["cur"]*1.2, color="gray25",size=0.2 ) })}} + 
  scale_y_continuous(limits = c(0,115)) +
  geom_gene_arrow(data=gid, aes(xmin = gene_seq_start, xmax = gene_seq_end, y = 105,forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "rect", xmin=c(137309445,136889611,136486121), xmax=c(137319445,136899611,136497889), ymin = -Inf,ymax=Inf,color = NA,fill = c("#629fd8ff","#c99516ff", "#5b863dff"),alpha=0.5) +
  annotate(geom = "tile", x = enh$baitID_mid, y = 105,fill = "#448838",width=5000,height=4) +
  annotate(geom = "text", x=enh$baitID_mid, y= 105,label=enh$name,size = 2) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 105,label=unique(gid$symbol),size = 2) +
  annotate(geom = "point", x = enh$baitID_mid, y = 115,fill = "red2",color = "red2", size = 1,shape=25) +
  labs(y = paste0(tis," normalized\ninteraction frequency")) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.title.y=element_text(size=7),
        axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

fig4b <- plot_grid(p0,p1,ncol=1,align = "v",rel_heights = c(1,10))


#hs271-Nr2f1, Pou5f2. FB
#hs271; chr13; Nr2f1, ENSMUST00000091458;Pou5f2,ENSMUST00000175955; y_limit=2
chrom="chr13"
y_limit_dna=4
y_limit_rna=4
ext=5000
cpgfile=c("encode_data/hs271_forebrain_CpG_E115_ENCFF232QBX.bed.gz.bg","encode_data/hs271_forebrain_CpG_E115_ENCFF890XTY.bed.gz.bg")
dnasefile="encode_data/hs271_forebrain_DNase_e11_5.bg"
rnaseqfile="encode_data/hs271_forebrain_RNAseq_e11_5.bg"

genename="Nr2f1"
txid="ENSMUST00000091458"
exon <- exons(edb,filter = ~ (tx_id == txid & gene_name == genename)) %>% as_tibble() %>% mutate(mid=(start+end)/2)
intron <- exon %>% mutate(intron_start=end,intron_end=lead(start)) %>% dplyr::filter(!is.na(intron_end)) %>% dplyr::select(gene_name,intron_start,intron_end,strand) %>% mutate(mid=(intron_start+intron_end)/2, width=intron_end-intron_start)
trans_lim <- exon %>% summarise(start=min(start),end=max(end),strand=unique(strand)) 
cd <- cdsBy(edb,filter = ~ (tx_id == txid & gene_name == genename)) %>% as_tibble() %>% mutate(mid=(start+end)/2)

p4 <- read_tsv(cpgfile[1]) %>% bind_rows(read_tsv(cpgfile[2])) %>%
  dplyr::filter(chr==chrom, start>=trans_lim$start-ext, end<=trans_lim$end+ext) %>%
  group_by(chr,start,end) %>% summarise(read_num=sum(read_num),perc=mean(perc)) %>% #pool replicates
  ggplot(aes(xmin=start, xmax=end, ymin=0, ymax=perc)) + geom_rect(fill="grey30", colour="grey30",size=0.3) +
  {if(trans_lim$strand=="+") {
    coord_cartesian(xlim=c(trans_lim$start-ext,trans_lim$start+ext))} else {
    coord_cartesian(xlim=c(trans_lim$end-ext,trans_lim$end+ext))} } + 
  annotate(geom = "tile", x=intron$mid, y=110,width=intron$width,height=1,fill="#629fd8ff", color="#629fd8ff") +
  annotate(geom = "tile", x=exon$mid, y= 110,width=exon$width,height=20,fill="white", color="#629fd8ff",size=0.5) +
  annotate(geom = "tile", x=cd$mid, y= 110,width=cd$width,height=20,fill="#629fd8ff", color="#629fd8ff") +
  {if(trans_lim$strand=="+") {
    annotate(geom = "text", x=trans_lim$start, y = 105, label=genename,size=2.5)} else {
    annotate(geom = "text", x=trans_lim$end, y = 105, label=genename,size=2.5)} } +
  scale_y_continuous(breaks = 100) +
  theme_classic() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_blank(),axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p7 <- read_tsv(dnasefile,col_names = c("chr","start","end","read_count")) %>%
  dplyr::filter(chr==chrom, start>=trans_lim$start-ext, end<=trans_lim$end+ext) %>%
  ggplot(aes(xmin=start, xmax=end, ymin=0, ymax=read_count)) + geom_rect(fill="grey30", colour="grey30",size=0.3) +
  {if(trans_lim$strand=="+") {
    coord_cartesian(xlim=c(trans_lim$start-ext,trans_lim$start+ext),ylim=c(0,y_limit_dna))} else {
    coord_cartesian(xlim=c(trans_lim$end-ext,trans_lim$end+ext),ylim=c(0,y_limit_dna))} } + 
  scale_y_continuous(breaks = y_limit_dna) +
  theme_classic() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_blank(),axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p1 <- read_tsv(rnaseqfile,col_names = c("chrom","chromStart","chromEnd","readcount")) %>% 
  dplyr::filter(chrom==chrom, chromStart>=trans_lim$start, chromEnd<=trans_lim$end) %>% mutate(readcount=if_else(readcount>y_limit_rna,y_limit_rna,readcount)) %>%
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=readcount)) + geom_rect(fill="#629fd8ff", colour="#629fd8ff", size=0.5) +
  {if(trans_lim$strand=="+") {
    coord_cartesian(xlim=c(trans_lim$start-ext,trans_lim$start+ext),ylim=c(0,y_limit_rna))} else {
    coord_cartesian(xlim=c(trans_lim$end-ext,trans_lim$end+ext),ylim=c(0,y_limit_rna))} } + 
  scale_y_continuous(breaks = y_limit_rna) +
  labs(y = "FB RNA-seq") +
  theme_classic() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_blank(),axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))


genename="Pou5f2"   #color used to be #c99516ff
txid="ENSMUST00000175955"
exon <- exons(edb,filter = ~ (tx_id == txid & gene_name == genename)) %>% as_tibble() %>% mutate(mid=(start+end)/2)
intron <- exon %>% mutate(intron_start=end,intron_end=lead(start)) %>% dplyr::filter(!is.na(intron_end)) %>% dplyr::select(gene_name,intron_start,intron_end,strand) %>% mutate(mid=(intron_start+intron_end)/2, width=intron_end-intron_start)
trans_lim <- exon %>% summarise(start=min(start),end=max(end),strand=unique(strand)) 
cd <- cdsBy(edb,filter = ~ (tx_id == txid & gene_name == genename)) %>% as_tibble() %>% mutate(mid=(start+end)/2)

p5 <- read_tsv(cpgfile[1]) %>% bind_rows(read_tsv(cpgfile[2])) %>%
  dplyr::filter(chr==chrom, start>=trans_lim$start-ext, end<=trans_lim$end+ext) %>%
  group_by(chr,start,end) %>% summarise(read_num=sum(read_num),perc=mean(perc)) %>% #pool replicates
  ggplot(aes(xmin=start, xmax=end, ymin=0, ymax=perc)) + geom_rect(fill="grey30", colour="grey30",size=0.3) +
  scale_y_continuous(breaks = 100) +
  {if(trans_lim$strand=="+") {
    coord_cartesian(xlim=c(trans_lim$start-ext,trans_lim$start+ext))} else {
    coord_cartesian(xlim=c(trans_lim$end-ext,trans_lim$end+ext))} } + 
  annotate(geom = "tile", x=intron$mid, y= 110,width=intron$width,height=2,fill="grey50", color="grey50") +
  annotate(geom = "tile", x=exon$mid, y= 110,width=exon$width,height=20,fill="white", color="grey50",size=0.5) +
  annotate(geom = "tile", x=cd$mid, y= 110,width=cd$width,height=20,fill="grey50", color="grey50") +
  {if(trans_lim$strand=="+") {
    annotate(geom = "text", x=trans_lim$start, y = 105, label=genename,size=2.5)} else {
    annotate(geom = "text", x=trans_lim$end, y = 105, label=genename,size=2.5)} } +
  theme_classic()  + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_blank(),axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p8 <- read_tsv(dnasefile,col_names = c("chr","start","end","read_count")) %>%
  dplyr::filter(chr==chrom, start>=trans_lim$start-ext, end<=trans_lim$end+ext) %>%
  ggplot(aes(xmin=start, xmax=end, ymin=0, ymax=read_count)) + geom_rect(fill="grey30", colour="grey30",size=0.3) +
  {if(trans_lim$strand=="+") {
    coord_cartesian(xlim=c(trans_lim$start-ext,trans_lim$start+ext),ylim=c(0,y_limit_dna))} else {
    coord_cartesian(xlim=c(trans_lim$end-ext,trans_lim$end+ext),ylim=c(0,y_limit_dna))} } + 
  scale_y_continuous(breaks = y_limit_dna) +
  theme_classic() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_blank(),axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p2 <- read_tsv(rnaseqfile,col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=trans_lim$start, chromEnd<=trans_lim$end) %>% mutate(readcount=if_else(readcount>y_limit_rna,y_limit_rna,readcount)) %>%
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=readcount)) + geom_rect(fill="grey50", colour="grey50", size=0.5) +
  {if(trans_lim$strand=="+") {
    coord_cartesian(xlim=c(trans_lim$start-ext,trans_lim$start+ext),ylim=c(0,y_limit_rna))} else {
    coord_cartesian(xlim=c(trans_lim$end-ext,trans_lim$end+ext),ylim=c(0,y_limit_rna))} } + 
  {if(trans_lim$strand=="+") {
    annotate(geom = "tile", x=trans_lim$start, y= y_limit_rna*0.95,width=3000,height=y_limit_rna*0.1)} else {
    annotate(geom = "tile", x=trans_lim$end, y= y_limit_rna*0.95,width=3000,height=y_limit_rna*0.1)} } + 
  {if(trans_lim$strand=="+") {
    annotate(geom = "text", x=trans_lim$start, y = y_limit_rna, label="3 kb",size=2.5)} else {
    annotate(geom = "text", x=trans_lim$end, y= y_limit_rna, label="3 kb",size=2.5)} } + 
  scale_y_continuous(breaks = y_limit_rna) +
  theme_classic() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_blank(),axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))


genename="hs271"
trans_lim <- data.frame(start=77890808,end=77894441) 
x_adjust=500

p6 <- read_tsv(cpgfile[1]) %>% bind_rows(read_tsv(cpgfile[2])) %>%
  dplyr::filter(chr==chrom, start>=trans_lim$start-ext+x_adjust, end<=trans_lim$end+ext+x_adjust) %>%
  group_by(chr,start,end) %>% summarise(read_num=sum(read_num),perc=mean(perc)) %>% #pool replicates
  ggplot(aes(xmin=start, xmax=end, ymin=0, ymax=perc)) + geom_rect(fill="grey30", colour="grey30",size=0.3) +
  coord_cartesian(xlim=c(trans_lim$start-ext+x_adjust,trans_lim$start+ext+x_adjust)) +
  annotate(geom = "tile", x=(trans_lim$end+trans_lim$start)/2, y= 110, width=trans_lim$end-trans_lim$start,height=20,fill="#5b863dff", color="#5b863dff") +
  annotate(geom = "text", x=trans_lim$start, y = 105, label=genename,size=2.5) +
  scale_y_continuous(breaks = 100) +
  theme_classic() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_blank(),axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p9 <- read_tsv(dnasefile,col_names = c("chr","start","end","read_count")) %>%
  dplyr::filter(chr==chrom, start>=trans_lim$start-ext+x_adjust, end<=trans_lim$end+ext+x_adjust) %>%
  ggplot(aes(xmin=start, xmax=end, ymin=0, ymax=read_count)) + geom_rect(fill="grey30", colour="grey30",size=0.3) +
  coord_cartesian(xlim=c(trans_lim$start-ext+x_adjust,trans_lim$start+ext+x_adjust),ylim=c(0,y_limit_dna)) + 
  scale_y_continuous(breaks = y_limit_dna) +
  theme_classic() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_blank(),axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p3 <- data.frame(start=trans_lim$start-1500,end=trans_lim$start+1500,ymin=2,ymax=2.2) %>%
  ggplot(aes(xmin=start, xmax=end, ymin=ymin, ymax=ymax)) + geom_rect(fill="black", colour=NA) +
  scale_y_continuous(breaks = 5,limits = c(0,5)) + scale_x_continuous(limits = c(trans_lim$start-ext,trans_lim$start+ext)) +
  annotate(geom = "text", x=trans_lim$start, y = 3, label="3kb",size=2.5) +
  theme_void() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_blank(),axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))



fig4c <- plot_grid(#plot_grid(p6,p9,p3,ncol=1,align = "v", axis = 'lr',rel_heights = c(1.2,1.1,1)),
                   plot_grid(p5,p8,p2,ncol=1,align = "v", axis = 'lr',rel_heights = c(1.2,1.1,1)),
                   plot_grid(p4,p7,p1,ncol=1,align = "v", axis = 'lr',rel_heights = c(1.2,1.1,1)),
                   #plot_grid(p10,p11,p12,ncol=1,align = "v", axis = 'lr',rel_heights = c(1.2,1.1,1)),
                   nrow=1,align = "hv",scale = 0.9)
fig4c


fig4bc <- plot_grid(fig4b,fig4c,ncol=1,labels = c("B","C"),align = "v",rel_heights = c(1,1.2),label_size = 12)
fig4bc
```


Fig3D. count ep_expressing pairs from now on, the interactions are count
based on function elements, so group by promoter coordinates

```{r}
#rna_seq
#ep_exp
#ep_exp_5kb

#use promoter as counting standard
### summarise the number of tissues they interact in (7 tissues)
p1 <- ep_exp_5kb %>% distinct(bait_name,promoter,tissue,bait_type) %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% 
  dplyr::count(bait_name,promoter,bait_type,name="interact_tot") %>% dplyr::count(interact_tot,bait_type) %>% 
  dplyr::filter(bait_type=="enh") %>%
  mutate(interact_tot=if_else(interact_tot==10,"10",if_else(interact_tot>6,"7-9",if_else(interact_tot>3,"4-6","1-3")))) %>% dplyr::count(bait_type,interact_tot,wt=n,name = "tot") %>% ungroup() %>% mutate(interact_tot=factor(interact_tot,levels = c("1-3","4-6","7-9","10")), perc=tot/sum(tot)) %>%
  ggplot(aes(x="",y=perc,fill=interact_tot)) + geom_bar(width = 1, stat = "identity", color = "white") + geom_text(aes(label=paste0(interact_tot,"\n",formatC(perc*100, format = "f", digits = 1),"%"), x = 1.2),position=position_stack(0.5),size=3) + #facet_wrap(~ bait_type) + 
  coord_polar("y", start=0) + theme_void() + 
  scale_fill_manual(name="Tissue No.",values=c("#eff3ffff","#b9d5eaff","#4292c6ff","#084594ff")) +
  theme(legend.direction='horizontal',legend.position = "bottom",
        legend.key.size = unit(3, 'mm'),
        legend.title = element_text(size=8),
        legend.text = element_text(size=6)) #+ labs(title="ep_expressing_pair_count_7_tissue")

tis4=c("brain","cf","limb","hrt","nt","tail","trunk") 
tis_group=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TL","TK"),tis_group=c(rep("brain",3),"cf",rep("limb",2),"hrt","nt","tail","trunk"))
### summarise the number of tissues they interact in (4 tissues)
p2 <- ep_exp_5kb %>% left_join(tis_group,by="tissue") %>% distinct(bait_name,promoter,tis_group,bait_type) %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% 
  dplyr::count(bait_name,promoter,bait_type,name="interact_tot") %>% dplyr::count(interact_tot,bait_type) %>% 
  dplyr::filter(bait_type=="enh") %>%
  group_by(bait_type) %>% mutate(perc=n/sum(n)) %>% ggplot(aes(x="",y=perc,fill=interact_tot)) + geom_bar(width = 1, stat = "identity", color = "white") + geom_text(aes(label=n, x = 1.2),position=position_stack(0.5),size=3) + 
  coord_polar("y", start=0) + theme_void() + scale_fill_distiller(name="Tissue region No.",direction = 1) +
  theme(legend.direction='horizontal',legend.position = "bottom",
        legend.key.size = unit(3, 'mm'),
        legend.title = element_text(size=8),
        legend.text = element_text(size=6))  #+ labs(title="ep_expressing_pair_count_4_tissue")
fig3d <- plot_grid(p1,p2,nrow=1,scale=0.9)
fig3d

ep_exp_5kb %>% distinct(bait_name,promoter,tissue,bait_type) %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% dplyr::count(bait_name,promoter,bait_type,name="interact_tot") %>% dplyr::count(interact_tot,bait_type) %>% dplyr::filter(bait_type=="enh") %>% group_by(bait_type) %>% mutate(perc=n/sum(n)) %>% 
  #summarise(sum=sum(n))
  dplyr::filter(interact_tot==10)

#how many percent of enhancers form stable interactions
ep_exp_5kb %>% distinct(bait_name,promoter,tissue,bait_type) %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% dplyr::count(bait_name,promoter,bait_type,name="interact_tot") %>% group_by(bait_name,bait_type) %>% dplyr::filter(interact_tot==max(interact_tot)) %>% ungroup() %>% distinct(bait_name,bait_type,interact_tot) %>% dplyr::count(interact_tot) %>% mutate(perc=n/sum(n))

#number of e-p interaction
ep_exp_5kb %>% distinct(bait_name,promoter,tissue,bait_type) %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% 
  dplyr::count(bait_name,promoter,bait_type,name="interact_tot") %>% dplyr::count(bait_type) %>% 
  dplyr::filter(bait_type=="enh")
```

Fig2B. plot e-p interactions metaplot

```{r}
#1.scale read count
#count_scaled

#2.get ep pairs
#ep_exp_5kb %>% distinct(bait_type)
#count(count_scaled,name,condition) %>% dplyr::filter(name %in% ep_exp_5kb$bait_name) %>% count(name,wt=n) %>% arrange(n) #still mm1502 and mm75 have to few reads, remove them

#3.get bait activity
bait_enh_act 
#bait_enh_act %>% distinct(class)

#4.plot metaplot
rm_sd <- function(df, val, k) {
  df %>%
    mutate(
      rmean = rollapply(val, k, FUN = mean, partial=T, align="center"),
      rstd = rollapply(val, width = k, FUN = sd, partial=T, align="center")
    )  #add the prefix to calculate sd successfully
}
```

 
```{r}
act_stage <- read_tsv("all_activity_from_e10_to_e12.tsv") %>% group_by(enhancer_chr,enhancer_start,enhancer_end,tissue,stage) %>% partition(new_cluster(10)) %>% dplyr::filter(signal==max(signal)) %>% dplyr::filter(overlapping==max(overlapping)) %>% collect() %>% ungroup() %>% dplyr::select(-c(chr,start,end,overlapping)) %>% mutate(tissue=str_remove_all(tissue, "_"), stage=str_replace(stage, "_", "") %>% toupper()) %>% as.data.table() %>% left_join(tis_corr,by="tissue") %>% dplyr::select(-tissue) %>% dplyr::select(tissue=new,everything())
setkey(act_stage,enhancer_chr,enhancer_start,enhancer_end)

new_rmap <- rmap %>% mutate(oeID_mid2=oeID_mid) %>% as.data.table()
setkey(new_rmap,oe_chr,oeID_mid,oeID_mid2)

tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")

#for enh baits 
bound <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% 
  dplyr::filter(! bait_name %in% c("mm1502","mm75")) %>% group_by(bait_name,promoter) %>% mutate(oeIDmid=as.integer(mean(otherEndID)), strand=if_else(target_start>bait_start,"+","-")) %>% ungroup() %>% #get the peak mid on each promoter and strandness
  distinct(bait_chr,bait_start,bait_end,bait_name,tissue,promoter,oeIDmid,strand,bait_type) %>%
  left_join(rmap,by=c("bait_chr"="oe_chr","oeIDmid"="otherEndID")) %>%
  mutate(up=if_else(strand=="+", as.integer((bait_start+bait_end)/2)-50000, oeID_mid-50000), down=if_else(strand=="+",  oeID_mid+50000, as.integer((bait_start+bait_end)/2)+50000), bait_mid= as.integer((bait_start+bait_end)/2)) %>% as.data.table() %>% foverlaps(new_rmap, by.x=c("bait_chr","up", "down"), type="any") %>% arrange(bait_name,promoter,tissue) %>% dplyr::select(bait_name,bait_mid,bait_type,promoter,pro_mid=i.oeID_mid,otherEndID,oeID_mid,tissue,strand,up,down) %>% complete(nesting(bait_name,bait_mid,bait_type,promoter,pro_mid,otherEndID,oeID_mid,strand,up,down),tissue=tis) %>%
  left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% 
  left_join(dplyr::select(count_scaled,name,otherEndID,scaled_Nav,condition),by=c("bait_name"="name","otherEndID","tissue"="condition")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>%
  group_by(bait_name,promoter,tissue) %>% 
  mutate(class=if_else(is.na(class),"NEG","POS"), #complete the enhancer activity
         sum=sum(scaled_Nav)) %>% dplyr::filter(sum>0) %>% # filter out enh-pro pairs that have no read counts
  mutate(tenkb_Nav=if_else(oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000,0,scaled_Nav)) %>% mutate(max_dep=max(tenkb_Nav), rescaled_Nav=if_else( (scaled_Nav>max_dep | oeID_mid==bait_mid) ,100,scaled_Nav/max_dep*100)) %>% ungroup() %>% #set upper threshold for read count and rescale to 0-100
  dplyr::select(-c(otherEndID,tenkb_Nav,max_dep,up,down,sum)) %>% 
  group_by(bait_name,promoter,tissue) %>% arrange(bait_name,promoter,tissue,oeID_mid) %>%
  group_modify(~ { .x %>% rm_sd(.x$rescaled_Nav,3)}) %>% 
  mutate(newID=if_else(strand=="+",
                       if_else(oeID_mid<=bait_mid+10000, (oeID_mid-bait_mid-10000)/100,  #bait_mid+10000 is 0
                               if_else(oeID_mid>=pro_mid-10000,(oeID_mid-pro_mid+10000)/100+500, #pro_mid-10000 is 500
                                       (oeID_mid-bait_mid-10000)/(pro_mid-bait_mid-20000)*500)),  #scale in the middle
                       if_else(oeID_mid>=bait_mid-10000, (bait_mid-oeID_mid-10000)/100,    #otherwise
                               if_else(oeID_mid<=pro_mid+10000,(pro_mid-oeID_mid+10000)/100+500,
                                       (bait_mid-10000-oeID_mid)/(bait_mid-pro_mid-20000)*500)) 
                       )) %>% ungroup()  #give new coordinates


#for in vivo act
a_vs_in <- distinct(bound,bait_name,class) %>% dplyr::filter(! is.na(class)) %>% dplyr::count(bait_name) %>% dplyr::filter(n==2)  #463

bound1 <- bound %>% arrange(bait_name,promoter,tissue,newID) %>% dplyr::filter(abs(pro_mid-bait_mid)>30000 & tissue !="TK") %>% dplyr::filter(bait_name %in% a_vs_in$bait_name) %>% group_by(bait_name,promoter,bait_type,class,newID) %>% summarise(m_Nav=mean(rescaled_Nav)) %>% ungroup() %>% pivot_wider(names_from = class, values_from = m_Nav) %>% mutate(ratio=POS/NEG, code="enh_invivo")


num_int=bound1 %>% distinct(bait_name,promoter) %>% nrow()
num_enh=bound1 %>% distinct(bait_name) %>% nrow()

#add 95% CI
fig2b <- bound1 %>%
  ggplot(aes(x=newID,y=ratio)) + 
  stat_summary_bin(data=function(x) subset(x, newID < -110), geom="ribbon", binwidth= 30, fun.data=mean_cl_boot, fill="lightblue") +
  stat_summary_bin(data=function(x) subset(x, newID > -90), geom="ribbon", binwidth= 30, fun.data=mean_cl_boot, fill="lightblue") +
  stat_summary_bin(data=function(x) subset(x, newID < -110), geom="line", binwidth= 30, fun=mean, color="#0073C2FF",size=.3) +
  stat_summary_bin(data=function(x) subset(x, newID > -90), geom="line", binwidth= 30, fun=mean, color="#0073C2FF",size=.3) +
  geom_rect(data=data.frame(xmin = c(-130,-5,495), xmax = c(-70,5,505)), aes(xmin=xmin,xmax=xmax, ymin = -Inf, ymax = Inf),
            fill = "gray80",inherit.aes = F)  +
  coord_cartesian(xlim=c(-250, 750)) + 

  scale_x_continuous(limits=c(-250, 750), #trans = squish_trans(1000, 9000,10), 
                     breaks=c(-200,-100,0,500,600,700),labels = c("bait-10kb", "bait", "bait+10kb", "pro-10kb", "pro", "pro+10kb")) + 
  labs(x="Position", y="Mean interaction frequency ratio\n(activ vs. inactive)", title = paste0(num_enh," enhancers with ",num_int, " interactions")
       ) + theme_bw() +
  theme(axis.text.x = element_text(angle = 15, vjust = 1, hjust = 1,size=6),axis.title.x=element_blank(),axis.title.y=element_text(size=7),axis.text.y = element_text(size=6),
        legend.position = c(0.6,0.75),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7), 
        legend.text = element_text(size=6),
        plot.title = element_text(size=7)) 

#stable interactions, main groups
interact_num <- ep_exp_5kb %>% left_join(tis_group,by="tissue") %>% distinct(bait_name,promoter,tis_group,bait_type) %>% 
  dplyr::count(bait_name,promoter,bait_type,name="interact_tot") %>% arrange(desc(interact_tot)) %>% dplyr::filter(interact_tot==max(interact_tot))
  
supfig2e <- bound1 %>% left_join(interact_num,by=c("bait_name","promoter","bait_type")) %>% 
  mutate(stable=if_else(is.na(interact_tot),"no","yes")) %>%
  ggplot(aes(x=newID,y=ratio,fill=stable, color=stable
             )) + 
  stat_summary_bin(data=function(x) subset(x, newID < -110), geom="ribbon", binwidth= 30, fun.data=mean_cl_boot,color=NA,alpha=0.5) +
  stat_summary_bin(data=function(x) subset(x, newID > -90), geom="ribbon", binwidth= 30, fun.data=mean_cl_boot,color=NA,alpha=0.5) +
  stat_summary_bin(data=function(x) subset(x, newID < -110), geom="line", binwidth= 30, fun=mean,size=.5) +
  stat_summary_bin(data=function(x) subset(x, newID > -90), geom="line", binwidth= 30, fun=mean,size=.5) +
  geom_rect(data=data.frame(xmin = c(-130,-5,495), xmax = c(-70,5,505)), aes(xmin=xmin,xmax=xmax, ymin = -Inf, ymax = Inf),
            fill = "gray80",inherit.aes = F)  +
  scale_x_continuous(limits=c(-250, 750), #trans = squish_trans(1000, 9000,10), 
                     breaks=c(-200,-100,0,500,600,700),labels = c("bait-10kb", "bait", "bait+10kb", "pro-10kb", "pro", "pro+10kb")) + 
  labs(x="Position", y="Mean interaction frequency ratio\n(activ vs. inactive)") + theme_bw() +
  theme(axis.text.x = element_text(angle = 15, vjust = 1, hjust = 1,size=6),axis.title.x=element_blank(),axis.title.y=element_text(size=7),axis.text.y = element_text(size=6),
        legend.position = c(0.6,0.75),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7), 
        legend.text = element_text(size=6),
        plot.title = element_text(size=7)) +
  scale_color_manual(values=c("#4d77b1e8","#ec7f26e3"),name = "Interactions", labels = c("Tissue-specific","Invariant")) + scale_fill_manual(values=c("#4d77b1e8","#ec7f26e3")) + guides(fill = "none")

fig2b
supfig2e
```

negative regions interactions with inactive regions or metaplot without
encode annotated regions

```{r}
encode <- read_tsv("~/enh_capC/encode/encode_peaks.bed",col_names = c("chr","start","end")) %>% as.data.table() %>% mutate(encode_peak="yes")
setkey(encode,chr,start,end)

tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")
#for enh baits 
bound <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% 
  dplyr::filter(! bait_name %in% c("mm1502","mm75")) %>% group_by(bait_name,promoter) %>% mutate(oeIDmid=as.integer(mean(otherEndID)), strand=if_else(target_start>bait_start,"+","-")) %>% ungroup() %>% #get the peak mid on each promoter and strandness
  distinct(bait_chr,bait_start,bait_end,bait_name,tissue,promoter,oeIDmid,strand,bait_type) %>%
  left_join(rmap,by=c("bait_chr"="oe_chr","oeIDmid"="otherEndID")) %>%
  mutate(up=if_else(strand=="+", as.integer((bait_start+bait_end)/2)-50000, oeID_mid-50000), down=if_else(strand=="+",  oeID_mid+50000, as.integer((bait_start+bait_end)/2)+50000), bait_mid= as.integer((bait_start+bait_end)/2)) %>% as.data.table() %>% foverlaps(new_rmap, by.x=c("bait_chr","up", "down"), type="any") %>% arrange(bait_name,promoter,tissue) %>% dplyr::select(bait_name,bait_mid,bait_type,promoter,pro_mid=i.oeID_mid,otherEndID,oeID_mid,tissue,strand,up,down) %>% complete(nesting(bait_name,bait_mid,bait_type,promoter,pro_mid,otherEndID,oeID_mid,strand,up,down),tissue=tis) %>%
  left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% 
  left_join(dplyr::select(count_scaled,name,otherEndID,scaled_Nav,condition),by=c("bait_name"="name","otherEndID","tissue"="condition")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>%
  group_by(bait_name,promoter,tissue) %>% 
  mutate(class=if_else(is.na(class),"NEG","POS"), #complete the enhancer activity
         sum=sum(scaled_Nav)) %>% dplyr::filter(sum>0) %>% # filter out enh-pro pairs that have no read counts
  mutate(tenkb_Nav=if_else(oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000,0,scaled_Nav)) %>% mutate(max_dep=max(tenkb_Nav), rescaled_Nav=if_else( (scaled_Nav>max_dep | oeID_mid==bait_mid) ,100,scaled_Nav/max_dep*100)) %>% ungroup() %>% #set upper threshold for read count and rescale to 0-100
  dplyr::select(-c(otherEndID,tenkb_Nav,max_dep,up,down,sum)) %>% 
  group_by(bait_name,promoter,tissue) %>% arrange(bait_name,promoter,tissue,oeID_mid) %>%
  group_modify(~ { .x %>% rm_sd(.x$rescaled_Nav,3)}) %>% 
  mutate(newID=if_else(strand=="+",
                       if_else(oeID_mid<=bait_mid+10000, (oeID_mid-bait_mid-10000)/100,  #bait_mid+10000 is 0
                               if_else(oeID_mid>=pro_mid-10000,(oeID_mid-pro_mid+10000)/100+500, #pro_mid-10000 is 500
                                       (oeID_mid-bait_mid-10000)/(pro_mid-bait_mid-20000)*500)),  #scale in the middle
                       if_else(oeID_mid>=bait_mid-10000, (bait_mid-oeID_mid-10000)/100,    #otherwise
                               if_else(oeID_mid<=pro_mid+10000,(pro_mid-oeID_mid+10000)/100+500,
                                       (bait_mid-10000-oeID_mid)/(bait_mid-pro_mid-20000)*500)) 
                       )) %>% ungroup() #give new coordinates


#for in vivo act
a_vs_in <- distinct(bound,bait_name,class) %>% dplyr::filter(! is.na(class)) %>% dplyr::count(bait_name) %>% dplyr::filter(n==2)  #463 enhancers

#+/- 20kb
bound1 <- bound %>% dplyr::filter(abs(pro_mid-bait_mid)>30000 & tissue !="TK") %>% dplyr::filter(bait_name %in% a_vs_in$bait_name) %>% 
  separate(promoter, c("chr", "B"),sep="_",extra = "merge",remove = F) %>% mutate(oeID_mid1=oeID_mid+20000) %>% mutate(oeID_mid=oeID_mid1-40000) %>% as.data.table() %>% foverlaps(encode, by.x=c("chr","oeID_mid", "oeID_mid1"), type="any") %>% 
  dplyr::filter(is.na(encode_peak) | newID < -10 | newID > 510) %>%
  distinct(bait_name,promoter,bait_type,class,newID,tissue,.keep_all = T)

bound2 <- bound1 %>% group_by(bait_name,promoter,bait_type,class,newID,encode_peak) %>% summarise(m_Nav=mean(rescaled_Nav)) %>% ungroup() %>% pivot_wider(names_from = class, values_from = m_Nav) %>% mutate(ratio=POS/NEG, code="enh_invivo")
num_int=bound2 %>% distinct(bait_name,promoter) %>% nrow()
num_enh=bound2 %>% distinct(bait_name) %>% nrow()

#add 95% CI
supfig10a <- bound2 %>% ggplot(aes(x=newID,y=ratio)) + 
  stat_summary_bin(data=function(x) subset(x, newID < -110), geom="ribbon", binwidth= 30, fun.data=mean_cl_boot, fill="lightblue") +
  stat_summary_bin(data=function(x) subset(x, newID > -90), geom="ribbon", binwidth= 30, fun.data=mean_cl_boot, fill="lightblue") +
  stat_summary_bin(data=function(x) subset(x, newID < -110), geom="line", binwidth= 30, fun=mean, color="#0073C2FF",size=.3) +
  stat_summary_bin(data=function(x) subset(x, newID > -90), geom="line", binwidth= 30, fun=mean, color="#0073C2FF",size=.3) +
  geom_rect(data=data.frame(xmin = c(-130,-5,495), xmax = c(-70,5,505)), aes(xmin=xmin,xmax=xmax, ymin = -Inf, ymax = Inf),
            fill = "gray80",inherit.aes = F)  +
  coord_cartesian(xlim=c(-250, 750)) + 

  scale_x_continuous(limits=c(-250, 750), #trans = squish_trans(1000, 9000,10), 
                     breaks=c(-200,-100,0,500,600,700),labels = c("bait-10kb", "bait", "bait+10kb", "pro-10kb", "pro", "pro+10kb")) + 
  labs(x="Position", y="Mean interaction frequency ratio\n(activ vs. inactive)", title = paste0("After removal of encode peaks\n",num_enh," enhancers with ",num_int, " interactions, remove +/-20kb")
       ) + theme_bw() +
  theme(axis.text.x = element_text(angle = 15, vjust = 1, hjust = 1,size=6),axis.title.x=element_blank(),axis.title.y=element_text(size=7),axis.text.y = element_text(size=6),
        legend.position = c(0.6,0.75),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7), 
        legend.text = element_text(size=6),
        plot.title = element_text(size=7)) 
supfig10a
```

#boxplot

```{r}
#promoter
data1 <- bound %>% dplyr::filter(abs(pro_mid-bait_mid)>30000 & tissue !="TK") %>% dplyr::filter(bait_name %in% a_vs_in$bait_name) %>% dplyr::filter(newID==600) %>% group_by(bait_name,promoter,bait_type,class,newID) %>% summarise(m_Nav=mean(rescaled_Nav)) %>% ungroup() %>% pivot_wider(names_from = class, values_from = m_Nav) %>% mutate(ratio=POS/NEG, code="target_promoter") %>% dplyr::filter(is.finite(ratio)) %>% group_by(bait_name,promoter) %>% slice_sample(n = 1) %>% ungroup() %>% dplyr::select(bait_name,promoter,ratio,code)
#before remove encode
data2 <- bound %>% dplyr::filter(abs(pro_mid-bait_mid)>30000 & tissue !="TK") %>% dplyr::filter(bait_name %in% a_vs_in$bait_name) %>% dplyr::filter(newID <500 & newID>0) %>% group_by(bait_name,promoter,bait_type,class,newID) %>% summarise(m_Nav=mean(rescaled_Nav)) %>% ungroup() %>% pivot_wider(names_from = class, values_from = m_Nav) %>% mutate(ratio=POS/NEG, code="neg_before_encode") %>% dplyr::filter(is.finite(ratio)) %>% group_by(bait_name,promoter) %>% slice_sample(n = 1) %>% ungroup() %>% dplyr::select(bait_name,promoter,ratio,code)

#neg_region 
data3 <- bound2 %>% dplyr::filter(is.na(encode_peak)) %>% dplyr::filter(is.finite(ratio)) %>% dplyr::filter(newID <500 & newID>0) %>% group_by(bait_name,promoter) %>% slice_sample(n = 1) %>% ungroup() %>% mutate(code="neg_after_encode") %>% dplyr::select(bait_name,promoter,ratio,code)
#move 2mb as neg
data4 <- bound %>% dplyr::filter(abs(pro_mid-bait_mid)>30000 & tissue !="TK") %>% dplyr::filter(bait_name %in% a_vs_in$bait_name) %>% distinct(bait_name,promoter,tissue,bait_mid,pro_mid,class) %>%
  separate(promoter, c("chr", "B"),sep="_",extra = "merge",remove = F) %>%
  left_join(rmap,by=c("chr"="oe_chr","pro_mid"="oeID_mid")) %>%
  mutate(oeID=if_else(bait_mid<pro_mid,otherEndID+666,otherEndID-666)) %>% 
  left_join(dplyr::select(count_scaled,name,otherEndID,scaled_Nav,condition),by=c("bait_name"="name","tissue"="condition","oeID"="otherEndID")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>%  group_by(bait_name,promoter,class) %>% summarise(m_Nav=mean(scaled_Nav)) %>% ungroup() %>% pivot_wider(names_from = class, values_from = m_Nav) %>% mutate(ratio=POS/NEG, code="2mb") %>% dplyr::filter(is.finite(ratio)) %>% dplyr::select(bait_name,promoter,ratio,code)

data <- bind_rows(data1,data2,data3) %>% mutate(code=factor(code,levels=c("target_promoter", "neg_before_encode", "neg_after_encode")))

stat.test <- data %>% group_by(code) %>% wilcox_test(ratio ~ 1, mu = 1,paired = F) %>% 
  adjust_pvalue(method = "BH") %>% add_significance("p.adj") %>% add_x_position(x = "code") %>% #separate the p values to x.axis
  mutate(p.adj=if_else(p.adj.signif=="ns","ns",formatC(p.adj, format = "g", digits = 3)))

supfig1p <- data %>% mutate(log2fc=log2(ratio)) %>% ggboxplot(x="code",y="log2fc",outlier.size=0.5) + geom_hline(yintercept=0, color="red",linetype=2) + 
  stat_pvalue_manual(stat.test, label = "p.adj",y.position=3, tip.length = 0.01,size=2.5) +
  scale_x_discrete(labels=c("Interacting promoters","Intervening regions","Intervening region\nwithout ENCODE elements")) + labs(x="",y="E-P interaction frequency in active / inactive tissues (log2)") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1,size=6), axis.title.y=element_text(size=7),axis.text.y = element_text(size=6)) 

supfig1p
```



Fig2D. differential interactions

```{r}
tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")

tis4=c("brain","cf","limb","hrt","nt","tail","trunk") 
tis_group=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TL","TK"),tis_group=c(rep("brain",3),"cf",rep("limb",2),"hrt","nt","tail","trunk"))
bait_enh_act_4group <- bait_enh_act %>% left_join(tis_group,by="tissue") %>% group_by(elementID,tis_group) %>% dplyr::filter(!is.na(tis_group)) %>% summarise(class=dplyr::first(class)) %>% ungroup() %>% mutate(value=1) %>%
  complete(elementID,tis_group=tis4,fill = list(value = 0)) %>% add_count(elementID,wt=value) 

j=0
for (ref in tis4) {
    j=j+1
    act_one <- bait_enh_act_4group %>% dplyr::filter(tis_group==ref & value == 1) %>% .$elementID
    
    data <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% #just keep ep pairs match in at least one tissue
      dplyr::filter((bait_type=="enh") & (bait_name %in% act_one)) %>% #only include the pos enh baits
      group_by(bait_name,promoter) %>% dplyr::filter(score==max(score)) %>% distinct(bait_name,promoter,.keep_all = T) %>% #filter this because one promoter link to multiple target fragment, and if target is not the same in different tissues, it would be some bias
    ungroup() %>% mutate(target_mid=(target_start+target_end)/2,start=target_mid-2500,end=target_mid+2500) %>% as.data.table() %>% foverlaps(., new_rmap, by.x=c("target_chr","start", "end"), type="any", nomatch=NULL) %>% ##find the reads around the target
    dplyr::select(-c(oeID_mid2,start,end,gene)) %>% complete(nesting(bait_name,promoter,target_chr, target_mid,otherEndID,oeID_mid), tissue=tis) %>% ##complete the tissue combination
    left_join(dplyr::select(count_scaled,bait_name=name,otherEndID,scaled_Nav,tissue=condition), by=c("bait_name","otherEndID","tissue")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>%  left_join(tis_group,by="tissue") %>% group_by(bait_name,promoter,tissue,tis_group) %>% summarise(sum=sum(scaled_Nav)) %>% left_join(bait_enh_act_4group,by=c("bait_name"="elementID","tis_group")) %>%
      dplyr::filter(!(tis_group!=ref & value==1) | (tis_group==ref & value==0)) %>%  group_by(bait_name,promoter,value) %>% summarise(mean=mean(sum)) %>% ungroup() %>% mutate(value=if_else(value==1,"ref_tis","vs_ref"),group=ref) 
    assign(paste0("data",j),data)
}

#in order to show p.adjust to the plot, calculate p value separately 
data <- bind_rows(data1,data2,data3,data4,data5,data6,data7) %>% add_count(bait_name,promoter,group) %>% dplyr::filter(n==2) %>% mutate(group=factor(group,levels=c("brain","cf","limb","hrt","nt","tail","trunk"))) %>% dplyr::filter(! group %in% c("hrt","nt","tail","trunk"))
num=dplyr::count(data,group) %>% mutate(pair=n/2)

stat.test <- data %>% group_by(group) %>% wilcox_test(mean ~ value, paired = T) %>% adjust_pvalue(method = "BH") %>% add_significance("p.adj") %>% add_x_position(x = "group") %>% 
  mutate(p.adj=if_else(p.adj.signif=="ns","ns",formatC(p.adj, format = "g", digits = 3)))

#plot without removing linked tissues.
library(scales)
#read count box plot of ref and vs_ref
p1 <- data %>% group_by(bait_name,promoter,group) %>% mutate(remo=if_else(is.na(mean),1,0),n=sum(remo)) %>% dplyr::filter(n<1) %>% ungroup() %>% ggboxplot( x = "group", y = "mean", color = "value") + 
  stat_pvalue_manual(stat.test, label = "p.adj",y.position=8.5, tip.length = 0.01) + theme_classic() +
  scale_color_manual(values = c("#0073C2FF", "#EFC000FF"), name = "Type", labels = c("Active", "Inactive")) + 
  scale_x_discrete(labels=c("Brain","Face","Limb","Heart","Neuraltube","Tail","Trunk")) +
  scale_y_continuous(trans=pseudo_log_trans(base = 2),limits = c(0,500),breaks = c(0,1,4,16,64,256)) +
  labs(x="",y="Mean normalized read count on other end",title = "All interactions") +
  theme(legend.position = "top", axis.text.x = element_text(angle = 15, vjust = 1, hjust = 1))

#fold-change
p1 <- data %>% group_by(bait_name,promoter,group) %>% mutate(remo=if_else(is.na(mean),1,0),n=sum(remo)) %>% dplyr::filter(n<1) %>% dplyr::select(-c(remo,n)) %>% ungroup() %>% pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% 
  ggplot(aes(x=group,y=log2fc)) + geom_boxplot(outlier.size = 0.5) + theme_bw() + scale_x_discrete(labels=c("Brain","Face","Limb")) + labs(x="",y="E-P interaction frequency in active / inactive tissues (log2)" ) +
    annotate(geom = "text", x = stat.test$group, y = 3, label=paste0("n=",num$pair,"\n", stat.test$p.adj.signif) , size=1.5) +
  theme(legend.position="none",axis.title.x = element_text(size=7,angle=0, vjust = 0.5, hjust = 1),
        axis.title.y=element_text(size=7),axis.text = element_text(size=6) )


data %>% group_by(bait_name,promoter,group) %>% mutate(remo=if_else(is.na(mean),1,0),n=sum(remo)) %>% dplyr::filter(n<1) %>% dplyr::select(-c(remo,n)) %>% ungroup() %>% pivot_wider(names_from = value, values_from = mean) %>% mutate(fc=ref_tis/vs_ref) %>% dplyr::filter(is.finite(fc)) %>% 
   ggbarplot(x="group",y="fc",add="mean_se") + labs(title="so indeed, the mean interaction fold-change is just 1.3~1.7 fold")

#relationships between stars and fc
#use unpaired wilcox test because enhancers in different star catogories are independent
tis_group=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TL","TK"),tis_group=c(rep("brain",3),"cf",rep("limb",2),"hrt","nt","tail","trunk"))
stars <- read_tsv("../GATC.frags.pos.merged.tiles.mappability.anno.repeats.tsv") %>% distinct(elementID,stars) %>% right_join(bait_enh_act,by="elementID") %>% left_join(tis_group,by="tissue") %>% distinct(elementID,stars,class,tis_group) %>% mutate(stars=if_else(stars==6,5,stars))

data <- bind_rows(data1,data2,data3,data4,data5,data6,data7) %>% add_count(bait_name,promoter,group) %>% dplyr::filter(n==2) %>% mutate(group=factor(group,levels=c("brain","cf","limb","hrt","nt","tail","trunk"))) %>% left_join(stars,by=c("bait_name"="elementID","group"="tis_group")) %>% mutate(stars=factor(stars,levels=c(3,4,5))) 

stat.test <- data %>% group_by(stars,group) %>% wilcox_test(mean ~ value, paired = F) %>% #paired samples
  adjust_pvalue(method = "BH") %>% add_significance("p.adj") %>% add_x_position(x = "group") %>% #separate the p values to x.axis
  mutate(p.adj=if_else(p.adj.signif=="ns","ns",formatC(p.adj, format = "g", digits = 3)))
num=dplyr::count(data,group,stars) %>% mutate(pair=n/2)

#remove tail and trunk due to they have less than 10 interactions in at least one star category
supfig2b <- data %>% dplyr::filter(! group %in% c("tail","trunk")) %>% group_by(bait_name,promoter,group) %>%
  pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% ungroup() %>% 
  ggboxplot(y="log2fc",x="stars",fill="stars",outlier.size = 0.5) + 
  stat_compare_means(label = "p.format", comparisons = list(c("3","4"),c("4","5"),c("3","5")),method = "wilcox.test",paired = FALSE,tip.length = 0.01,label.y = c(2.8,3.2,3.6),size=2.5) + # Add pairwise comparisons p-value
  theme_bw() + labs(x="In vivo enhancer strength",y="E-P interaction frequency in\nactive / inactive tissues (log2)") +
  facet_grid(~ group) +
  theme_bw() + scale_fill_manual(values=c("#D2E2FF", "#97AAFF","#7990FF","#5A77E3")) +
  theme(axis.text.x = element_text(size=7),axis.title = element_text(size=8),axis.text.y = element_text(size=7),
        legend.position="none",
        plot.title = element_text(size=7))

p2 <- data %>% group_by(bait_name,promoter,group) %>% dplyr::filter(group=="brain") %>% 
  pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% ungroup() %>%
  ggboxplot(y="log2fc",x="stars",fill="stars",outlier.size = 0.5) + 
  stat_compare_means(label = "p.format", comparisons = list(c("3","4"),c("4","5"),c("3","5")),method = "wilcox.test",paired = FALSE,tip.length = 0.01,label.y = c(2.2,2.4,2.6),size=1.5) + 
  theme_bw() + labs(x="In vivo\nenhancer\nstrength\nin brain",y="E-P interaction frequency in active / inactive tissues (log2)") +
  theme(legend.position="none",axis.title.y = element_text(size=7,angle=0, vjust = 0.5, hjust = 1),
        axis.title.x=element_text(size=7),axis.text = element_text(size=6) ) + 
  scale_fill_manual(values=c("#D2E2FF", "#97AAFF","#7990FF","#5A77E3")) 
p2 <- ggpar(p2, orientation = "horiz")

fig2cd <- p1
supfig2b


fig2cd

```

Suppfig2fg. differential interactions in stable interactions

```{r}

data <- bind_rows(data1,data2,data3,data4,data5,data6,data7) %>% add_count(bait_name,promoter,group) %>% dplyr::filter(n==2) %>% left_join(interact_num,by=c("bait_name","promoter")) %>% mutate(group=factor(group,levels=c("brain","cf","limb","hrt","nt","tail"))) %>% dplyr::filter(! is.na(interact_tot)) %>% dplyr::filter(group!="tail")

stat.test <- data %>% group_by(group) %>% wilcox_test(mean ~ value, paired = T) %>% 
  adjust_pvalue(method = "BH") %>% add_significance("p.adj") %>% add_x_position(x = "group") %>% 
  mutate(p.adj=if_else(p.adj.signif=="ns","ns",formatC(p.adj, format = "g", digits = 2)))
num=dplyr::count(data,group) %>% mutate(pair=n/2)

#fold-change
supfig2f <- data %>% group_by(bait_name,promoter,group) %>% mutate(remo=if_else(is.na(mean),1,0),n=sum(remo)) %>% dplyr::filter(n<1) %>% dplyr::select(-c(remo,n)) %>% ungroup() %>% pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% ggplot(aes(y=group,x=log2fc)) + geom_boxplot(outlier.size = 0.5) + theme_bw() + scale_x_continuous(limits = c(-1.5,2.5)) + scale_y_discrete(labels=c("Brain","Face","Limb","Heart","Neuraltube")) + labs(y="",x="Invariant E-P interaction frequency in active / inactive tissues (log2)") +
    annotate(geom = "text", y = stat.test$group, x = 2.3, label=paste0("n=",num$pair,"\np=", stat.test$p.adj) , size=2.5) +
  theme(axis.text.x = element_text(size=7),axis.title.x = element_text(size=8),axis.text.y = element_text(size=7),
        legend.position = c(0.6,0.75),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7), 
        legend.text = element_text(size=6),
        plot.title = element_text(size=7))


#comparison between number of tissues enhancer active in and interaction stability
test <- ep_exp_5kb %>% distinct(bait_name,promoter,tissue,bait_type) %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% #just keep ep pairs match in at least one tissue
  dplyr::count(bait_name,promoter,bait_type,name="interact_tot") %>% dplyr::filter(bait_type=="enh") %>%
  group_by(bait_name) %>% dplyr::filter(interact_tot==max(interact_tot)) %>% ungroup() %>% distinct(bait_name,interact_tot) %>% 
  left_join(dplyr::count(bait_enh_act,elementID,name="active_tot"),by=c("bait_name"="elementID")) %>% 
  mutate(group=if_else(interact_tot<5,"tissue-specific",if_else(interact_tot>9,"invariant","no"))) %>% mutate(group=factor(group,levels = c("tissue-specific","invariant","no"))) 
stat.test <- test  %>% wilcox_test(active_tot ~ group, paired = F) %>% 
  adjust_pvalue(method = "fdr") %>% add_significance("p.adj") %>% add_x_position(x = "group")
  
supfig2g <- test %>% dplyr::filter(group != "no") %>% ggboxplot(y="active_tot",x="group",outlier.size = 0.1) + scale_y_continuous(limits =c(0,9)) +
   stat_pvalue_manual(dplyr::filter(stat.test,group1 != "no" & group2 != "no"), label = "p.adj",y.position=9,tip.length = 0.03,size=2) + # Add pairwise comparisons p-value
  labs(y="Number of tissues that\nenhancer bait active in",x="") + theme_bw() +
  theme(axis.text.x = element_text(size=7,angle = 30, vjust = 1, hjust = 1),axis.title.y = element_text(size=8),axis.text.y = element_text(size=7), axis.title.x=element_blank(),
        plot.title = element_text(size=7))
supfig2f
supfig2g

```

plot all ten tissues

```{r}
tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")
j=0
for (ref in tis) {
    j=j+1
    act_one <- bait_enh_act %>% dplyr::filter(tissue==ref & class == "POS") %>% .$elementID
    
    data <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% #just keep ep pairs match in at least one tissue
      dplyr::filter((bait_type=="enh") & (bait_name %in% act_one)) %>% #only include the pos enh baits
      group_by(bait_name,promoter) %>% dplyr::filter(score==max(score)) %>% distinct(bait_name,promoter,.keep_all = T) %>% 
    ungroup() %>% mutate(target_mid=(target_start+target_end)/2,start=target_mid-2500,end=target_mid+2500) %>% as.data.table() %>% foverlaps(., new_rmap, by.x=c("target_chr","start", "end"), type="any", nomatch=NULL) %>% ##find the reads around the target
    dplyr::select(-c(oeID_mid2,start,end,gene)) %>% complete(nesting(bait_name,promoter,target_chr, target_mid,otherEndID,oeID_mid), tissue=tis) %>% ##complete the tissue combination
    left_join(dplyr::select(count_scaled,bait_name=name,otherEndID,scaled_Nav,tissue=condition), by=c("bait_name","otherEndID","tissue")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>% group_by(bait_name,promoter,tissue) %>% summarise(sum=sum(scaled_Nav)) %>% left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% mutate(value=if_else(is.na(class),0,1)) %>%
      dplyr::filter(!(tissue!=ref & value==1) | (tissue==ref & value==0)) %>%  group_by(bait_name,promoter,value) %>% summarise(mean=mean(sum)) %>% ungroup() %>% mutate(value=if_else(value==1,"ref_tis","vs_ref"),group=ref) 
    assign(paste0("data",j),data)
}

#fold-change for non significant tissues (tail and trunk)
data <- bind_rows(data1,data2,data3,data4,data5,data6,data7,data8,data9,data10) %>% add_count(bait_name,promoter,group) %>% dplyr::filter(n==2) %>% mutate(group=factor(group,levels=tis)) 
num=dplyr::count(data,group) %>% mutate(pair=n/2)

stat.test <- data %>% group_by(group) %>% wilcox_test(mean ~ value, paired = T) %>%
  adjust_pvalue(method = "BH") %>% add_significance("p.adj") %>% add_x_position(x = "group") %>%
  mutate(p.adj=if_else(p.adj.signif=="ns","ns",formatC(p.adj, format = "g", digits = 3)))

#fold-change
supfig2a <- data %>% group_by(bait_name,promoter,group) %>% mutate(remo=if_else(is.na(mean),1,0),n=sum(remo)) %>% dplyr::filter(n<1) %>% dplyr::select(-c(remo,n)) %>% ungroup() %>% pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% 
  ggplot(aes(x=group,y=log2fc)) + geom_boxplot(outlier.size = 0.3) + theme_bw() + 
  labs(y="E-P interaction frequency in\nactive / inactive tissues (log2)",x="") +
  annotate(geom = "text", x= stat.test$group, y = 3.6, label=paste0("n=",num$pair,"\np=", stat.test$p.adj) , size=2.5) +
  theme(axis.text.x = element_text(size=7),axis.title.y = element_text(size=8),axis.text.y = element_text(size=7),
        legend.position = c(0.6,0.75),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7), 
        legend.text = element_text(size=6),
        plot.title = element_text(size=7))

supfig2a

```


what about the interaction frequencies for inactive genes? definition of
active and inactive genes, average expression across tissues, TPM>0.5

```{r}
neg_pair <- ep_exp_5kb %>% dplyr::filter(bait_type=="enh") %>% separate(gene,paste0("gene_",seq(1:15)),sep=",") %>% pivot_longer(!c(target_chr, target_start,target_end,bait_chr,bait_start,bait_end,bait_name,score,tissue,promoter,overlapping_bp, target_mid, otherEndID,dist,overlap_len,bait_type), names_to = "name", values_to = "gene_name") %>% mutate(gene_name=gsub(" ","",gene_name)) %>%
  dplyr::filter(!is.na(gene_name)) %>% dplyr::filter(tissue %in% c("FB","MB","HB","NT","CF","FL","HL","HR")) %>% 
  left_join(rna_seq_exp,by=c("gene_name","tissue")) %>% mutate(exp=if_else(mean_TPM<0.5,"no","yes")) %>% #try different stringency
  left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% 
  dplyr::filter(exp=="no") %>%
  arrange(bait_name,gene_name,tissue) %>% 
  distinct(bait_name,promoter) %>% unite("pair",c("bait_name","promoter"))


tis=c("FB","MB","HB","NT","CF","FL","HL")  #,"HR","TL","TK"  no expression data, so don't know if express or not
j=0
for (ref in tis) {
    j=j+1
    act_one <- bait_enh_act %>% dplyr::filter(tissue==ref & class == "POS") %>% .$elementID
    
    data <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% mutate(exp_gene=if_else(pair %in% neg_pair$pair,"neg",NULL)) %>% mutate(exp_gene=if_else(pair %in% pos_pos_int$pair,"pos",exp_gene)) %>% dplyr::filter(!is.na(exp_gene)) %>% 
      dplyr::filter((bait_type=="enh") & (bait_name %in% act_one)) %>% #only include the pos enh baits
      group_by(bait_name,promoter) %>% dplyr::filter(score==max(score)) %>% distinct(bait_name,promoter,.keep_all = T) %>% 
    ungroup() %>% mutate(target_mid=(target_start+target_end)/2,start=target_mid-2500,end=target_mid+2500) %>% as.data.table() %>% foverlaps(., new_rmap, by.x=c("target_chr","start", "end"), type="any", nomatch=NULL) %>% ##find the reads around the target
    dplyr::select(-c(oeID_mid2,start,end,gene)) %>% complete(nesting(bait_name,promoter,target_chr, target_mid,otherEndID,oeID_mid,exp_gene), tissue=tis) %>% ##complete the tissue combination
    left_join(dplyr::select(count_scaled,bait_name=name,otherEndID,scaled_Nav,tissue=condition), by=c("bait_name","otherEndID","tissue")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>% group_by(bait_name,promoter,tissue,exp_gene) %>% summarise(sum=sum(scaled_Nav)) %>% left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% mutate(value=if_else(is.na(class),0,1)) %>%
      dplyr::filter(!(tissue!=ref & value==1) | (tissue==ref & value==0)) %>%  group_by(bait_name,promoter,value,exp_gene) %>% summarise(mean=mean(sum)) %>% ungroup() %>% mutate(value=if_else(value==1,"ref_tis","vs_ref"),group=ref) 
    assign(paste0("data",j),data)
}

#fold-change for non significant tissues (tail and trunk)
data <- bind_rows(data1,data2,data3,data4,data5,data6,data7) %>% #,data8,data9,data10
  add_count(bait_name,promoter,group) %>% dplyr::filter(n==2) %>%
  mutate(group=factor(group,levels=tis),exp_gene=factor(exp_gene,levels = c("pos","neg"))) 

num=dplyr::count(data,group,exp_gene) %>% mutate(pair=n/2)

stat.test <- data %>% group_by(group,exp_gene) %>% wilcox_test(mean ~ value, paired = T) %>% 
  adjust_pvalue(method = "BH") %>% add_significance("p.adj") %>% add_x_position(x = "group") %>% mutate(x=if_else(exp_gene=="pos",xmin,xmax)) %>% #separate the p values to x.axis
  mutate(p.adj=if_else(p.adj.signif=="ns","ns",formatC(p.adj, format = "g", digits = 2)))

#fold-change
supfig2a1 <- data %>% group_by(bait_name,promoter,group) %>% mutate(remo=if_else(is.na(mean),1,0),n=sum(remo)) %>% dplyr::filter(n<1) %>% dplyr::select(-c(remo,n)) %>% ungroup() %>% pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% 
  ggplot(aes(x=group,y=log2fc,color=exp_gene)) + geom_boxplot(outlier.size = 0.3) + theme_bw() + 
  labs(y="E-P interaction frequency in\nactive / inactive tissues (log2)",x="",title="TPM>0.5 or TPM<0.5") +
  annotate(geom = "text", x= stat.test$x, y = rep(c(3.8,3.2),7), label=paste0("n=",num$pair,"\np=", stat.test$p.adj) , size=2,angle=45) +
    scale_color_manual(values=c("#ec7f26e3","#4d77b1e8"),name = "Expressing gene", labels = c("Yes","No")) +
  theme(axis.text.x = element_text(size=7),axis.title.y = element_text(size=8),axis.text.y = element_text(size=7),
        legend.position = "top",
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7), 
        legend.text = element_text(size=6),
        plot.title = element_text(size=7))

supfig2a1
```

what about interactions within TAD or not (\<2Mb) or more than 2mb and trans

```{r}
ep_exp_trans <- read_tsv("~/ensembl_regulatory/chicago3_ditag_promoter_target_gene_preliminary_result_including_trans.tsv") %>% mutate(target_mid=as.integer((target_start+target_end)/2)) %>% left_join(rmap,by=c("target_chr"="oe_chr","target_mid"="oeID_mid")) %>% group_by(bait_name,tissue,otherEndID) %>% mutate(promoter_start=promoter_start+5000,promoter_end=promoter_end-5000) %>%
  mutate(dist=abs((promoter_start+promoter_end)/2-target_mid), overlap_len=if_else(target_end>=promoter_start & target_end<=promoter_end, target_end-promoter_start+1, if_else(target_start>=promoter_start & target_start<=promoter_end, promoter_end-target_start+1,0))) %>% dplyr::filter(overlap_len==max(overlap_len)) %>% dplyr::filter(dist==min(dist)) %>% ungroup() %>% 
  mutate(promoter_start=promoter_start-5000,promoter_end=promoter_end+5000) %>%
  unite("promoter",starts_with("promoter")) %>% #group them by promoter coordinate
  mutate(bait_type=if_else(bait_name %in% pos_enh_bait$bait_name, "enh", if_else(bait_name %in% prom_bait$bait_name, "pro", "neg"))) %>% mutate(ep_dist=abs((bait_start+bait_end)/2-target_mid)) %>%
  as.data.table() %>% foverlaps(tad, by.x=c("target_chr","target_start", "target_end"), type="any") %>% foverlaps(tad, by.x=c("bait_chr","bait_start", "bait_end"), type="any") %>% mutate(tad=if_else(ep_dist<2000000,if_else(id==i.id,"within","across"),">2Mb")) %>% dplyr::filter(!is.na(tad))

#only golden set
ep_exp_trans <- ep_exp_trans %>% unite("pair",c("bait_name","promoter"),remove = F) %>% dplyr::filter(pair %in% pos_pos_int$pair)


tis=c("FB","MB","HB","NT","CF","FL","HL","HR")  
j=0
for (ref in tis) {
    j=j+1
    act_one <- bait_enh_act %>% dplyr::filter(tissue==ref & class == "POS") %>% .$elementID
    
    data <- ep_exp_trans %>% 
      dplyr::filter((bait_type=="enh") & (bait_name %in% act_one)) %>% 
      group_by(bait_name,promoter) %>% dplyr::filter(score==max(score)) %>% distinct(bait_name,promoter,.keep_all = T) %>% 
    ungroup() %>% mutate(target_mid=(target_start+target_end)/2,start=target_mid-2500,end=target_mid+2500) %>% as.data.table() %>% foverlaps(., new_rmap, by.x=c("target_chr","start", "end"), type="any", nomatch=NULL) %>% ##find the reads around the target
    dplyr::select(-c(oeID_mid2,start,end,gene)) %>% complete(nesting(bait_name,promoter,target_chr, target_mid,otherEndID,oeID_mid,tad), tissue=tis) %>% ##complete the tissue combination
    left_join(dplyr::select(count_scaled,bait_name=name,otherEndID,scaled_Nav,tissue=condition), by=c("bait_name","otherEndID","tissue")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>% group_by(bait_name,promoter,tissue,tad) %>% summarise(sum=sum(scaled_Nav)) %>% left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% mutate(value=if_else(is.na(class),0,1)) %>%
      dplyr::filter(!(tissue!=ref & value==1) | (tissue==ref & value==0)) %>%  group_by(bait_name,promoter,value,tad) %>% summarise(mean=mean(sum)) %>% ungroup() %>% mutate(value=if_else(value==1,"ref_tis","vs_ref"),group=ref) 
    assign(paste0("data",j),data)
}

#fold-change for non significant tissues (tail and trunk)
data <- bind_rows(data1,data2,data3,data4,data5,data6,data7,data8) %>% 
  add_count(bait_name,promoter,group) %>% dplyr::filter(n==2) %>%
  group_by(bait_name,promoter) %>% mutate(min=min(mean)) %>% ungroup() %>% dplyr::filter(min>0) %>%   mutate(group=factor(group,levels=tis),tad=factor(tad,levels=c("within","across",">2Mb"))) %>% dplyr::filter(tad != ">2Mb") 


num=dplyr::count(data,group,tad) %>% mutate(pair=n/2)

stat.test <- data %>% group_by(group,tad) %>% wilcox_test(mean ~ value, paired = T) %>% 
  adjust_pvalue(method = "BH") %>% add_significance("p.adj") %>% add_x_position(x = "group") %>% 
  mutate(p.adj=if_else(p.adj.signif=="ns","ns",formatC(p.adj, format = "g", digits = 2)),x=if_else(tad=="within",xmin,if_else(tad=="across",x,xmax)))

#fold-change for each tissue
data %>% group_by(bait_name,promoter,group) %>% mutate(remo=if_else(is.na(mean),1,0),n=sum(remo)) %>% dplyr::filter(n<1) %>% dplyr::select(-c(remo,n)) %>% ungroup() %>% 
  #mutate(mean=mean+0.00001) %>% 
  pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% 
  #ggplot(aes(x=group,y=mean, color=tad,fill=value)) + 
  ggplot(aes(x=group,y=log2fc, color=tad)) + 
  geom_boxplot(outlier.size = 0.3) + theme_bw() +
  coord_cartesian(ylim= c(-2,4)) +
  labs(y="E-P interaction frequency in\nactive / inactive tissues (log2)",x="",
       title = "golden set"
       #title = "All interactions"
       ) +
  annotate(geom = "text", x= stat.test$x, y = rep(c(4,3.5),8), label=paste0("n=",num$pair,"\np=", stat.test$p.adj) , size=2,angle=45) +
  scale_color_manual(values=c("#dd8581ff","#e9b86fff","#77b27bff"),name = "TAD boundary") +
  theme(axis.text.x = element_text(size=7),axis.title.y = element_text(size=8),axis.text.y = element_text(size=7),
        legend.position = "top",
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7), 
        legend.text = element_text(size=6),
        plot.title = element_text(size=7))

#combine
stat.test <- data %>% group_by(bait_name,promoter,value,tad) %>% summarise(mean=mean(mean)) %>% group_by(tad) %>% wilcox_test(mean ~ value, paired = T) %>% add_x_position(x="tad") %>% 
  mutate(p=formatC(p, format = "g", digits = 2))

num <- data %>% distinct(bait_name,promoter,tad) %>% dplyr::count(tad)

supfig1m <- data %>% group_by(bait_name,promoter,value,tad) %>% summarise(mean=mean(mean)) %>% ungroup() %>% #combine all tissues
  pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% 
  ggboxplot(x="tad",y="log2fc",color="tad",outlier.size = 0.3) + theme_bw() + 
  labs(y="E-P interaction frequency in\nactive / inactive tissues (log2)",x="") +
  annotate(geom = "text", x= num$tad, y = 3, label=paste0("n = ",num$n,"\np = ",stat.test$p) , size=3) +
  theme(axis.text.x = element_text(size=7),axis.title.y = element_text(size=8),axis.text.y = element_text(size=7),
        legend.position = "none",
        plot.title = element_text(size=7))
supfig1m

```

fig4 heatmap for interaction frequency across tissues #golden set: 469 enhancers, 969 interactions

```{r}
library("pheatmap")
tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")
pos_pos_int %>% dplyr::filter(grepl("hs2308",pair))
ep_exp_5kb %>% dplyr::filter(bait_name=="hs2308")

test <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% group_by(bait_name,promoter) %>% dplyr::filter(score==max(score)) %>% distinct(bait_name,promoter,.keep_all = T) %>% ungroup() %>% mutate(target_mid=(target_start+target_end)/2,start=target_mid-2500,end=target_mid+2500) %>% as.data.table() %>% foverlaps(., new_rmap, by.x=c("target_chr","start", "end"), type="any", nomatch=NULL) %>% ##find the reads around the target
  dplyr::select(-c(oeID_mid2,start,end,gene)) %>% complete(nesting(bait_name,promoter,target_chr, target_mid,otherEndID,oeID_mid,pair), tissue=tis) %>% ##complete the tissue combination
  left_join(dplyr::select(count_scaled,bait_name=name,otherEndID,scaled_Nav,tissue=condition), by=c("bait_name","otherEndID","tissue")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>% group_by(bait_name,promoter,tissue,pair) %>% summarise(sum=sum(scaled_Nav)) %>% left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% ungroup()

test %>% distinct(pair) %>% nrow() #969 interactions

df <- test %>% dplyr::select(pair,tissue,sum) %>% pivot_wider(names_from = pair, values_from = sum) %>% column_to_rownames("tissue") %>% scale()
df %>% as_tibble(rownames = "tissue") %>% dplyr::select(tissue, contains("ZRS") | contains("hs654") | contains("hs266") | contains("hs267") | contains("hs268") | contains("hs853") | contains("hs1431") | contains("hs2306")) %>% column_to_rownames("tissue") %>% pheatmap(clustering_method="ward.D")


df <- test %>% dplyr::select(pair,tissue,sum) %>% pivot_wider(names_from = pair, values_from = sum) %>% column_to_rownames("tissue")
pheatmap(df,show_colnames = F,clustering_method="ward.D")

```

proportion to the max of each ep interaction

```{r}
#library(heatmaply)
df <- test %>% group_by(pair) %>% mutate(ratio=sum/max(sum)) %>% ungroup() %>% dplyr::select(pair,tissue,ratio) %>% pivot_wider(names_from = tissue, values_from = ratio) %>% column_to_rownames("pair")

df %>% as_tibble(rownames = "tissue") %>% dplyr::select(tissue, contains("ZRS") | contains("hs654") | contains("hs266") | contains("hs267") | contains("hs268") | contains("hs853") | contains("hs1431") | contains("hs2308")) %>% column_to_rownames("tissue") %>% pheatmap(clustering_method="ward.D")

#find optimal k
# function to compute total within-cluster sum of square 
library(cluster)
avg_sil <- function(k) {
  km.res <- kmeans(df, centers = k, nstart = 25)
  ss <- silhouette(km.res$cluster, dist(df))
  mean(ss[, 3])
}
# Compute and plot wss for k = 2 to k = 15
k.values <- 2:15

#k-means
set.seed(42)
k <- df %>% kmeans(10,nstart=30)

#ward.D cluster k-means groups
hclust <- k$centers %>% dist() %>% hclust( method = "ward.D")
plot(hclust)
hclust$order

int_num <- ep_exp_5kb %>% distinct(bait_name,promoter,tissue,bait_type) %>% unite("pair",c("bait_name","promoter"),remove = F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% dplyr::count(pair,name="interact_tot") 

pair_col <- int_num %>% mutate(interact_tot=if_else(interact_tot>9,1,0)) %>% mutate(interact_tot=factor(interact_tot)) %>% column_to_rownames("pair")
Var1 = c("#CA4B00", "#FEF5EC")
names(Var1) = c("1", "0")
ann_colors = list(interact_tot = Var1)


dfc <- cbind(df, id=seq(nrow(df)), cluster=k$cluster)
# Add id sort, the id number ordered by cluster 
dfc$idsort <- dfc$id[order(dfc$cluster)]
dfc$idsort <- order(dfc$idsort)

# use reshape2::melt to create data.frame in long format
dfm <- dfc %>% rownames_to_column("pair") %>% left_join(int_num,by="pair") %>% mutate(interact_tot=if_else(interact_tot>9,1,0)) %>% pivot_longer(!c("id", "idsort","pair")) %>% dplyr::filter(name != "cluster") %>% mutate(name = factor(name, levels = c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK","interact_tot")))

ggplot(dfm, aes(x=name, y=idsort)) + geom_tile(aes(fill=value)) + scale_fill_gradient(low="white", high="blue") + labs(title="k=5")


colfunc <- colorRampPalette(c("white","#DEEFBD","#006F4B"))
cols <- colfunc(50)
  
#or two color breaks
makeColorRampPalette <- function(colors, cutoff.fraction, num.colors.in.palette)
{
  stopifnot(length(colors) == 4)
  ramp1 <- colorRampPalette(c("white","#BEE09F","#00513B"))(num.colors.in.palette * cutoff.fraction)
  ramp2 <- colorRampPalette(c("white","white","#2e5a9f"))(num.colors.in.palette * (1 - cutoff.fraction))
  return(c(ramp1, ramp2))
}

#add annotation color
pair_act <- mutate(test,enh_act=if_else(is.na(class),2,3),tissue=paste0("act_",tissue)) %>% dplyr::select(pair,tissue,enh_act) %>% pivot_wider(names_from = tissue, values_from = enh_act) %>% dplyr::select(pair,act_FB,act_MB,act_HB,act_NT,act_CF,act_LB=act_FL,act_HR,act_TL,act_TK)
pair_col <- int_num %>% mutate(interact_tot=if_else(interact_tot>9,1,0)) %>% mutate(interact_tot=factor(interact_tot)) %>% left_join(as_tibble(k$cluster, rownames = "pair"),by="pair") %>% column_to_rownames("pair")

Var1 = c("#d97d28", "white")
names(Var1) = c("1", "0")
ann_colors = list(interact_tot = Var1, value=hcl.colors(10, "Purples"))

#customized order
ord=c(1,6,10,3,7,5,8,4,2,9)
m <- test %>% group_by(pair) %>% mutate(ratio=sum/max(sum),tissue=paste0("int_",tissue)) %>% dplyr::select(pair,tissue,ratio) %>% pivot_wider(names_from = tissue, values_from = ratio) %>% left_join(as_tibble(k$cluster, rownames = "pair"),by="pair") %>% 
  left_join(pair_act,by="pair") %>% 
  left_join(data.frame(value=ord,id=seq(1,10)), by="value") %>% rowwise() %>% mutate(sum=sum(across(int_CF:int_TL)),blank=1.5,int_LB=(int_FL+int_HL)/2) %>% ungroup() %>% arrange(id,sum) %>% column_to_rownames("pair") %>% dplyr::select(int_FB,int_MB,int_HB,int_NT,int_CF,int_LB,int_HR,int_TL,int_TK,blank,starts_with("act")) 
distmat <- dist(t(m))


cutoff.distance <- 20  
cols <- makeColorRampPalette(c("white", "red",    # distances 0 to 3 colored from white to red
                               "green", "black"), # distances 3 to max(distmat) colored from green to black
                             cutoff.distance / max(distmat),
                             100)

fig2e <- m %>% pheatmap(show_rownames = F,clustering_method="ward.D", 
           annotation_row = pair_col,annotation_colors = ann_colors, 
           #cellwidth = 8, cellheight = 2,
           fontsize=8,  #size for text
           annotation_legend=F,
           #breaks=c(0,0.9,1),
           #main="k-means",
           cluster_rows=F,cluster_cols=F,color = cols) 

```

also plot logistic regression and odd ratio for each tissue

```{r}
tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")

test <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% group_by(bait_name,promoter) %>% dplyr::filter(score==max(score)) %>% distinct(bait_name,promoter,.keep_all = T) %>% ungroup() %>% mutate(target_mid=(target_start+target_end)/2,start=target_mid-2500,end=target_mid+2500) %>% as.data.table() %>% foverlaps(., new_rmap, by.x=c("target_chr","start", "end"), type="any", nomatch=NULL) %>% ##find the reads around the target
  dplyr::select(-c(oeID_mid2,start,end,gene)) %>% complete(nesting(bait_name,promoter,target_chr, target_mid,otherEndID,oeID_mid,pair), tissue=tis) %>% ##complete the tissue combination
  left_join(dplyr::select(count_scaled,bait_name=name,otherEndID,scaled_Nav,tissue=condition), by=c("bait_name","otherEndID","tissue")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>% group_by(bait_name,promoter,tissue,pair) %>% summarise(sum=sum(scaled_Nav)) %>% left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% ungroup()

pair_chi <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% distinct(pair,tissue,score) %>%  mutate(tissue=if_else(tissue %in% c("FL","HL"), "LB", tissue)) %>% group_by(pair,tissue) %>% summarise(score=mean(score)) %>% ungroup() %>% mutate(tissue=paste0("score_",tissue),score=1) %>% pivot_wider(names_from = tissue, values_from = score, values_fill = 0)

data <- test %>% group_by(pair) %>% mutate(ratio=sum/max(sum),tissue=paste0("int_",tissue)) %>% dplyr::select(pair,tissue,ratio) %>% pivot_wider(names_from = tissue, values_from = ratio) %>%
  left_join(pair_act,by="pair") %>% mutate(int_LB=(int_FL+int_HL)/2) %>% dplyr::select(-int_FL,-int_HL) %>%
  left_join(pair_chi,by=c("pair")) %>% 
  pivot_longer(!pair, names_to = "name",values_to = "num") %>% separate(name,c("type","tissue"),sep="_") %>%
  pivot_wider(names_from = type, values_from = num) %>% arrange(pair,tissue) %>% 
  mutate(act=if_else(act==2,0,1)) 
num=nrow(distinct(data,pair))

#use chi-squre to calculate p value, num is the same for all tissues
rm(dataset)
for (ref in c("FB","MB","HB","NT","CF","LB","HR","TL","TK")) {
  fmod <- glm(int~act, dplyr::filter(data, tissue==ref), family = "binomial") 
  pval <- pchisq(summary(fmod)$null-summary(fmod)$deviance, summary(fmod)$df.null-summary(fmod)$df.residual, lower.tail=FALSE) %>% formatC(format = "g", digits = 3)
  odd_rat <- dplyr::filter(data, tissue==ref) %>% ungroup() %>% dplyr::count(act,score) %>% mutate(score=if_else(score==0,"no","yes")) %>%
    pivot_wider(names_from = score, values_from = n) %>% mutate(ratio=yes/no) 
  odd = (dplyr::filter(odd_rat,act==1)$ratio / dplyr::filter(odd_rat,act==0)$ratio) %>% formatC(format = "g", digits = 3)
    if (!exists("dataset")) {
      dataset <- data.frame(tissue=ref,p_value=pval,odd=odd)
    } else {
      dataset <- dataset %>% add_row(tissue=ref,p_value=pval,odd=odd)
    }
}
dataset <- dataset %>% mutate(tissue=factor(tissue,levels=c("FB","MB","HB","NT","CF","LB","HR","TL","TK")))

data %>% mutate(tissue=factor(tissue,levels=c("FB","MB","HB","NT","CF","LB","HR","TL","TK")))  %>% 
  ggplot(aes(x=int, y=act,color=tissue)) +
  geom_point(alpha=.05) +
  facet_wrap(~ tissue) +
  stat_smooth(method="glm", se=F, 
              fullrange=TRUE,  #view the full logistic distribution
              method.args = list(family=binomial)) +
  geom_text(data    = dataset,
  mapping = aes(x = 0.1, y = 0.5, label = paste0("P = ",p_value,"\nodd ratio = ",odd)),hjust   = -0.1,vjust   = -1,size=2.5) +
  
  labs(y="Enhancer activity",x="Interaction frequencies") + 
  theme(plot.title = element_text(size=6), axis.title=element_text(size=6), axis.text=element_text(size=5))

#plot across all tissues
fmod <- glm(int~act, data, family = "binomial") 
pval <- pchisq(summary(fmod)$null-summary(fmod)$deviance, summary(fmod)$df.null-summary(fmod)$df.residual, lower.tail=FALSE) %>% formatC(format = "g", digits = 3)
odd_rat <- data %>% ungroup() %>% dplyr::count(act,score) %>% mutate(score=if_else(score==0,"no","yes")) %>%
  pivot_wider(names_from = score, values_from = n) %>% mutate(ratio=yes/no) 
odd = (dplyr::filter(odd_rat,act==1)$ratio / dplyr::filter(odd_rat,act==0)$ratio) %>% formatC(format = "g", digits = 3)

supfig2k <- data %>% ggplot(aes(x=int, y=act)) +
  geom_point(alpha=.01,color="#5891f8ff",size=0.5) +
  stat_smooth(method="glm", se=T, color="black",
              fullrange=TRUE,  #view the full logistic distribution
              method.args = list(family=binomial)) +
  annotate("text", x=0.1, y=0.9, hjust = 0, vjust = 2.5, label=paste0("P = ",pval,"\nEP num = ",num,"\nOdd ratio = ", odd),size=2.5) +
  labs(y="Enhancer activity",x="Relative nteraction frequencies") + theme_bw() +
  theme(axis.title=element_text(size=7), axis.text=element_text(size=6))
supfig2k

```

heat map for gene activity and ep interaction, scatter plot

```{r}
tis=c("FB","MB","HB","NT","CF","FL","HL","HR")  #,"TL","TK" no expression data
test <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% group_by(bait_name,promoter) %>% dplyr::filter(score==max(score)) %>% distinct(bait_name,promoter,.keep_all = T) %>% ungroup() %>% mutate(target_mid=(target_start+target_end)/2,start=target_mid-2500,end=target_mid+2500) %>% as.data.table() %>% foverlaps(., new_rmap, by.x=c("target_chr","start", "end"), type="any", nomatch=NULL) %>% ##find the reads around the target
  dplyr::select(-c(oeID_mid2,start,end)) %>% complete(nesting(bait_name,promoter,target_chr, target_mid,otherEndID,oeID_mid,pair,gene), tissue=tis) %>% ##complete the tissue combination
  left_join(dplyr::select(count_scaled,bait_name=name,otherEndID,scaled_Nav,tissue=condition), by=c("bait_name","otherEndID","tissue")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>% group_by(bait_name,promoter,tissue,pair,gene) %>% summarise(int_sum=sum(scaled_Nav)) %>% left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% ungroup() %>% mutate(len=str_count(gene,",")) %>% arrange(desc(len)) %>% separate(gene,paste0("gene",seq(1,13)),sep=",") %>% pivot_longer(!c(bait_name,promoter,tissue,pair,int_sum,class,len),names_to = "name",values_to = "gene_name") %>% dplyr::filter(!is.na(gene_name)) %>%
  mutate(gene_name=gsub(" ","",gene_name)) %>% dplyr::filter(!is.na(gene_name)) %>% #dplyr::filter(grepl("Hoxa1",gene_name)) %>% 
  left_join(rna_seq_exp,by=c("gene_name","tissue")) %>% dplyr::filter(stage=="E115") %>% 
  group_by(pair,gene_name) %>% mutate(ratio=mean_TPM/max(mean_TPM)) %>%  #get expression ratio
  group_by(bait_name,promoter,pair,tissue,int_sum,class) %>% summarise(m_ra=mean(ratio)) %>% ungroup() %>%  #multiple target genes, so integrate
  as.data.table()
test %>% distinct(pair) %>% nrow() #969 interactions

#k-means
set.seed(42)
df <- test %>% group_by(pair) %>% mutate(ratio=int_sum/max(int_sum)) %>% dplyr::select(pair,tissue,ratio) %>% pivot_wider(names_from = tissue, values_from = ratio) %>% column_to_rownames("pair")
k <- df %>% kmeans(10,nstart=30)

#ward.D cluster k-means groups
hclust <- k$centers %>% dist() %>% hclust( method = "ward.D")
plot(hclust)
hclust$order
nrow(df) # how many interactions 

int_num <- ep_exp_5kb %>% distinct(bait_name,promoter,tissue,bait_type) %>% unite("pair",c("bait_name","promoter"),remove = F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% dplyr::count(pair,name="interact_tot") 

```

calculate the correlation between interaction and activity or expression

```{r}
data <- data2 %>% group_by(pair) %>% mutate(ratio=int_sum/max(int_sum),tissue=paste0("int_",tissue)) %>% dplyr::select(pair,tissue,ratio) %>% pivot_wider(names_from = tissue, values_from = ratio) %>%
  left_join(pair_act,by="pair") %>% mutate(int_LB=(int_FL+int_HL)/2) %>% dplyr::select(-int_FL,-int_HL) %>%
  left_join(pair_exp,by="pair") %>%
  pivot_longer(!pair, names_to = "name",values_to = "num") %>% separate(name,c("type","tissue"),sep="_") %>% pivot_wider(names_from = type, values_from = num) %>% arrange(pair,tissue) %>% mutate(act=if_else(act==2,0,1),exp=exp-4) 
num=nrow(distinct(data,pair))
#use chi-squre to calculate p value
fmod <- glm(int~act, data, family = "binomial") ##"full" mod
summary(fmod)
#To get the significance for the overall model we use the following command:
#pchisq(Null deviance - Residual deviance, degrees of freedom - degrees of freedom, lower.tail=FALSE) #lower.tail=F so that we can get the right tail probability (p value)
pval <- pchisq(summary(fmod)$null-summary(fmod)$deviance, summary(fmod)$df.null-summary(fmod)$df.residual, lower.tail=FALSE) %>% formatC(format = "g", digits = 3)

p1 <- data %>% ggplot(aes(x=int, y=act#,color=tissue
                    )) +
  geom_point(alpha=.01,color="#5891f8ff",size=0.5) +
  stat_smooth(method="glm", se=T, color="black",
              fullrange=TRUE,  #view the full logistic distribution
              method.args = list(family=binomial)) +
  annotate("text", x=0.1, y=0.9, hjust = 0, vjust = 2.5, label=paste0("P= ",pval,"\nep num=",num),size=2.5) +
  labs(y="Enhancer activity",x="Interaction frequencies",title=paste0(tag,", enhancer activity vs. interaction frequencies")) + theme_classic() +
  theme(plot.title = element_text(size=6), axis.title=element_text(size=6), axis.text=element_text(size=5))



fmod <- glm(exp~act, data, family = "binomial") ##"full" mod
pval <- pchisq(summary(fmod)$null-summary(fmod)$deviance, summary(fmod)$df.null-summary(fmod)$df.residual, lower.tail=FALSE) %>% formatC(format = "g", digits = 3)
p2 <- data %>% ggplot(aes(x=exp, y=act#,color=tissue
                    )) +
  geom_point(alpha=.01,color="#5891f8ff",size=0.5) +
  stat_smooth(method="glm", se=T, color="black",
              fullrange=TRUE,  #view the full logistic distribution
              method.args = list(family=binomial)) +
  annotate("text", x=0.1, y=0.9, hjust = 0, vjust = 2.5, label=paste0("P= ",pval,"\nep num=",num),size=2.5) +
  labs(y="Enhancer activity",x="Expression level",title=paste0(tag,", enhancer activity vs. expression level")) + theme_classic() +
  theme(plot.title = element_text(size=6), axis.title=element_text(size=6), axis.text=element_text(size=5))

  
library(ggpmisc)
p3<- data %>% ggplot(aes(x=int, y=exp#,color=tissue
                    )) +
  geom_point(alpha=.1,color="#A60B00",size=0.5) +
  stat_smooth(method="lm",color="black", se=T) +
  stat_poly_eq(aes(label = paste(..rr.label..)),label.x.npc = "left", label.y.npc = 0.85, formula = y ~ x, parse = TRUE, size = 2) +
  stat_fit_glance(method = "lm", method.args = list(formula = y ~ x), geom = "text",aes(label = paste("P = ", formatC(..p.value..,format = "g", digits = 4), sep = "")),label.x = 0.1,size=2) +
  labs(y="Expression level",x="Interaction frequencies",title=paste0(tag,", interaction frequencie vs. expression level")) + theme_classic() + 
  theme(plot.title = element_text(size=6), axis.title=element_text(size=6), axis.text=element_text(size=5))

plot_grid(as.ggplot(figtemp1),p2,p3,nrow=1,rel_widths = c(3,1,1))

```

whether part of the brain is significantly higer then other part, only brain and also plot an example hs1172 chr13_78184481_78211082 FB

```{r}
tis3=c("FB","MB","HB")

targ="hs1172"
dplyr::filter(bmap,name==targ)

chrom="chr13"
gene_limit=c(78057768,78705499)

#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid

exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis3,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis3) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

tis_col=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TK","TL"),color=c("#629fd8ff","#716bbdff","#c881bbff","#c1d17dff","#c99516ff","#c75b7cff","#c83737ff","#cfcf3cff","#5b863dff","#73d758ff"))
enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>%  mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "any")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)

p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=1) +
  annotate(geom = "text",x=gene_limit[2]-500000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=1) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#order tissues
ord=c("text","anno","gene",paste0(rep(tis3,rep(2,length(tis3))),rep(c("-1",""),length(tis3)))) %>% rev()
supfig2c <- readcount_bait %>% 
  tidyr::replace_na(list(rescaled_Nav=0, score.y=0)) %>% 
  add_row(condition=c(paste0(tis3,rep("-1",length(tis3))),"gene","anno","text")) %>%
  mutate(condition=factor(condition, levels = ord)) %>%
  ggplot(aes(oeID_mid, condition, fill= rescaled_Nav, width = width)) + geom_tile() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = condition, xend = oeID_mid, yend = condition), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "tile", x = enh$baitID_mid, y = "gene",fill = "#448838",width=enh$wid,height=0.5) +
  annotate(geom = "text", x=enh$baitID_mid, y= "gene",label=enh$name,size = 2) +
  annotate(geom = "point", x = enh$baitID_mid, y = "anno", fill = "red2",color="red2", size = 1,shape=25) +
  annotate(geom = "text", x = enh$baitID_mid, y = "text", label="bait",size=3) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = "gene",forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= "gene",label=unique(gid$symbol),size = 2.5) +
  scale_fill_gradient(low="white", high="#552492",name="Normalized interaction frequencies") +
  scale_y_discrete(labels=c("FB-1"="","MB-1"="","HB-1"="","anno"="","text"="","gene"="")) +
  coord_cartesian(xlim=gene_limit) +
  theme_bw() + theme(axis.title=element_blank(),axis.ticks = element_blank(),
                     axis.text.x=element_blank(),panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                     axis.text.y=element_text(size=6),
                     plot.margin = margin(0, 0, 0, 0),
                     legend.position = "bottom",
                     legend.direction='horizontal',
                     legend.key.size = unit(2, 'mm'),
                     legend.title = element_text(size=6), 
                     legend.text = element_text(size=5))
supfig2c <- plot_grid(p0,supfig2c, ncol=1, align = "v", rel_heights = c(0.1,2))
supfig2c
```

```{r}
#get the bait enhancers that are active in at least one tissue in each brain tissues
tis3=c("FB","MB","HB")

j=0
for (ref in tis3) {
    j=j+1
    act_one <- bait_enh_act %>% dplyr::filter(tissue==ref & class == "POS") %>% .$elementID
    
    data <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>% #just keep ep pairs match in at least one tissue
      dplyr::filter((bait_type=="enh") & (bait_name %in% act_one) & (tissue %in% tis3)) %>% #only include the pos enh baits
      group_by(bait_name,promoter) %>% dplyr::filter(score==max(score)) %>% distinct(bait_name,promoter,.keep_all = T) %>% #filter this because one promoter link to multiple target fragment, and if target is not the same in different tissues, it would be some bias
      ungroup() %>% mutate(target_mid=(target_start+target_end)/2,start=target_mid-2500,end=target_mid+2500) %>% as.data.table() %>% foverlaps(., new_rmap, by.x=c("target_chr","start", "end"), type="any", nomatch=NULL) %>% ##find the reads around the target
      dplyr::select(-c(oeID_mid2,start,end,gene)) %>% complete(nesting(bait_name,promoter,target_chr, target_mid,otherEndID,oeID_mid), tissue=tis3) %>% ##complete the tissue combination
      left_join(dplyr::select(count_scaled,bait_name=name,otherEndID,scaled_Nav,tissue=condition), by=c("bait_name","otherEndID","tissue")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>% group_by(bait_name,promoter,tissue) %>% summarise(sum=sum(scaled_Nav)) %>% left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% mutate(value=if_else(is.na(class),0,1)) %>%
      dplyr::filter(!(tissue!=ref & value==1) | (tissue==ref & value==0)) %>% 
      add_count(bait_name,promoter) %>% dplyr::filter(n==3) %>%
      group_by(bait_name,promoter,value) %>% summarise(mean=mean(sum)) %>% ungroup() %>% mutate(value=if_else(value==1,"ref_tis","vs_ref"),group=ref) 
    assign(paste0("data",j),data)
}

#in order to show p.adjust to the plot, calculate p value separately 
data <- bind_rows(data1,data2,data3) %>% add_count(bait_name,promoter,group) %>% dplyr::filter(n==2) %>% mutate(group=factor(group,levels=c("FB","MB","HB")))
#data %>% distinct(bait_name,promoter,group) %>% count(group)
num=dplyr::count(data,group) %>% mutate(pair=n/2)

stat.test <- data %>% group_by(group) %>% wilcox_test(mean ~ value, paired = T) %>%
  adjust_pvalue(method = "BH") %>% add_significance("p.adj") %>% add_x_position(x = "group") %>%   mutate(p.adj=if_else(p.adj.signif=="ns","ns",formatC(p.adj, format = "g", digits = 3)))

#fold-change
supfig2d <- data %>% #dplyr::select(-class) %>% 
  pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% 
  ggplot(aes(y=group,x=log2fc)) + geom_boxplot(outlier.size = 0.5) + theme_bw() + labs(y="",x="E-P interaction frequency in\nactive / inactive tissues (log2) in brain") +
    annotate(geom = "text", y = stat.test$group, x = 2.4, label=paste0("n=",num$pair,"\np=", stat.test$p.adj) , size=2.5) +
  theme(axis.text.x = element_text(size=7),axis.title.x = element_text(size=8),axis.text.y = element_text(size=7),
        legend.position = c(0.6,0.75),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7), 
        legend.text = element_text(size=6),
        plot.title = element_text(size=7))

supfig2d
```

evolutionary DNA conservation score for enhancers interacting stably or not

```{r}
files <- list.files(path="~/enh_capC/phyloP", pattern=".bg$") 
rm(phyloP)
for (path in files){
  temp_dataset <- read_tsv(file=paste0("~/enh_capC/phyloP/",path),col_names = c("chr","start","end","score")) %>% add_column(bait_name=str_replace_all(path,".bg", ""))
  if (!exists("phyloP")) {
    phyloP <- temp_dataset 
  } else {
    phyloP <- bind_rows(phyloP,temp_dataset)
  }
}

supfig2h <- ep_exp_5kb %>% distinct(across(starts_with("bait")),tissue,promoter) %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>%
  dplyr::count(bait_name,promoter,bait_type,name="interact_tot") %>% group_by(bait_name,bait_type) %>% summarise(interact_tot=max(interact_tot)) %>% ungroup() %>% distinct() %>% left_join(phyloP, by="bait_name") %>% group_by(bait_name,interact_tot,bait_type) %>% summarise(mean_phyloP=mean(score)) %>%
  mutate(group=if_else(interact_tot<5,"tissue-specific",if_else(interact_tot==10,"invariant","no"))) %>% dplyr::filter(group!="no") %>%
  #only include tissue-specific (<=4) and invariant ones (=10)
  ggboxplot(y="mean_phyloP",x="group",outlier.size = 0.1) + scale_y_continuous(limits =c(0,4)) +
  theme_bw() + labs(y="Average phyloP score",x="") +
  stat_compare_means(label = "p.format", comparisons = list(c("tissue-specific","invariant")),method = "wilcox.test",paired = FALSE,tip.length = 0.03,label.y = 3.5,size=2) + # Add pairwise comparisons p-value
  theme(axis.text.x = element_text(size=7,angle = 30, vjust = 1, hjust = 1),axis.title.y = element_text(size=8),axis.text.y = element_text(size=7), axis.title.x=element_blank(), plot.title = element_text(size=7))
supfig2h
```

GO for stable interacting genes Biological Process (BP); Cellular
Component (CC); Molecular Function (MF)

```{r}
library(clusterProfiler)
library(org.Mm.eg.db)
library("ggpmisc")

gene <- ep_exp_5kb %>% left_join(tis_group,by="tissue") %>% distinct(bait_name,promoter,tis_group,bait_type,gene) %>% dplyr::count(bait_name,promoter,bait_type,gene,name="interact_tot") %>% arrange(desc(interact_tot)) %>% dplyr::filter(interact_tot==max(interact_tot) & bait_type=="enh") %>% .$gene %>% str_split(., ",",simplify = T) %>% as.character() %>% str_remove_all(.," ") %>% unique() 

anno <- AnnotationDbi::select(org.Mm.eg.db,keys=gene,columns=c("ENTREZID","ENSEMBL","GENENAME"),keytype="SYMBOL") %>% dplyr::filter(!is.na(ENSEMBL))
  
ego <- enrichGO(gene          = anno$ENTREZID,
                OrgDb         = org.Mm.eg.db,
                ont           = "MF",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
        readable      = TRUE)
df <- head(ego,n=20) %>% as_tibble() %>% separate(GeneRatio,into = c("on_gene","on_tot"),sep="/",remove = F) %>% separate(BgRatio,into = c("bg_gene","bg_tot"),sep="/") %>% mutate(across(on_gene:bg_tot, as.numeric)) %>% mutate(FoldEnrich=formatC( (on_gene/on_tot)/(bg_gene/bg_tot), format = "g", digits = 2), qvalue=formatC(qvalue, format = "g", digits = 2)) %>% 
  dplyr::select(ID,MolecularFunction=Description,GeneRatio,FoldEnrich,qvalue)

ego <- enrichGO(gene          = anno$ENTREZID,
                OrgDb         = org.Mm.eg.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
        readable      = TRUE)
df1 <- head(ego,n=5) %>% as_tibble() %>% separate(GeneRatio,into = c("on_gene","on_tot"),sep="/",remove = F) %>% separate(BgRatio,into = c("bg_gene","bg_tot"),sep="/") %>% mutate(across(on_gene:bg_tot, as.numeric)) %>% mutate(FoldEnrich=formatC( (on_gene/on_tot)/(bg_gene/bg_tot), format = "g", digits = 2), qvalue=formatC(qvalue, format = "g", digits = 2)) %>% 
  dplyr::select(ID,BiologicalProcess=Description,GeneRatio,FoldEnrich,qvalue)

supfig2i <- ggplot() + theme_void() +                                              # Add table to ggplot2 plot
  annotate(geom = "table", x = 0,  y = 0, label = list(df),size = 1.5) +
  annotate(geom = "table", x = 0,  y = 1, label = list(df1),size = 1.5) 
supfig2i
```

```{r}
supfig2de <- plot_grid(NULL,supfig2c,supfig2d,nrow=1, rel_widths = c(0.3,1.5,1),labels = c("","E","F"), label_size = 12)
supfig2bc <- plot_grid(supfig2b,supfig2j,nrow=1, rel_widths = c(5,1),labels = c("C","D"), label_size = 12)
supfig2ab <- plot_grid(supfig2a,supfig2k,nrow=1, rel_widths = c(2,1),labels = c("A","B"), label_size = 12)
supfig2 <- plot_grid(supfig2ab,supfig2bc,supfig2de,  ncol=1,rel_heights = c(1,1,0.7),scale = 0.9)
supfig2


supfig2ef <- plot_grid(supfig2e,supfig2f, labels = c("A", "B"), label_size = 12, nrow=1,rel_widths = c(0.9,1))
supfig2gh <- plot_grid(supfig2g,supfig2h, labels = c("C", "D"), label_size = 12, ncol=1,rel_heights = c(1,1), align = "v")
supfig2ghi <- plot_grid(supfig2gh,supfig2i, labels = c("", "E"), label_size = 12, nrow=1,rel_widths = c(0.3,1))
supfig3 <- plot_grid(supfig2ef,supfig2ghi, labels = c("", ""), label_size = 12, ncol=1,rel_heights = c(1,2))
supfig3

```

Fig3B. stable ep interaction example (hs699-EBf3 / hs2308-Dlx5|Dlx6 / ZRS-shh)

```{r}
tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")

targ="hs699"
dplyr::filter(bmap,name==targ)

chrom="chr7"
gene_limit=c(136400000,137400000)

#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

tis_col=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TK","TL"),color=c("#629fd8ff","#716bbdff","#c881bbff","#c1d17dff","#c99516ff","#c75b7cff","#c83737ff","#cfcf3cff","#5b863dff","#73d758ff"))
enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "within")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)

#scale bar
p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=3) +
  annotate(geom = "text",x=gene_limit[2]-500000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=3) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#ctcf get upper limit of 1 
p1 <- read_tsv("~/enh_capC/encode/bigwig/bg/hs699_GSM5501396_E12_brain_ChIP_CTCF.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>%  mutate(rc=readcount/sum(length)*length) %>% 
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>1,10,norm_rc/0.1)) %>%

  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#678cccff", colour="#678cccff", size=0.1) +
  labs(y = "WB CTCF") +
  scale_x_continuous(limits = gene_limit) +
  #scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p2 <- read_tsv("~/enh_capC/encode/bigwig/bg/hs699_GSM5501398_E12_PFL_ChIP_CTCF.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>%  mutate(rc=readcount/sum(length)*length) %>% 
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>1,10,norm_rc/0.1)) %>%

    ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#678cccff", colour="#678cccff", size=0.1) +
  labs(y = "FL CTCF") +
  scale_x_continuous(limits = gene_limit) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()
fig3b <- readcount_bait %>% #dplyr::filter(condition == "FB") %>%
  tidyr::replace_na(list(rescaled_Nav=0, score.y=0)) %>% #should fill NA or not?
  add_row(condition=c(paste0(tis,rep("-1",length(tis))),"anno","text","gene")) %>%
  mutate(condition=factor(condition, levels = ord)) %>%
  ggplot(aes(oeID_mid, condition, fill= rescaled_Nav, width = width)) + geom_tile() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = condition, xend = oeID_mid, yend = condition), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "tile", x = enh$baitID_mid, y = "gene",fill = "#448838",width=enh$wid,height=0.5) +
  annotate(geom = "text", x=enh$baitID_mid, y= "gene",label=enh$name,size = 2) +
  annotate(geom = "point", x = enh$baitID_mid, y = "anno", fill = "red2",color="red2", size = 1,shape=25) +
  annotate(geom = "text", x = enh$baitID_mid, y = "text", label="bait",size=2) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = "gene",forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=gid$mid, y= "gene",label=gid$symbol,size = 2) +
  scale_fill_gradient(low="white", high="#552492",name="normalized\nread count") +
  scale_y_discrete(labels=c("FB-1"="","MB-1"="","HB-1"="","CF-1"="","FL-1"="","HL-1"="","HR-1"="","NT-1"="","TL-1"="","TK-1"="","anno"="","text"="","gene"="")) +
  theme_bw() + theme(axis.title=element_blank(),axis.ticks = element_blank(),
                     axis.text.x=element_blank(),axis.text.y=element_text(size=6),
                     panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                     plot.margin = margin(0, 0, 0, 0),
                     legend.position = "none")
fig3b <- plot_grid(p0,fig3b,p1,p2, ncol=1, align = "v", rel_heights = c(0.1,2,0.15,0.15))
fig3b

```

dynamic ep interaction example (hs1431)
```{r}
tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")
targ="hs1431"
dplyr::filter(bmap,name==targ)

chrom="chr16"
gene_limit=c(14610000,15220000)
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

tis_col=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TK","TL"),color=c("#629fd8ff","#716bbdff","#c881bbff","#c1d17dff","#c99516ff","#c75b7cff","#c83737ff","#cfcf3cff","#5b863dff","#73d758ff"))
enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "within")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)


#scale bar
p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=3) +
  annotate(geom = "text",x=gene_limit[2]-500000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=3) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

p1 <- read_tsv("encode_data/hs1431_GSM5501396_E12_brain_ChIP_CTCF.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>1,10,norm_rc/0.1)) %>%

  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#678cccff", colour="#678cccff", size=0.1) +
  labs(y = "WB CTCF") +
  scale_x_continuous(limits = gene_limit) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p2 <- read_tsv("encode_data/hs1431_GSM5501398_E12_PFL_ChIP_CTCF.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>1,10,norm_rc/0.1)) %>%

    ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#678cccff", colour="#678cccff", size=0.1) +
  labs(y = "FL CTCF") +
  scale_x_continuous(limits = gene_limit) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()
fig3a <- readcount_bait %>% #dplyr::filter(condition == "FB") %>%
  tidyr::replace_na(list(rescaled_Nav=0, score.y=0)) %>% #should fill NA or not?
  add_row(condition=c(paste0(tis,rep("-1",length(tis))),"anno","text","gene")) %>%
  mutate(condition=factor(condition, levels = ord)) %>%
  ggplot(aes(oeID_mid, condition, fill= rescaled_Nav, width = width)) + geom_tile() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = condition, xend = oeID_mid, yend = condition), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "tile", x = enh$baitID_mid, y = "gene",fill = "#448838",width=enh$wid,height=0.5) +
  annotate(geom = "text", x=enh$baitID_mid, y= "gene",label=enh$name,size = 2) +
  annotate(geom = "point", x = enh$baitID_mid, y = "anno", fill = "red2",color="red2", size = 1,shape=25) +
  annotate(geom = "text", x = enh$baitID_mid, y = "text", label="bait",size=2) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = "gene",forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=gid$mid, y= "gene",label=gid$symbol,size = 2) +
  scale_fill_gradient(low="white", high="#552492",name="normalized\nread count") +
  scale_y_discrete(labels=c("FB-1"="","MB-1"="","HB-1"="","CF-1"="","FL-1"="","HL-1"="","HR-1"="","NT-1"="","TL-1"="","TK-1"="","anno"="","text"="","gene"="")) +
  theme_bw() + theme(axis.title=element_blank(),axis.ticks = element_blank(),
                     axis.text.x=element_blank(),axis.text.y=element_text(size=6),
                     panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                     plot.margin = margin(0, 0, 0, 0),
                     legend.position = "none")
fig3a <- plot_grid(p0,fig3a,p1,p2, ncol=1, align = "v", rel_heights = c(0.1,2,0.15,0.15))

#Fig3C. stable ep interaction example (shh)
tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")
targ="ZRS"

chrom="chr5"
gene_limit=c(28320000,29400000)
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100


tis_col=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TK","TL"),color=c("#629fd8ff","#716bbdff","#c881bbff","#c1d17dff","#c99516ff","#c75b7cff","#c83737ff","#cfcf3cff","#5b863dff","#73d758ff"))
enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "within")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)


#scale bar
p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=3) +
  annotate(geom = "text",x=gene_limit[2]-500000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=3) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

p1 <- read_tsv("encode_data/ZRS_GSM5501396_E12_brain_ChIP_CTCF.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>1,10,norm_rc/0.1)) %>%

  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#678cccff", colour="#678cccff", size=0.1) +
  labs(y = "WB CTCF") +
  scale_x_continuous(limits = gene_limit) +
  #scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p2 <- read_tsv("encode_data/ZRS_GSM5501398_E12_PFL_ChIP_CTCF.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>1,10,norm_rc/0.1)) %>%

    ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#678cccff", colour="#678cccff", size=0.1) +
  labs(y = "FL CTCF") +
  scale_x_continuous(limits = gene_limit) +
  #scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()
fig3c <- readcount_bait %>% #dplyr::filter(condition == "FB") %>%
  tidyr::replace_na(list(rescaled_Nav=0, score.y=0)) %>% #should fill NA or not?
  add_row(condition=c(paste0(tis,rep("-1",length(tis))),"anno","text","gene")) %>%
  mutate(condition=factor(condition, levels = ord)) %>%
  ggplot(aes(oeID_mid, condition, fill= rescaled_Nav, width = width)) + geom_tile() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = condition, xend = oeID_mid, yend = condition), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "tile", x = enh$baitID_mid, y = "gene",fill = "#448838",width=enh$wid,height=0.5) +
  annotate(geom = "text", x=enh$baitID_mid, y= "gene",label=enh$name,size = 2) +
  annotate(geom = "point", x = enh$baitID_mid, y = "anno", fill = "red2",color="red2", size = 1,shape=25) +
  annotate(geom = "text", x = enh$baitID_mid, y = "text", label="bait",size=2) +
  #annotate(geom = "tile", x=gid$mid, y= "gene",width=gid$wid,height=0.6,fill="#868686FF", color="#868686FF") +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = "gene",forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=gid$mid, y= "gene",label=gid$symbol,size = 2) +
  scale_fill_gradient(low="white", high="#552492",name="Normalized\ninteraction\nfrequencies") +
  scale_y_discrete(labels=c("FB-1"="","MB-1"="","HB-1"="","CF-1"="","FL-1"="","HL-1"="","HR-1"="","NT-1"="","TL-1"="","TK-1"="","anno"="","text"="","gene"="")) +
  theme_bw() + theme(axis.title=element_blank(),axis.ticks = element_blank(),
                     axis.text.x=element_blank(),axis.text.y=element_text(size=6),
                     panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                     plot.margin = margin(0, 0, 0, 0),
                     legend.position = c(0.2,0.95),
                     legend.direction='horizontal',
                     legend.key.size = unit(3, 'mm'),
                     legend.title = element_text(size=7), 
                     legend.text = element_text(size=6))
fig3c <- plot_grid(p0,fig3c,p1,p2, ncol=1, align = "v", rel_heights = c(0.1,2,0.15,0.15))

fig3abc <- plot_grid(fig3a,fig3b,fig3c,nrow=1,labels = c("A","B","C"),label_size = 12)
fig3abc
```

#hs1430 in this window (chr6-52299727-53270006)

```{r}
tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")
targ="hs1430"
dplyr::filter(bmap,name==targ)

chrom="chr6"
gene_limit=c(52200000,53200000)
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100


tis_col=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TK","TL"),color=c("#629fd8ff","#716bbdff","#c881bbff","#c1d17dff","#c99516ff","#c75b7cff","#c83737ff","#cfcf3cff","#5b863dff","#73d758ff"))
enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "within")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)


#scale bar
p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=3) +
  annotate(geom = "text",x=gene_limit[2]-500000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=3) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()
test <- readcount_bait %>% #dplyr::filter(condition == "FB") %>%
  tidyr::replace_na(list(rescaled_Nav=0, score.y=0)) %>% #should fill NA or not?
  add_row(condition=c(paste0(tis,rep("-1",length(tis))),"anno","text","gene")) %>%
  mutate(condition=factor(condition, levels = ord)) %>%
  ggplot(aes(oeID_mid, condition, fill= rescaled_Nav, width = width)) + geom_tile() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = condition, xend = oeID_mid, yend = condition), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "tile", x = enh$baitID_mid, y = "gene",fill = "#448838",width=enh$wid,height=0.5) +
  annotate(geom = "text", x=enh$baitID_mid, y= "gene",label=enh$name,size = 2) +
  annotate(geom = "point", x = enh$baitID_mid, y = "anno", fill = "red2",color="red2", size = 1,shape=25) +
  annotate(geom = "text", x = enh$baitID_mid, y = "text", label="bait",size=2) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = "gene",forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=gid$mid, y= "gene",label=gid$symbol,size = 2) +
  scale_fill_gradient(low="white", high="#552492",name="normalized\nread count") +
  scale_y_discrete(labels=c("FB-1"="","MB-1"="","HB-1"="","CF-1"="","FL-1"="","HL-1"="","HR-1"="","NT-1"="","TL-1"="","TK-1"="","anno"="","text"="","gene"="")) +
  theme_bw() + theme(axis.title=element_blank(),axis.ticks = element_blank(),
                     axis.text.x=element_blank(),axis.text.y=element_text(size=6),
                     panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                     plot.margin = margin(0, 0, 0, 0),
                     legend.position = "none")
           legend.text = element_text(size=6)
test <- plot_grid(p0,test, ncol=1, align = "v", rel_heights = c(0.1,2))
test

```

Fig3E. overlap chicago peaks (expressing ep pairs) with ensembl CTCF annotation, see if stable interactions

```{r}
invariant_emb_ctcf <- read_tsv("invariant_embryonic_CTCF.tsv",col_names = c("chr","start","end","num")) %>% mutate(type="CTCF") %>% as.data.table()
setkey(invariant_emb_ctcf,chr,start,end)

setDT(ep_exp_5kb)
#ctcf overlapping interactions for promoters in ep pairs
data <- ep_exp_5kb %>% unite("pair",c("bait_name","promoter"),remove=F) %>% dplyr::filter(pair %in% pos_pos_int$pair) %>%
  foverlaps(., invariant_emb_ctcf, by.x=c("target_chr","target_start", "target_end"), type="any") %>% group_by(bait_name,promoter,tissue,bait_type) %>% arrange(type) %>% dplyr::filter(row_number()==1) %>% ungroup() %>%
  distinct(bait_name,promoter,tissue,type,bait_type) %>% 
  left_join(bmap,by=c("bait_name"="name")) %>% mutate(start=start-5000,end=end+5000) %>% as.data.table() %>% foverlaps(., invariant_emb_ctcf, type="any") %>% group_by(bait_name,promoter,tissue) %>% arrange(type) %>% dplyr::filter(row_number()==1) %>% ungroup() %>% #filter the first ctcf site in each ep pair, either end (bait or oe) overlaps CTCF are counted
  add_count(bait_name,promoter,bait_type,name="No_tis") %>% distinct(bait_name,promoter,bait_type,.keep_all = T) %>%
  dplyr::count(i.type,type,No_tis,bait_type) %>% mutate(rowcount=if_else(is.na(i.type) & is.na(type),"None",if_else(is.na(i.type),"Bait", if_else(is.na(type),"Promoter","Both")))) %>% group_by(No_tis,bait_type) %>% mutate(perc=n/sum(n)*100,rowcount=factor(rowcount,levels = c("None","Promoter","Bait","Both"))) %>% ungroup() %>% dplyr::filter(bait_type=="enh")


fig3e <- data %>% ggplot(aes(x=No_tis,y=perc,fill=rowcount,group=bait_type)) + geom_bar(stat="identity") + #facet_wrap(~ bait_type, scales = "free") +
  scale_fill_manual(values = c("#868686FF","#EFC000FF","#0073C2FF","#CD534CFF"),name="CTCF peak") + theme_minimal() +
  theme(axis.title=element_text(size=8),axis.text=element_text(size=6),
        plot.title = element_text(size = 6), legend.position = "top",
        legend.key.size = unit(2, 'mm'),
        legend.title = element_text(size=6), 
        legend.text = element_text(size=6)) +
  labs(x="# of tissues", y = "% overlapping with ±5kb\naround CTCF peaks"#,title="Overlap with invariant embryonic CTCF peaks\nwith promoters (5kb) or bait enhancers (5kb distance)"
       )

fig3de <- plot_grid(fig3d,fig3e,nrow=1,labels = c("D","E"),label_size = 12,rel_widths = c(2,1))
fig3gh <- plot_grid(NULL,fig3h,nrow=1,labels = c("G","H"),label_size = 12,rel_widths = c(1,1))
fig3 <- plot_grid(fig3abc,fig3de,fig3gh,ncol=1,labels = c("","",""),rel_heights = c(1.3,1,1))

fig3
```



find the enhancers according to <Enhancer redundancy provides phenotypic robustness in mammalian development>

```{r}
#ee_CTCF_e10_12_5kb

### summarise the number of tissues they interact in (10 tissues)
p1 <- ee_CTCF_e10_12_5kb %>% unite(enhancer,starts_with("enhancer"),remove = F) %>% distinct(bait_name,enhancer,tissue,bait_type) %>% dplyr::count(bait_name,enhancer,bait_type, name="interact_tot") %>% dplyr::count(interact_tot,bait_type) %>% 
  dplyr::filter(bait_type=="enh") %>%
  group_by(bait_type) %>% mutate(perc=n/sum(n)) %>% #summarise(total=sum(n))
  ggplot(aes(x="",y=perc,fill=interact_tot)) + geom_bar(width = 1, stat = "identity", color = "white") + #facet_wrap(~ bait_type) +
  geom_text(aes(label=paste0(formatC(perc*100, format = "f", digits = 1),"%"), x = 1.2),position=position_stack(0.5),size=3) + coord_polar("y", start=0) + theme_void() + scale_fill_distiller(name="Tissue No.",direction = 1, breaks=c(2,4,6,8,10)) + theme(legend.direction='horizontal',legend.position = "bottom",
        legend.key.size = unit(3, 'mm'),
        legend.title = element_text(size=8),
        legend.text = element_text(size=6)) #+ labs(title="ee_pair_count_7_tissue")

### summarise the number of tissues they interact in (7 tissues)
p2 <- ee_CTCF_e10_12_5kb %>% unite(enhancer,starts_with("enhancer"),remove = F) %>% left_join(tis_group,by="tissue") %>% distinct(bait_name,enhancer,tis_group,bait_type) %>% dplyr::count(bait_name,enhancer,bait_type, name="interact_tot") %>% dplyr::count(interact_tot,bait_type) %>% 
  dplyr::filter(bait_type=="enh") %>%
  group_by(bait_type) %>% mutate(perc=n/sum(n)) %>% ggplot(aes(x="",y=perc,fill=interact_tot)) + geom_bar(width = 1, stat = "identity", color = "white") + #facet_wrap(~ bait_type) + 
  geom_text(aes(label=n, x = 1.2),position=position_stack(0.5),size=3) + coord_polar("y", start=0) + theme_void() + scale_fill_distiller(name="Tissue region No.",direction = 1) + theme(legend.direction='horizontal',legend.position = "bottom",
        legend.key.size = unit(3, 'mm'),
        legend.title = element_text(size=8),
        legend.text = element_text(size=6)) # + labs(title="ee_pair_count_4_tissue")
plot_grid(p1,p2,nrow=1,scale=0.9)

fig5c <- p1  
```

first filter out the other end that is with signal lower than 5
Fig5d. plot e-e interactions metaplot only signal \> 5

```{r}
#1.scale read count
#count_scaled

#2.get ee pairs, only signal > 5
#get activity of the other-end enhancers, because I include chicago peaks within 5kb away from enhancer peaks, so need to extend these peaks 5kb both sides
#only these tissues have ha3k27ac signal
tis_corr <- data.frame(tissue=c("embryonic_facial_prominence","forebrain","heart","hindbrain","limb","limb","liver","midbrain","neuraltube"),new=c("CF","FB","HR","HB","FL","HL","LV","MB","NT"))

ee_CTCF_exp <- ee_CTCF_e10_12 %>% unite(enhancer,starts_with("enhancer"),remove = F) %>% dplyr::filter(bait_type=="enh") %>% left_join(dplyr::filter(act_stage,stage=="E115"),by=c("enhancer_chr","enhancer_start","enhancer_end","tissue")) %>% arrange(bait_name,enhancer,tissue) %>% dplyr::select(enhancer,starts_with("target"),bait_name,tissue,score,signal,q_value,bait_type) %>% dplyr::filter(signal>5) %>% distinct()

#use all the h3k27ac around 5kb of chicago peaks
###filter out other end that lower than 5
ee_CTCF_5kb_exp <- ee_CTCF_e10_12_5kb %>% unite(enhancer,starts_with("enhancer"),remove = F) %>% as.data.table() %>% dplyr::filter(bait_type=="enh") %>% foverlaps(dplyr::filter(act_stage,stage=="E115"),type="any") %>% dplyr::filter(tissue==i.tissue) %>%
  arrange(bait_name,enhancer,tissue) %>% dplyr::select(enhancer,starts_with("target"),starts_with("bait"),tissue,score,signal,q_value,bait_type) %>% dplyr::filter(signal>5) %>% group_by(enhancer,bait_name,tissue,bait_type) %>% dplyr::filter(signal==max(signal)) %>% dplyr::filter(score==max(score)) %>% ungroup() %>% distinct()

```

```{r}
tis4=c("brain","cf","limb","hrt","nt","tail","trunk") 
tis_group=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TL","TK"),tis_group=c(rep("brain",3),"cf",rep("limb",2),"hrt","nt","tail","trunk"))
new_rmap <- rmap %>% mutate(oeID_mid2=oeID_mid) %>% as.data.table()
setkey(new_rmap,oe_chr,oeID_mid,oeID_mid2)

#3.plot metaplot
#get the read counts on 50kb around all ee pairs

rm_sd <- function(df, val, k) {
  df %>%
    mutate(
      rmean = rollapply(val, k, FUN = mean, partial=T, align="center"),
      rstd = rollapply(val, width = k, FUN = sd, partial=T, align="center")
    )  #add the prefix to calculate sd successfully
}

tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")

bound <- ee_CTCF_5kb_exp %>% 
  mutate(target_mid=as.integer((target_start+target_end)/2)) %>% left_join(rmap,by=c("target_chr"="oe_chr","target_mid"="oeID_mid")) %>%
  group_by(bait_name,enhancer) %>% mutate(oeIDmid=as.integer(mean(otherEndID)), strand=if_else(target_start>bait_start,"+","-")) %>% ungroup() %>% #get the peak mid on each enhancer and strandness
  distinct(bait_chr,bait_start,bait_end,bait_name,tissue,enhancer,oeIDmid,strand,bait_type) %>% left_join(rmap,by=c("bait_chr"="oe_chr","oeIDmid"="otherEndID")) %>%
  mutate(up=if_else(strand=="+", as.integer((bait_start+bait_end)/2)-50000, oeID_mid-50000), down=if_else(strand=="+",  oeID_mid+50000, as.integer((bait_start+bait_end)/2)+50000), bait_mid= as.integer((bait_start+bait_end)/2)) %>% as.data.table() %>% foverlaps(new_rmap, by.x=c("bait_chr","up", "down"), type="any") %>% arrange(bait_name,enhancer,tissue) %>% dplyr::select(bait_name,bait_mid,bait_type,enhancer,enh_mid=i.oeID_mid,otherEndID,oeID_mid,tissue,strand,up,down) %>% complete(nesting(bait_name,bait_mid,bait_type,enhancer,enh_mid,otherEndID,oeID_mid,strand,up,down),tissue=tis) %>%
  left_join(bait_enh_act,by=c("bait_name"="elementID","tissue")) %>% 
  left_join(select(count_scaled,name,otherEndID,scaled_Nav,condition),by=c("bait_name"="name","otherEndID","tissue"="condition")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>%
  group_by(bait_name,enhancer,tissue) %>% 
  mutate(class=if_else(is.na(class),"NEG","POS"), #complete the enhancer activity
         sum=sum(scaled_Nav)) %>% dplyr::filter(sum>0) %>% # filter out enh-enh pairs that have no read counts
  mutate(tenkb_Nav=if_else(oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000,0,scaled_Nav)) %>% mutate(max_dep=max(tenkb_Nav), rescaled_Nav=if_else( (scaled_Nav>max_dep | oeID_mid==bait_mid) ,100,scaled_Nav/max_dep*100)) %>% ungroup() %>% #set upper threshold for read count and rescale to 0-100
  dplyr::select(-c(otherEndID,tenkb_Nav,max_dep,up,down,sum)) %>% 
  group_by(bait_name,enhancer,tissue) %>% arrange(bait_name,enhancer,tissue,oeID_mid) %>%
  group_modify(~ { .x %>% rm_sd(.x$rescaled_Nav,3)}) %>% 
  mutate(newID=if_else(strand=="+",
                       if_else(oeID_mid<=bait_mid+10000, (oeID_mid-bait_mid-10000)/100,  #bait_mid+10000 is 0
                               if_else(oeID_mid>=enh_mid-10000,(oeID_mid-enh_mid+10000)/100+500, #enh_mid-10000 is 500
                                       (oeID_mid-bait_mid-10000)/(enh_mid-bait_mid-20000)*500)),  #scale in the middle
                       if_else(oeID_mid>=bait_mid-10000, (bait_mid-oeID_mid-10000)/100,    #otherwise
                               if_else(oeID_mid<=enh_mid+10000,(enh_mid-oeID_mid+10000)/100+500,
                                       (bait_mid-10000-oeID_mid)/(bait_mid-enh_mid-20000)*500)) 
                       )) %>% ungroup()   #give new coordinates

a_vs_in <- bound %>% distinct(bait_name,class,bait_type) %>% dplyr::filter(! is.na(class)) %>% dplyr::count(bait_name,bait_type) %>% dplyr::filter(bait_type=="enh" & n==2) 

num_int=bound %>% dplyr::filter(abs(enh_mid-bait_mid)>30000) %>% dplyr::filter(bait_name %in% a_vs_in$bait_name) %>% distinct(bait_name,enhancer) %>% nrow()
num_enh=bound %>% dplyr::filter(abs(enh_mid-bait_mid)>30000) %>% dplyr::filter(bait_name %in% a_vs_in$bait_name) %>% distinct(bait_name) %>% nrow()

fig5d <- bound %>% arrange(bait_name,enhancer,tissue,newID) %>% dplyr::filter(abs(enh_mid-bait_mid)>30000 & tissue !="TK") %>% dplyr::filter(bait_name %in% a_vs_in$bait_name) %>% group_by(bait_name,enhancer,bait_type,class,newID) %>%  summarise(m_Nav=mean(rescaled_Nav)) %>% ungroup() %>% pivot_wider(names_from = class, values_from = m_Nav) %>% mutate(ratio=POS/NEG, code="enh_invivo") %>% #dplyr::filter(newID>590 &newID<610 )
ggplot(aes(x=newID,y=ratio#,color=code
             )) + 
  stat_summary_bin(data=function(x) subset(x, newID < -110), geom="ribbon", binwidth= 35, fun.data=mean_cl_boot, fill="lightblue") +
  stat_summary_bin(data=function(x) subset(x, newID > -90), geom="ribbon", binwidth= 35, fun.data=mean_cl_boot, fill="lightblue") +
  stat_summary_bin(data=function(x) subset(x, newID < -110), geom="line", binwidth= 35, fun=mean, color="#0073C2FF",size=.3) +
  stat_summary_bin(data=function(x) subset(x, newID > -90), geom="line", binwidth= 35, fun=mean, color="#0073C2FF",size=.3) +
  geom_rect(data=data.frame(xmin = c(-130,-5,495), xmax = c(-70,5,505)), aes(xmin=xmin,xmax=xmax, ymin = -Inf, ymax = Inf),
            fill = "gray80",inherit.aes = F)  +
  scale_x_continuous(limits=c(-250, 750), #trans = squish_trans(1000, 9000,10), 
                     breaks=c(-200,-100,0,500,600,700),labels = c("bait-10kb", "bait", "bait+10kb", "enh-10kb", "enh", "enh+10kb")) + 
  labs(x="Position", y="Mean interaction frequency ratio\n(activ vs. inactive)", title = paste0(num_enh," enhancers with ",num_int, " interactions")) + theme_bw() +
  theme(axis.text.x = element_text(angle = 15, vjust = 1, hjust = 1,size=6),axis.title.x=element_blank(),axis.title.y=element_text(size=7),axis.text.y = element_text(size=6),
        legend.position = c(0.6,0.75),
        legend.key.size = unit(2.5, 'mm'),
        legend.title = element_text(size=7), 
        legend.text = element_text(size=6),
        plot.title = element_text(size=7)) 
fig5d
```

fig5e fold-change

```{r}
j=0
for (ref in tis4) {
    j=j+1
    act_one <- bait_enh_act_4group %>% dplyr::filter(tis_group==ref & value == 1) %>% .$elementID
    data <- ee_CTCF_5kb_exp %>% dplyr::filter((bait_type=="enh") & (bait_name %in% act_one)) %>%#only pos enh baits
            group_by(bait_name,enhancer) %>% arrange(bait_name,enhancer) %>% dplyr::filter(score==max(score)) %>% distinct(bait_name,enhancer,.keep_all = T) %>% 
    ungroup() %>% mutate(target_mid=(target_start+target_end)/2,start=target_mid-2500,end=target_mid+2500) %>% 
      
      as.data.table() %>% foverlaps(., new_rmap, by.x=c("target_chr","start", "end"), type="any", nomatch=NULL) %>% ##find the reads around the target
      dplyr::select(-c(oeID_mid2,start,end)) %>% complete(nesting(bait_name,enhancer,target_chr, target_mid,otherEndID,oeID_mid), tissue=tis) %>% ##complete the tissue combination
      left_join(dplyr::select(count_scaled,bait_name=name,otherEndID,scaled_Nav,tissue=condition), by=c("bait_name","otherEndID","tissue")) %>% tidyr::replace_na(list(scaled_Nav=0)) %>% left_join(tis_group,by="tissue") %>% group_by(bait_name,enhancer,tissue,tis_group) %>% summarise(sum=sum(scaled_Nav)) %>% left_join(bait_enh_act_4group,by=c("bait_name"="elementID","tis_group")) %>%
      dplyr::filter(!(tis_group!=ref & value==1) | (tis_group==ref & value==0)) %>%  group_by(bait_name,enhancer,value) %>% summarise(mean=mean(sum)) %>% ungroup() %>% 
      #dplyr::filter(!is.na(value)) %>% ###weird, some values are NA #check later
      mutate(value=if_else(value==1,"ref_tis","vs_ref"),group=ref)
    assign(paste0("data",j),data)
}

data <- bind_rows(data1,data2,data3,data4,data5,data6,data7) %>% add_count(bait_name,enhancer,group) %>% dplyr::filter(n==2) %>% dplyr::filter(group %in% c("brain","cf","limb")) %>% mutate(group=factor(group,levels=c("brain","cf","limb"))) 

num=dplyr::count(data,group) %>% mutate(pair=n/2) %>% dplyr::filter(group %in% c("brain","cf","limb"))
stat.test <- data %>% group_by(group) %>% wilcox_test(mean ~ value, paired = T) %>% #paired samples
  adjust_pvalue(method = "BH") %>% add_significance("p.adj") %>% add_x_position(x = "group") %>% #separate the p values to x.axis
  mutate(p.adj=formatC(p.adj, format = "g", digits = 2)) %>% dplyr::filter(group %in% c("brain","cf","limb"))

#read count box plot of ref and vs_ref, no tail h3k27ac data
p1 <- data %>% group_by(bait_name,enhancer,group) %>% mutate(remo=if_else(is.na(mean),1,0),n=sum(remo)) %>% dplyr::filter(n<1) %>% ungroup() %>% #filter out na pairs
  ggboxplot(x = "group", y = "mean", color = "value") + 
  stat_pvalue_manual(stat.test, label = "p.adj.signif",y.position=8.5, tip.length = 0.01) + theme_classic() +
  scale_color_manual(values = c("#0073C2FF", "#EFC000FF"), name = "Type", labels = c("Active", "Inactive")) + 
  scale_x_discrete(labels=c("Brain","Face","Limb","Heart","Neuraltube")) +
  scale_y_continuous(trans=pseudo_log_trans(base = 2),limits = c(0,500),breaks = c(0,1,4,16,64,256)) +
  labs(x="",y="Mean normalized read count on other end") +
  theme(legend.position = "top", axis.text.x = element_text(angle = 15, vjust = 1, hjust = 1))


#fold-change
p1 <- data %>%  group_by(bait_name,enhancer,group) %>% mutate(remo=if_else(is.na(mean),1,0),n=sum(remo)) %>% dplyr::filter(n<1) %>% dplyr::select(-c(remo,n)) %>% ungroup() %>% pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>%
  ggplot(aes(x=group,y=log2fc)) + geom_boxplot(outlier.size = 0.5) + theme_bw() + scale_x_discrete(labels=c("Brain","Face","Limb")) +     
  labs(y="",x="E-E interaction frequency in active / inactive tissues (log2)") +
    annotate(geom = "text", x = stat.test$group, y = 2.5, label=paste0("n=",num$pair,"\n", stat.test$p.adj.signif) , size=1.5) +   theme(legend.position="none",axis.title.x = element_text(size=7,angle=0, vjust = 0.5, hjust = 1),
        axis.title.y=element_text(size=7),axis.text = element_text(size=6)  ) 

#fold-change for other tissues
data <- bind_rows(data1,data2,data3,data4,data5,data6,data7) %>% add_count(bait_name,enhancer,group) %>% dplyr::filter(n==2) %>% dplyr::filter(group %in% c("nt","hrt","tail","trunk")) %>% mutate(group=factor(group,levels=c("nt","hrt","tail","trunk"))) 

num=dplyr::count(data,group) %>% mutate(pair=n/2) %>% dplyr::filter(group %in% c("nt","hrt","tail","trunk"))
stat.test <- data %>% group_by(group) %>% wilcox_test(mean ~ value, paired = T) %>% 
  adjust_pvalue(method = "BH") %>% add_significance("p.adj") %>% add_x_position(x = "group") %>% 
  mutate(p.adj=if_else(p.adj.signif=="ns","ns",formatC(p.adj, format = "g", digits = 3))) %>% dplyr::filter(group %in% c("nt","hrt","tail","trunk")) 

#fold-change
supfig6c <- data %>% group_by(bait_name,enhancer,group) %>% mutate(remo=if_else(is.na(mean),1,0),n=sum(remo)) %>% dplyr::filter(n<1) %>% dplyr::select(-c(remo,n)) %>% ungroup() %>% pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% 
  ggplot(aes(x=group,y=log2fc)) + geom_boxplot(outlier.size = 0.3) + theme_bw() + scale_x_discrete(labels=c("Neural tube","Heart","Tail","Trunk"))  + labs(y="E-E interaction frequency in\nactive / inactive tissues (log2)",x="") +
    annotate(geom = "text", x = stat.test$group, y = 2.5, label=paste0("n=",num$pair,"\np=", stat.test$p.adj) , size=2.5) + 
  theme(axis.title.y = element_text(size=8), axis.text=element_text(size=7))
supfig6c


#relationships between stars and fc
data <- bind_rows(data1,data2,data3,data4,data5,data6,data7) %>% add_count(bait_name,enhancer,group) %>% dplyr::filter(n==2) %>% mutate(group=factor(group,levels=c("brain","cf","limb","hrt","nt","tail","trunk"))) %>% left_join(stars,by=c("bait_name"="elementID","group"="tis_group")) %>% mutate(stars=factor(stars,levels=c(3,4,5)))

stat.test <- data %>% group_by(stars,group) %>% wilcox_test(mean ~ value, paired = T) %>% #paired samples
  adjust_pvalue(method = "BH") %>% add_significance("p.adj") %>% add_x_position(x = "group") %>%   mutate(p.adj=formatC(p.adj, format = "g", digits = 2)) #%>% dplyr::filter(group=="brain")
num=dplyr::count(data,group,stars) %>% mutate(pair=n/2)


p2 <- data %>% group_by(bait_name,enhancer,group) %>% 
  pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% ungroup() %>% ggplot(aes(x=stars,y=log2fc,fill=stars)) + geom_boxplot(outlier.size = 0.5) + theme_bw() + labs(x="Stars",y="Log2 fold-change") +
  facet_grid(~ group) +
  geom_text(aes(x=stars, y=4, label=p.adj.signif), data=stat.test, vjust=1) +
  theme_bw() + theme(legend.position="none") + scale_fill_manual(values=c("#D2E2FF", "#97AAFF","#7990FF","#5A77E3"))

p2 <- data %>% group_by(bait_name,enhancer,group) %>%
  dplyr::filter(group=="brain") %>% 
  pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% ungroup() %>%
  
  ggboxplot(y="log2fc",x="stars",fill="stars",outlier.size = 0.5) + 
  stat_compare_means(label = "p.format", comparisons = list(c("3","4"),c("4","5"),c("3","5")),method = "wilcox.test",paired = FALSE,tip.length = 0.01,label.y = c(2.2,2.4,2.6),size=1.5) +
  theme_bw() + labs(x="In vivo\nenhancer\nstrength\nin brain",y="E-E interaction frequency in active / inactive tissues (log2)") +
  theme(legend.position="none",axis.title.y = element_text(size=7,angle=0, vjust = 0.5, hjust = 1),
        axis.title.x=element_text(size=7),axis.text = element_text(size=6),
        ) + 
  scale_fill_manual(values=c("#D2E2FF", "#97AAFF","#7990FF","#5A77E3")) 

p2 <- ggpar(p2, orientation = "horiz")

#remove heart, tail and trunk due to less than 10
supfig6d <- data %>% dplyr::filter(! group %in% c("hrt","tail","trunk")) %>% 
  group_by(bait_name,enhancer,group) %>% 
  pivot_wider(names_from = value, values_from = mean) %>% mutate(log2fc=log2(ref_tis/vs_ref)) %>% ungroup() %>% 
  ggplot(aes(x=stars,y=log2fc,fill=stars)) + geom_boxplot(outlier.size = 0.5) + theme_bw() + labs(x="In vivo enhancer strength",y="E-E interaction frequency\nin active / inactive tissues (log2)") +
  facet_grid(~ group) +
  stat_compare_means(label = "p.format", comparisons = list(c("3","4"),c("4","5"),c("3","5")),method = "wilcox.test",paired = FALSE,tip.length = 0.01,label.y = c(2.4,2.7,3),size=3) + # Add pairwise comparisons p-value
  theme_bw() + theme(legend.position="none",axis.title = element_text(size=8)) + scale_fill_manual(values=c("#D2E2FF", "#97AAFF","#7990FF","#5A77E3"))
supfig6d

fig5e <- p1
fig5e
```

Fig5a mir9-2 locus

```{r}
targ=c("hs268","hs267","hs266","hs853")
dplyr::filter(bmap,name==targ)

chrom="chr13"
gene_limit=c(83558457,84861438)
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID
tis=c("FB","FL")
  
readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=sample(targ,size=nrow(.),replace=T),condition=sample(tis,size=nrow(.),replace=T)) %>% 
  complete(nesting(otherEndID,oe_chr,oeID_mid,name,width), condition=tis) %>%
  complete(nesting(otherEndID,oe_chr,oeID_mid,condition,width), name=targ) %>%
  left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% dplyr::select(midpoint=baitID_mid,everything()) %>% mutate( scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) 

max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>midpoint-10000 & oeID_mid<midpoint+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

tis_col=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TK","TL"),color=c("#629fd8ff","#716bbdff","#c881bbff","#c1d17dff","#c99516ff","#c75b7cff","#c83737ff","#cfcf3cff","#5b863dff","#73d758ff"))
enh = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2]) %>% mutate(wid=end-start)


#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "within")

#in UCSC they only use ENSMUST00000131907 to annotate C130071C03Rik, so change the coordinates for C130071C03Rik 
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% dplyr::filter((symbol=="C130071C03Rik" & tx_id=="ENSMUST00000131907") | symbol!="C130071C03Rik") %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% 
  dplyr::filter(gene_biotype %in% c("protein_coding","miRNA","lincRNA")) %>%
  summarise(tx_seq_start=min(tx_seq_start),tx_seq_end=max(tx_seq_end)) %>% mutate(wid=tx_seq_end-tx_seq_start,mid=(tx_seq_start+tx_seq_end)/2)

#scale bar
p0 <- data.frame(oeID_mid=rep(gene_limit[2]-50000,2), condition=c("scale_bar","text"), width=c(100000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-50000, y = "text", label="100kb",size=2.5) +
  annotate(geom = "text",x=gene_limit[2]-1000000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=3) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#h3k27ac get upper limit of 20 and scale to 10
p1 <- read_tsv("encode_data/mir9_forebrain_H3K27ac_e11_5.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>20,10,norm_rc/2)) %>%

  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#c99516ff", colour="#c99516ff", size=0.1) +
  scale_x_continuous(limits = gene_limit) +
  scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0)) + labs(y="FB H3K27ac")

p2 <- read_tsv("encode_data/mir9_limb_H3K27ac_e11_5.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>20,10,norm_rc/2)) %>%

  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#c99516ff", colour="#c99516ff", size=0.1) +
  scale_x_continuous(limits = gene_limit) +
  scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0)) + labs(y="LB H3K27ac")

#h3k4me3 get upper limit of 30 and scale to 10
p3 <- read_tsv("encode_data/mir9_forebrain_H3K4me3_e11_5.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>30,10,norm_rc/3)) %>%
  
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#599231ff", colour="#599231ff", size=0.1) +
  scale_x_continuous(limits = gene_limit) +
  scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0)) + labs(y="FB H3K4me3")

p4 <- read_tsv("encode_data/mir9_limb_H3K4me3_e11_5.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>30,10,norm_rc/3)) %>%
  
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#599231ff", colour="#599231ff", size=0.1) +
  scale_x_continuous(limits = gene_limit) +
  scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0)) + labs(y="LB H3K4me3")

#ctcf get upper limit of 1 
p5 <- read_tsv("encode_data/mir9_GSM5501396_E12_brain_ChIP_CTCF.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>1,10,norm_rc/0.1)) %>%

  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#678cccff", colour="#678cccff", size=0.1) +
  scale_x_continuous(limits = gene_limit) +
  scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0)) + labs(y="WB CTCF")

p6 <- read_tsv("encode_data/mir9_GSM5501398_E12_PFL_ChIP_CTCF.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  mutate(length=chromEnd-chromStart,gr=if_else(length>9,row_number(), as.integer(NA))) %>% mutate(gr=na.locf(gr,na.rm = FALSE,fromLast=T)) %>%
  group_by(gr,chrom) %>% partition(new_cluster(10)) %>% mutate(rc=readcount/sum(length)*length) %>% collect() %>%
  summarise(chromStart=min(chromStart),chromEnd=max(chromEnd),norm_rc=sum(rc),length=sum(length)) %>% ungroup() %>%
  mutate(norm_rc=if_else(norm_rc>1,10,norm_rc/0.1)) %>%

  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=norm_rc)) + geom_rect(fill="#678cccff", colour="#678cccff", size=0.1) +
  scale_x_continuous(limits = gene_limit) +
  scale_y_continuous(limits = c(0,10)) +
    theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0)) + labs(y="FL CTCF")

#order tissues
plot_tis=c("FB","FL")

#add curve, remove FL arches
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste(condition,name,"arch",sep="-"))

enh_anno <- dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2]) %>% mutate(seq_strand=".",wid=end-start+1) %>% arrange(start)

ord=c("text","anno","gene", paste(rep(plot_tis,rep((length(targ)*2),length(plot_tis))),rep(paste0(rep(targ,c(2,2,2,2)),c("-arch","")),length(plot_tis)),sep="-")) %>% rev() #%>% .[!grepl('FL.*arch',.)]

fig5a <- readcount_bait %>% tidyr::replace_na(list(rescaled_Nav=0, score.y=0)) %>% 
  dplyr::filter(condition %in% plot_tis) %>% unite("code",c("condition","name"),sep="-",remove = F) %>%
  add_row(code=c("anno","text","gene",distinct(df,condition)$condition)) %>%
  mutate(code=factor(code, levels = ord)) %>%
  ggplot(aes(oeID_mid, code, fill=rescaled_Nav,width = width)) + 
  geom_tile() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = condition, xend = oeID_mid, yend = condition), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1) })}} + 
  annotate(geom = "point", x = enh_anno$baitID_mid, y = "anno",fill = "#77b27bff",color="#77b27bff", size = 1,shape=25) +
  annotate(geom = "text", x=enh_anno$baitID_mid, y= "text",label=enh_anno$name,size = 2) +
  
  annotate(geom = "point", x = rep(enh$baitID_mid,2), y = rev(ord[grep("arch",ord)]), fill = "red2",color="red2", size = 1,shape=25) +
  geom_gene_arrow(data=gid, aes(xmin = tx_seq_start, xmax = tx_seq_end, y = "gene",forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=gid$mid, y= "gene",label=gid$symbol,size = 2) + 
  #scale_fill_manual(values =c("#39ad22","#7241d2","#e8412d","#d63fc9"),name="baits") +
  #scale_alpha_continuous(name="normalized\nread count",range=c(0,1)) +
  annotate(geom="tile",x=rep(enh$baitID_mid,2), y=paste(rep(plot_tis,c(4,4)),rep(enh$name,2),sep="-"),width=20000, fill=NA, color=rep(c("#629fd8ff", "#5b863dff","#c99516ff","#c83737ff"),2),size=0.8) +
  scale_fill_gradient(low="white", high="#552492",name="normalized\ninteraction\nfrequencies") +
  scale_y_discrete(labels=c("anno"="","text"="","gene"="")) +
  theme_bw() + theme(axis.title=element_blank(),axis.ticks = element_blank(),
                     axis.text.x=element_blank(),panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                     axis.text.y=element_text(size=7,angle = 0),
                     plot.margin = margin(0, 0, 0, 0),
                     legend.position = c(0.2,0.95),
                     legend.direction="horizontal",
                     legend.spacing=unit(0, 'mm'),
                     legend.key.size = unit(3, 'mm'),
                     legend.title = element_text(size=8), 
                     legend.text = element_text(size=6)) 

fig5a <- plot_grid(p0,fig5a,p1,p2,p3,p4,p5,p6, ncol=1, align = "v", rel_heights = c(0.1,2,0.2,0.2,0.2,0.2,0.2,0.2))

fig5a
```

other examples

```{r}
#first example
targ=c("mm1165","hs746","mm428","mm427")
dplyr::filter(bmap,name==targ)

chrom="chr5"
gene_limit=c(37554764,38206723)
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID
tis=c("CF","FB")

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=sample(targ,size=nrow(.),replace=T),condition=sample(tis,size=nrow(.),replace=T)) %>%
  complete(nesting(otherEndID,oe_chr,oeID_mid,name,width), condition=tis) %>%
  complete(nesting(otherEndID,oe_chr,oeID_mid,condition,width), name=targ) %>%
  left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% dplyr::select(midpoint=baitID_mid,everything()) %>% mutate( scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>midpoint-10000 & oeID_mid<midpoint+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

enh = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2]) %>% mutate(wid=end-start)

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "within")

gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% dplyr::filter((symbol=="C130071C03Rik" & tx_id=="ENSMUST00000131907") | symbol!="C130071C03Rik") %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% 
  dplyr::filter(gene_biotype %in% c("protein_coding","miRNA","lincRNA")) %>%
  #dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_seq_start=min(tx_seq_start),tx_seq_end=max(tx_seq_end)) %>% mutate(wid=tx_seq_end-tx_seq_start,mid=(tx_seq_start+tx_seq_end)/2)

plot_tis=c("CF","FB")

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste(condition,name,"arch",sep="-")) %>% dplyr::filter(!grepl("FB",condition))

enh_anno <- dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2]) %>% mutate(seq_strand=".",wid=end-start+1) %>% arrange(start)

ord=c("text","anno","gene", paste(rep(plot_tis,rep((length(targ)*2),length(plot_tis))),rep(paste0(rep(targ,c(2,2,2,2)),c("-arch","")),length(plot_tis)),sep="-")) %>% rev() %>% .[!grepl('FB.*arch',.)]

#scale bar
p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=2) +
  annotate(geom = "text",x=gene_limit[2]-500000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=2) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

supfig6a <- readcount_bait %>% tidyr::replace_na(list(rescaled_Nav=0, score.y=0)) %>% #should fill NA or not?
  dplyr::filter(condition %in% plot_tis) %>% unite("code",c("condition","name"),sep="-",remove = F) %>%
  add_row(code=c("anno","text","gene",distinct(df,condition)$condition)) %>%
  mutate(code=factor(code, levels = ord)) %>%
  ggplot(aes(oeID_mid, code, fill=rescaled_Nav,width = width)) + 
  geom_tile() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = condition, xend = oeID_mid, yend = condition), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1) })}} + 
  annotate(geom = "point", x = enh_anno$baitID_mid, y = "anno",fill = "#77b27bff",color="#77b27bff", size = 1,shape=15) +
  annotate(geom = "text", x=enh_anno$baitID_mid, y= "text",label=enh_anno$name,size = 2) +
  geom_gene_arrow(data=gid, aes(xmin = tx_seq_start, xmax = tx_seq_end, y = "gene",forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=gid$mid, y= "gene",label=gid$symbol,size = 2) + 
  scale_fill_gradient(low="white", high="#552492",name="normalized\ninteraction\nfrequencies") +
  scale_y_discrete(labels=c("anno"="","text"="","gene"="")) +
  theme_bw() + theme(axis.title=element_blank(),axis.ticks = element_blank(),
                     axis.text.x=element_blank(),panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                     axis.text.y=element_text(size=7,angle = 0),
                     plot.margin = margin(0, 0, 0, 0),
                     legend.position = c(0.2,0.95),
                     legend.direction="horizontal",
                     legend.spacing=unit(0, 'mm'),
                     legend.key.size = unit(3, 'mm'),
                     legend.title = element_text(size=8), 
                     legend.text = element_text(size=6)) 
supfig6a <- plot_grid(p0,supfig6a,ncol=1,rel_heights = c(1,10),align = "v")

#second example
targ=c("hs1315","mm1403")
dplyr::filter(bmap,name==targ)

chrom="chr13"
gene_limit=c(39098000,41000000)
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID
tis=c("NT","FL")

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=sample(targ,size=nrow(.),replace=T),condition=sample(tis,size=nrow(.),replace=T)) %>% 
  complete(nesting(otherEndID,oe_chr,oeID_mid,name,width), condition=tis) %>%
  complete(nesting(otherEndID,oe_chr,oeID_mid,condition,width), name=targ) %>%
  left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% dplyr::select(midpoint=baitID_mid,everything()) %>% mutate( scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>midpoint-10000 & oeID_mid<midpoint+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

enh = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2]) %>% mutate(wid=end-start)

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "within")

gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% dplyr::filter((symbol=="C130071C03Rik" & tx_id=="ENSMUST00000131907") | symbol!="C130071C03Rik") %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% 
  dplyr::filter(gene_biotype %in% c("protein_coding","miRNA","lincRNA")) %>%
  summarise(tx_seq_start=min(tx_seq_start),tx_seq_end=max(tx_seq_end)) %>% mutate(wid=tx_seq_end-tx_seq_start,mid=(tx_seq_start+tx_seq_end)/2)

plot_tis=c("NT","FL")

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste(condition,name,"arch",sep="-"))
enh_anno <- dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2]) %>% mutate(seq_strand=".",wid=end-start+1) %>% arrange(start)

ord=c("text","anno","gene", paste(rep(plot_tis,rep((length(targ)*2),length(plot_tis))),rep(paste0(rep(targ,c(2,2)),c("-arch","")),length(plot_tis)),sep="-")) %>% rev() #%>% .[!grepl('CF.*arch',.)]


#scale bar
p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]- 25000, y = "text", label="50kb",size=2) +
  annotate(geom = "text",x=gene_limit[2]-500000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=2) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

supfig6b <- readcount_bait %>% tidyr::replace_na(list(rescaled_Nav=0, score.y=0)) %>% 
  dplyr::filter(condition %in% plot_tis) %>% unite("code",c("condition","name"),sep="-",remove = F) %>%
  add_row(code=c("anno","text","gene",distinct(df,condition)$condition)) %>%
  mutate(code=factor(code, levels = ord)) %>%
  ggplot(aes(oeID_mid, code, fill=rescaled_Nav,width = width)) + 
  geom_tile() + 
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = condition, xend = oeID_mid, yend = condition), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1) })}} + 
  annotate(geom = "point", x = enh_anno$baitID_mid, y = "anno",fill = "#77b27bff",color="#77b27bff", size = 1,shape=15) +
  annotate(geom = "text", x=enh_anno$baitID_mid, y= "text",label=enh_anno$name,size = 2) +
  geom_gene_arrow(data=gid, aes(xmin = tx_seq_start, xmax = tx_seq_end, y = "gene",forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=gid$mid, y= "gene",label=gid$symbol,size = 2) + 
  scale_fill_gradient(low="white", high="#552492",name="normalized\ninteraction\nfrequencies") +
  scale_y_discrete(labels=c("anno"="","text"="","gene"="")) +
  theme_bw() + theme(axis.title=element_blank(),axis.ticks = element_blank(),
                     axis.text.x=element_blank(),panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                     axis.text.y=element_text(size=7,angle = 0),
                     plot.margin = margin(0, 0, 0, 0),
                     legend.position = c(0.2,0.95),
                     legend.direction="horizontal",
                     legend.spacing=unit(0, 'mm'),
                     legend.key.size = unit(3, 'mm'),
                     legend.title = element_text(size=8), 
                     legend.text = element_text(size=6)) 
supfig6b <- plot_grid(p0,supfig6b,ncol=1,rel_heights = c(1,5),align = "v")

supfig6ab <- plot_grid(supfig6a,supfig6b,rel_widths = c(1,1),ncol=2,labels = c("A","B"),label_size = 12)
supfig6cd <- plot_grid(NULL,supfig6c,rel_widths = c(1,1),nrow=1,labels = c("C","D"),label_size = 12)
supfig6 <- plot_grid(supfig6ab,supfig6cd,supfig6d,rel_heights = c(1,1,1.2),ncol=1,labels = c("","","E"),label_size = 12)
supfig6

```

```{r}
fig5ab <- plot_grid(as.ggplot(fig2e),fig2a,rel_widths = c(1.5,1),nrow=1,labels = c("A","B"),label_size = 12)
fig5cf <- plot_grid(fig2b,fig2cd,fig5d,fig5e,rel_widths = c(2,1),nrow=1,labels = c("C","D","E","F"),label_size = 12)
fig5 <- plot_grid(fig5ab,fig5cf,rel_heights = c(1.5,1),ncol=1)
ggsave("../manuscript/version_4_figure/fig5.pdf",width=183,height = 160,units="mm")
fig5

```

fig3h plot lz

```{r}
rm_sd <- function(df, k) {
  df %>%
    mutate(
      rmean = rollapply(readcount, k, FUN = mean, partial=T, align="center"),
      rstd = rollapply(readcount, width = k, FUN = sd, partial=T, align="center")
    )  #add the prefix to calculate sd successfully
}

targ="ZRS"
dplyr::filter(bmap,name==targ)

chrom="chr5"
gene_limit=c(28320000,29400000)

bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid
enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "within")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)


#scale bar
p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=2) +
  annotate(geom = "text",x=gene_limit[2]-500000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=2) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
  
#add curve
df <- read_tsv("~/enh_capC/enhancer_specificity/chicago3/LZ-FL.ibed") %>% dplyr::filter(bait_name=="lacZ_truncated") %>% mutate(midpoint=29315902, oeID_mid=(otherEnd_start+otherEnd_end)/2) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1))

p1 <- read_tsv("~/enh_capC/enhancer_specificity/LZ-FL/test/LZ-FL_chr5_29314500_29316005_ZRS.1kb.bg", col_names = c("chr","chromStart","chromEnd","readcount")) %>% dplyr::filter(chr==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  complete(chr, chromStart = full_seq(chromStart, period = 1000), fill = list(readcount = 0)) %>%
  mutate(chromEnd=chromStart+1000, oeID_mid=abs((chromStart+chromEnd)/2), readcount=replace(readcount,(oeID_mid >= bait_mid-3000 & oeID_mid <= bait_mid+5000),NA)) %>% 
  group_by(chr) %>% rm_sd(.,9) %>% ungroup() %>%
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20",size=0.3) +
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = 15, xend = oeID_mid, yend = 15), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1) })}} + 
  scale_x_continuous(limits = gene_limit) +
  coord_cartesian(ylim= c(0,22)) +
  geom_gene_arrow(data=gid, aes(xmin = gene_seq_start, xmax = gene_seq_end, y = 20,forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 20,label=unique(gid$symbol),size = 2) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0,vjust = 0.5),axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0)) + labs(y="WT")

#add curve
df <- read_tsv("~/enh_capC/enhancer_specificity/chicago3/LZ-FL.ibed") %>% dplyr::filter(bait_name=="ZRS") %>% mutate(midpoint=29315902, oeID_mid=(otherEnd_start+otherEnd_end)/2) %>% mutate(cur=if_else(midpoint>oeID_mid,0.1,-0.1))

p2 <- read_tsv("~/enh_capC/enhancer_specificity/LZ-FL/test/LZ-FL_lacZ_truncated_1_1946_lacZ_truncated.1kb.bg", col_names = c("chr","chromStart","chromEnd","readcount")) %>% dplyr::filter(chr==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  complete(chr, chromStart = full_seq(chromStart, period = 1000), fill = list(readcount = 0)) %>%
  mutate(chromEnd=chromStart+1000, oeID_mid=abs((chromStart+chromEnd)/2), readcount=replace(readcount,(oeID_mid >= bait_mid-3000 & oeID_mid <= bait_mid+5000),NA)) %>% 
  group_by(chr) %>% rm_sd(.,9) %>% ungroup() %>%
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20",size=0.3) +
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = 15, xend = oeID_mid, yend = 15), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1) })}} + 
  scale_x_continuous(limits = gene_limit) +
  coord_cartesian(ylim= c(22,0)) +
  geom_gene_arrow(data=gid, aes(xmin = gene_seq_start, xmax = gene_seq_end, y = 20,forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(2, "mm"), arrowhead_height = unit(2, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 20,label=unique(gid$symbol),size = 2) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=7,angle = 0,vjust = 0.5),axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0)) + labs(y="LZ")

fig3h <- plot_grid(p0,p1,p2,ncol=1,align="v",rel_heights = c(0.5,5,5))
fig3h
```

in R

```{r}
library(argparser)
library(Chicago)

rm(chinputs)
for(file in c("LZ-FB","LZ-FL","LZ-HL","LZ-LB1","LZ-LB2")) {
  temp_dataset <- read_tsv(file=paste0("~/enh_capC/enhancer_specificity/new_chinputs/",file,".chinput"),skip=1) %>% add_column(tissue=file)
  if (!exists("chinputs")) {
    chinputs <- temp_dataset 
  } else {
    chinputs <- bind_rows(chinputs,temp_dataset)
  }
}

#chr5    29313900        29317905        484133  ZRS
tis=c("FB","MB","HB","NT","CF","FL","HL","HR","TL","TK")

all_baits <- c(176796,176958,274559,357362,476464,484009,484133,484144,484178,622184,657758,736998)

targ=484144

chrom="chr5"
gene_limit=c(28311839,29450000)

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "any")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)


chinputs %>% group_by(tissue) %>% mutate(library_sum=sum(N)) %>% left_join(rmap,by="otherEndID") %>% dplyr::filter(baitID %in% c(484133,484144) & oe_chr==chrom & (! otherEndID %in% all_baits)) %>% dplyr::filter(oeID_mid<gene_limit[2] & oeID_mid>gene_limit[1]) %>% mutate(tenkb=if_else(oeID_mid>29315902-30000 & oeID_mid<29315902+30000, "yes","no")) %>% group_by(baitID,tenkb) %>% mutate(N=N/library_sum*10000) %>% #normalized by library size
  mutate(ten_max=max(N)) %>% group_by(baitID) %>% mutate(scal_fac=min(ten_max), scaled_N=if_else(N>=scal_fac, 100, N/scal_fac*100), name=if_else(baitID==484133,"ZRS","lacZ"), group=paste0(name,"_",tissue)) %>% ggplot(aes(oeID_mid, group, fill= scaled_N, width = otherEndLen)) + geom_tile() + 
    geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = "gene",forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(4, "mm"), arrowhead_height = unit(4, "mm"), arrowhead_width = unit(1, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= "gene",label=unique(gid$symbol),size = 2.5) +
  scale_fill_gradient(low="white", high="#552492",name="normalized read count") + theme_bw() + theme(axis.title=element_blank(),axis.ticks = element_blank(), axis.text.x=element_blank(),panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                     plot.margin = margin(0, 0, 0, 0))

chinputs %>% left_join(rmap,by="otherEndID") %>% mutate(library_sum=sum(N)) %>% dplyr::filter(baitID %in% c(484133,484144) & oe_chr==chrom & (! otherEndID %in% all_baits)) %>% dplyr::filter(oeID_mid<gene_limit[2] & oeID_mid>gene_limit[1]) %>% mutate(tenkb=if_else(oeID_mid>29315902-20000 & oeID_mid<29315902+20000, "yes","no")) %>% group_by(baitID,tenkb) %>%
  mutate(N=N/library_sum*10000) %>% #normalized by library size
  mutate(ten_max=max(N)) %>% group_by(baitID) %>% mutate(scal_fac=min(ten_max), scaled_N=if_else(N>=scal_fac, 100, N/scal_fac*100), name=if_else(baitID==484133,"ZRS","lacZ"), group=paste0(name,"_",tissue)) %>% dplyr::filter(oeID_mid>28456815-10000 & oeID_mid<28467256) %>% group_by(group,name,tissue) %>% summarise(total=sum(N)) %>% ggplot(aes(x=tissue,y=total,fill=name)) + geom_bar(stat="identity", position = "dodge")
```

RNA-seq data analysis

```{r}
library(DESeq2)
RNAseqPath="~/enh_capC/RNA-seq"
sampleFiles <- grep("Aligned.count.tsv",list.files(RNAseqPath),value=TRUE) %>% grep("MB",.,value=TRUE)
sampleCondition <- sub("EKZC_(.*)\\dAligned.count.tsv","\\1",sampleFiles)
sampleTable <- data.frame(sampleName = sampleFiles,
                          fileName = sampleFiles,
                          condition = sampleCondition)
sampleTable$condition <- factor(sampleTable$condition)

ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                       directory = RNAseqPath,
                                       design= ~ condition)
#filter out low expressed genes, below is to remove sum of read count from all samples has lower than 2
ddsHTSeq <- ddsHTSeq[(rowSums(counts(ddsHTSeq)) >= 20),]
#defined the comparison level
ddsHTSeq$condition <- factor(ddsHTSeq$condition, levels = c("MB_WT","MB_HS"))

ddsHTSeq <- DESeq(ddsHTSeq)
res <- results(ddsHTSeq)

plotMA(res) 
fig7b <- res %>% as.data.frame() %>% rownames_to_column("gene") %>% mutate(logpadj=-log10(padj), thres=if_else(padj<0.05,"padj<0.05","ns")) %>% ggplot(aes(x=log2FoldChange,y=logpadj, color=thres)) + geom_point(size=2) + geom_text_repel(data = ~subset(., logpadj > 0.4), aes(label=gene)) + theme_bw() + theme(legend.position = c(0.8,0.25)) +
  scale_color_manual(values=c("#868686FF", "#0073C2FF"), name = NULL, labels = c("Not Sig.", expression(p[adj]<0.05))) +
  labs(title="hs654 deletion", x="log2 fold change", y="-log10 adjusted p-value") #0.7 of wt
fig7b


sampleFiles <- grep("Aligned.count.tsv",list.files(RNAseqPath),value=TRUE) %>% grep("FB",.,value=TRUE)
sampleCondition <- sub("EKZC_(.*)\\dAligned.count.tsv","\\1",sampleFiles)
sampleTable <- data.frame(sampleName = sampleFiles,
                          fileName = sampleFiles,
                          condition = sampleCondition)
sampleTable$condition <- factor(sampleTable$condition)

ddsHTSeq <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                       directory = RNAseqPath,
                                       design= ~ condition)
#filter out low expressed genes, below is to remove sum of read count from all samples has lower than 2
ddsHTSeq <- ddsHTSeq[(rowSums(counts(ddsHTSeq)) >= 20),]
#defined the comparison level
ddsHTSeq$condition <- factor(ddsHTSeq$condition, levels = c("FB_WT","FB_HS","FB_MM"))

ddsHTSeq <- DESeq(ddsHTSeq)
res1 <- results(ddsHTSeq, contrast=c("condition","FB_HS","FB_WT"))
res2 <- results(ddsHTSeq, contrast=c("condition","FB_MM","FB_WT"))

plotMA(res1) #it can tell that after shrinkage is better
plotMA(res2) #it can tell that after shrinkage is better
fig7d <- res1 %>% as.data.frame() %>% rownames_to_column("gene") %>% mutate(logpadj=-log10(padj), thres=if_else(padj<0.05,"padj<0.05","ns")) %>% ggplot(aes(x=log2FoldChange,y=logpadj, color=thres)) + geom_point(size=2) + geom_text_repel(data = ~subset(., logpadj > 1.3 | gene %in% c("Tmem161b")), aes(label=gene),segment.color = '#cccccc',nudge_x=0.3*(-1)^seq(2,6),nudge_y=3) + theme_bw() + theme(legend.position = c(0.8,0.25)) + scale_color_manual(values=c("#868686FF", "#0073C2FF"), name = NULL, labels = c("Not Sig.", expression(p[adj]<0.05))) + labs(title="hs267-853 deletion", x="log2 fold change", y="-log10 adjusted p-value")  #0.4 of the wt

p2 <- res2 %>% as.data.frame() %>% rownames_to_column("gene") %>% mutate(logpadj=-log10(padj), thres=if_else(padj<0.05,"padj<0.05","ns")) %>% ggplot(aes(x=log2FoldChange,y=logpadj, color=thres)) + geom_point(size=2) + geom_text_repel(data = ~subset(., logpadj > 1.3 | gene=="Tmem161b"), aes(label=gene)) + theme_bw() + theme(legend.position = c(0.8,0.25)) + scale_color_manual(values=c("#868686FF", "#0073C2FF"), name = NULL, labels = c("Not Sig.", expression(p[adj]<0.05))) + labs(title="mm1501 deletion", x="log2 fold change", y="-log10 adjusted p-value") 
p2
fig7bd <- plot_grid(fig7b,fig7d,labels = c("B","D"), nrow=1, align="h")
ggsave("../manuscript/version_4_figure/fig7bd.pdf",width = 8.5, height = 4)
fig7bd

```

plot read count on chromosome in detail

```{r}
gene_limit=c(91300000,91500000)

grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1], gene_limit[2])), type = "any")
gid <- exons(edb,columns = c("gene_id","tx_id", "exon_idx"), filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% dplyr::select(gene_name,seq_strand,gene_biotype,gene_id,tx_id,exon_idx,exon_seq_start,exon_seq_end) %>% dplyr::filter(! gene_biotype %in% c("processed_pseudogene","TEC")) %>% group_by(tx_id) %>% mutate(min=min(exon_seq_start),max=max(exon_seq_end),len=max-min, mid=as.integer((min+max)/2)) %>% group_by(gene_name) %>% dplyr::filter(len==max(len))
intron <- gid %>% mutate(start=lag(exon_seq_end),end=exon_seq_start) %>% dplyr::select(start,end) %>% dplyr::filter(!is.na(start))

p1 <- readcount_bait %>% fill(midpoint,.direction = "downup") %>%   mutate(rescaled_Nav=if_else(is.na(rescaled_Nav),0,rescaled_Nav)) %>%
  group_by(condition,name) %>% rm_sd(.,5) %>% ungroup() %>%
  mutate(rmean=replace(rmean,oeID_mid==bait_mid,NA)) %>% #remove bait read count
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20") + theme_bw() + 
  scale_y_continuous(limits = c(0,max_dep$max + 25)) +
  geom_gene_arrow(data=dplyr::filter(gid,exon_idx==max(exon_idx)), aes(xmin = exon_seq_start, xmax = exon_seq_end, y = 1.2*max_dep$max,forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(4, "mm"), arrowhead_height = unit(4, "mm"), arrowhead_width = unit(1, "mm")) +
  geom_gene_arrow(data=dplyr::filter(gid,exon_idx!=max(exon_idx)), aes(xmin = exon_seq_start, xmax = exon_seq_end, y = 1.2*max_dep$max,forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(4, "mm"), arrowhead_height = unit(4, "mm"), arrowhead_width = unit(0, "mm")) +
    geom_gene_arrow(data=intron, aes(xmin = start, xmax = end, y = 1.2*max_dep$max), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(.1, "mm"), arrowhead_height = unit(.1, "mm"), arrowhead_width = unit(0, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 1.2*max_dep$max,label=unique(gid$gene_name),size = 2.5) +
  annotate(geom = "point", x = enh$baitID_mid, y = 1*max_dep$max, colour = "red", size = 3,shape=6) +
  annotate(geom = "text", x = enh$baitID_mid, y = 1.1*max_dep$max, label=enh$name,size=3) +
  labs(y = "MB smoothed\nread count") +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.title.y=element_text(size=8,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#plot
p2 <- read_tsv("~/enh_capC/RNA-seq/hs654_EKZC_MB_WT1Aligned.sorted.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  #mutate(readcount=if_else(readcount>20,20,readcount),norm_rc=readcount/2) %>%
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=readcount)) + geom_rect(fill="#c99516ff", colour="#c99516ff", size=0.1) +
  labs(y = "WT") +
  scale_x_continuous(limits = gene_limit) +
  scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=8,angle = 0, vjust = 0.5),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p3 <- read_tsv("~/enh_capC/RNA-seq/hs654_EKZC_MB_HS1Aligned.sorted.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  #mutate(readcount=if_else(readcount>20,20,readcount),norm_rc=readcount/2) %>%
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=readcount)) + geom_rect(fill="#c99516ff", colour="#c99516ff", size=0.1) +
  labs(y = "MT") +
  scale_x_continuous(limits = gene_limit) +
  scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=8,angle = 0, vjust = 0.5),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
plot_grid(p0, p1, p2,p3, ncol=1, align = "v", rel_heights = c(0.4,3,1,1))

```

mir9

```{r}
targ="hs268"

chrom="chr13"
gene_limit=c(83558457,84861438)
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

bait_mid=dplyr::filter(bmap,name %in% targ)$baitID_mid
enh = dplyr::filter(bmap,name %in% c("hs268","hs266","hs267","hs853")) %>% add_row(chr="chr13",start=84344904,end=84347924,baitID=NA,name="mm1501",baitID_mid=84346414)

tis="FB"
readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=sample(targ,size=nrow(.),replace=T),condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav) %>% distinct(name,condition,.keep_all = T)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

tis_col=data.frame(tissue=c("FB","MB","HB","CF","FL","HL","HR","NT","TK","TL"),color=c("#629fd8ff","#716bbdff","#c881bbff","#c1d17dff","#c99516ff","#c75b7cff","#c83737ff","#cfcf3cff","#5b863dff","#73d758ff"))

p0 <- data.frame(oeID_mid=rep(gene_limit[2]-50000,2), condition=c("scale_bar","text"), width=c(100000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.5) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-50000, y = "text", label="100kb",size=3) +
  #annotate(geom = "text",x=gene_limit[2]-1000000, y = "scale_bar", label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=4) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "any")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2) #%>% dplyr::filter(! symbol %in% c("Pou5f2","Mctp1"))


rm_sd <- function(df, k) {
  df %>%
    mutate(
      rmean = rollapply(rescaled_Nav, k, FUN = mean, partial=T, align="center"),
      rstd = rollapply(rescaled_Nav, width = k, FUN = sd, partial=T, align="center")
    )  
}

p1 <- readcount_bait %>% fill(midpoint,.direction = "downup") %>%   mutate(rescaled_Nav=if_else(is.na(rescaled_Nav),0,rescaled_Nav)) %>%
  group_by(condition,name) %>% rm_sd(.,5) %>% ungroup() %>%
  mutate(rmean=replace(rmean,oeID_mid %in% bait_mid,NA)) %>% #remove bait read count
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20") + theme_bw() + 
  scale_y_continuous(limits = c(0,max_dep$max + 25)) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = 1.2*max_dep$max,forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(4, "mm"), arrowhead_height = unit(4, "mm"), arrowhead_width = unit(1, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 1.2*max_dep$max,label=unique(gid$symbol),size = 2.5) +
  annotate(geom = "point", x = enh$baitID_mid, y = 1*max_dep$max, colour = "red", size = 3,shape=6) +
  annotate(geom = "text", x = enh$baitID_mid, y = 1.1*max_dep$max, label=enh$name,size=3) +

  labs(y = "FB smoothed\nread count") +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.title.y=element_text(size=8,angle = 0, vjust = 0.5),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))


#plot
p2 <- read_tsv("~/enh_capC/RNA-seq/mir9_EKZC_FB_WT1Aligned.sorted.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  #mutate(readcount=if_else(readcount>20,20,readcount),norm_rc=readcount/2) %>%
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=readcount)) + geom_rect(fill="#c99516ff", colour="#c99516ff", size=0.1) +
  labs(y = "WT") +
  scale_x_continuous(limits = gene_limit) +
  scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=8,angle = 0, vjust = 0.5),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p3 <- read_tsv("~/enh_capC/RNA-seq/mir9_EKZC_FB_HS1Aligned.sorted.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  #mutate(readcount=if_else(readcount>20,20,readcount),norm_rc=readcount/2) %>%
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=readcount)) + geom_rect(fill="#c99516ff", colour="#c99516ff", size=0.1) +
  labs(y = "MT") +
  scale_x_continuous(limits = gene_limit) +
  scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=8,angle = 0, vjust = 0.5),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
p4 <- read_tsv("~/enh_capC/RNA-seq/mir9_EKZC_FB_MM1Aligned.sorted.bg",col_names = c("chrom","chromStart","chromEnd","readcount")) %>% dplyr::filter(chrom==chrom, chromStart>=gene_limit[1], chromEnd<=gene_limit[2]) %>%
  ggplot(aes(xmin=chromStart, xmax=chromEnd, ymin=0, ymax=readcount)) + geom_rect(fill="#c99516ff", colour="#c99516ff", size=0.1) +
  labs(y = "mm1501") +
  scale_x_continuous(limits = gene_limit) +
  scale_y_continuous(limits = c(0,10)) +
  theme_bw() + theme(legend.position="none", axis.title.x=element_blank(),
                     axis.title.y=element_text(size=8,angle = 0, vjust = 0.5),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

plot_grid(p0, p1, p2,p3,p4, ncol=1, align = "v", rel_heights = c(0.4,3,1,1,1))
```

fig6 plot disease interactions

```{r}
#hs1523,  hs342

rm_sd <- function(df, k) {
  df %>%
    mutate(
      rmean = rollapply(rescaled_Nav, k, FUN = mean, partial=T, align="center"),
      rstd = rollapply(rescaled_Nav, width = k, FUN = sd, partial=T, align="center")
    )  #add the prefix to calculate sd successfully
}

tis="FB"

targ="hs342"

chrom="chr12"
gene_limit=c(49121092,50469462)


#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid

exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "any")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)

p0 <- data.frame(oeID_mid=rep(gene_limit[2]-50000,2), condition=c("scale_bar","text"), width=c(100000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-50000, y = "text", label="100kb",size=1) +
  annotate(geom = "text",x=gene_limit[2]-1000000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=1) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

p1 <- data.frame(xmin=29389815,	xmax=29391649, ymin = 0, ymax = 1) %>% ggplot(aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax)) + geom_rect(fill = "#448838", alpha=0.6) + 
  annotate(geom = "point", x = c(29390806,29390876), y = 1.5,fill = "#cb63a4ff",color = "#cb63a4ff", size = 1.5) +
  annotate("segment", x = c(29390806,29390876), xend = c(29390806,29390876), y = 0, yend = 1.5,colour = "#cb63a4ff") + 
  annotate("text", x = c(29390806,29390876), y=2, label=c("T>C","T>G"),size = 2) + 
  theme_void() + scale_y_continuous(limits = c(0,3)) + scale_x_continuous(limits = c(29388815,29392649),breaks = c(29390000,29390500,29391000,29391500)) + labs(x="human genome: chr14") +
  theme(axis.title.x=element_text(size=7), axis.text.x=element_text(size=6),
        axis.ticks.x= element_line(size = .5),axis.ticks.length=unit(0.1,"cm"))

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()

fig6b <- readcount_bait %>% fill(midpoint,.direction = "downup") %>%   mutate(rescaled_Nav=if_else(is.na(rescaled_Nav),0,rescaled_Nav)) %>%
  group_by(condition,name) %>% rm_sd(.,5) %>% ungroup() %>%
  mutate(rmean=replace(rmean,oeID_mid==bait_mid,NA)) %>% #remove bait read count
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20",size=0.3) + theme_bw() + 
  annotate(geom = "rect", xmin = 49381195, xmax = 49386204, ymin = -Inf, ymax = Inf, fill = "#629fd8ff") +
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = 78, xend = oeID_mid, yend = 78), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "point", x = enh$baitID_mid, y = 85,fill = "red2",color = "red2", size = 1,shape=25) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = 90, forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(3, "mm"), arrowhead_height = unit(3, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 90,label=unique(gid$symbol),size = 2) +
  annotate(geom = "tile", x = enh$baitID_mid, y = 90,fill = "#448838",width=enh$wid,height=3) +
  annotate(geom = "text", x=enh$baitID_mid+50000, y = 90,label="hs1523",size = 2) +
  scale_y_continuous(limits = c(0,100)) +

  labs(y = "Normalized\ninteraction\nfrequencies\nin forebrain") +
  coord_cartesian(xlim=gene_limit) + 
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

fig6b <- plot_grid(p1, p0, fig6b, ncol=1, align = "v", rel_heights = c(0.6,0.2,2))
fig6b

```

```{r}
tis="FL"

targ="hs1428"

chrom="chr3"
gene_limit=c(99000000,99900000)


#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid

exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "any")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)

p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=1) +
  annotate(geom = "text",x=gene_limit[2]-100000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=1) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

p1 <- data.frame(x=c(118263673,118785710), y= c(0,0)) %>% ggplot(aes(x = x, y = y)) + geom_line() + 
  annotate(geom = "point", x = 118486368, y = 0,fill = "#448838",color = "#448838", size = 1.5) + 
  theme_void() + scale_y_continuous(limits = c(-0.1,0.1)) + scale_x_continuous(limits = c(117263673,119785710)) + labs(x="human genome: chr1") +
  theme(axis.title.x=element_text(size=7), axis.text.x=element_text(size=6),
        axis.ticks.x= element_line(size = .5),axis.ticks.length=unit(0.1,"cm"))

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()

fig6a <- readcount_bait %>% fill(midpoint,.direction = "downup") %>% 
  mutate(rescaled_Nav=if_else(is.na(rescaled_Nav),0,rescaled_Nav)) %>%
  group_by(condition,name) %>% rm_sd(.,5) %>% ungroup() %>%
  mutate(rmean=replace(rmean,oeID_mid==bait_mid,NA)) %>% #remove bait read count
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20",size=0.3) + theme_bw() + 
  annotate(geom = "rect", xmin = 99251909, xmax = 99256153, ymin = -Inf, ymax = Inf, fill = "#629fd8ff") +
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = 78, xend = oeID_mid, yend = 78), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "point", x = enh$baitID_mid, y = 85,fill = "red2",color = "red2", size = 1,shape=25) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = 90, forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(3, "mm"), arrowhead_height = unit(3, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 90,label=unique(gid$symbol),size = 2) +
  annotate(geom = "tile", x = enh$baitID_mid, y = 90,fill = "#448838",width=enh$wid,height=3) +
  annotate(geom = "text", x=enh$baitID_mid+50000, y = 90,label=enh$name,size = 2) +
  scale_y_continuous(limits = c(0,100)) +
  annotate(geom = "rect", xmin = 99467760, xmax = 99838294, ymin = 95, ymax = 100, fill = "#cb63a4ff") +

  labs(y = "Normalized\ninteraction\nfrequencies\nin forelimb") +
  coord_cartesian(xlim=gene_limit) + 
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

fig6a <- plot_grid(p1,p0,fig6a, ncol=1, align = "v", rel_heights = c(0.6,0.2,2))

fig6 <- plot_grid(fig6a,fig6b, ncol=1, align = "v", rel_heights = c(1,1))
ggsave("../manuscript/version_4_figure/fig6.pdf",width=89,height = 90,units="mm")
fig6
```

other supplement examples

```{r}
#hs1507/mm1036/mm1042 limb 
tis="FL"
targ="hs1507"

chrom="chr1"
gene_limit=c(74788119,77634678)

#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>% #here, it doesn't matter that midpoint is NA, because midpoint is used to determine how to plot the curve, if midpoint is NA, that means there's no reads on it, so no interactions, no need to plot.
  left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "any")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)

p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=1) +
  annotate(geom = "text",x=gene_limit[2]-100000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=1) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()

supfig5a <- readcount_bait %>% fill(midpoint,.direction = "downup") %>% 
  mutate(rescaled_Nav=if_else(is.na(rescaled_Nav),0,rescaled_Nav)) %>%
  group_by(condition,name) %>% rm_sd(.,5) %>% ungroup() %>%
  mutate(rmean=replace(rmean,oeID_mid==bait_mid,NA)) %>% #remove bait read count
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20",size=0.3) + theme_bw() + 
  annotate(geom = "rect", xmin = 77509970, xmax = 77519037, ymin = -Inf, ymax = Inf, fill = "#629fd8ff") +
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = 78, xend = oeID_mid, yend = 78), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "point", x = enh$baitID_mid, y = 85,fill = "red2",color = "red2", size = 1,shape=25) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = 90, forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(3, "mm"), arrowhead_height = unit(3, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 90,label=unique(gid$symbol),size = 2) +
  annotate(geom = "tile", x = enh$baitID_mid, y = 90,fill = "#448838",width=enh$wid,height=3) +
  annotate(geom = "text", x=enh$baitID_mid+200000, y = 90,label=enh$name,size = 2) +
  scale_y_continuous(limits = c(0,100)) +
  annotate(geom = "rect", xmin = 74788119, xmax = 75902804, ymin = 95, ymax = 100, fill = "#cb63a4ff") +
  annotate(geom = "text", x = 75902804+1000000, y = 96,label="Polydactyly-associated duplication (P1)",size = 2, color = "#cb63a4ff") +

  labs(y = "Normalized\ninteraction\nfrequencies\nin forelimb") +
  coord_cartesian(xlim=gene_limit) +  
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
supfig5a <- plot_grid(p0,supfig5a, ncol=1, align = "v", rel_heights = c(0.2,2))
supfig5a
```

```{r}
#ZRS
tis="FL"
targ="ZRS"

chrom="chr5"
gene_limit=c(28320000,29400000)

#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "any")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)

p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=1) +
  annotate(geom = "text",x=gene_limit[2]-100000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=1) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

test <- data.frame(coord=c(105,287,334,396,401,402,404,406,407,417,446,463,515,555,619,621,739,743,769), 
name=c("C>G","C>A","T>G","C>T","A>G,C","C>T","G>T,C,A","A>G","T>A","A>G","T>A","T>G","C>A","G>A","C>T","C>G","A>G","T>G","T>C"))
p1 <- data.frame(xmin=1,	xmax=789, ymin = 0, ymax = 1) %>% ggplot(aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax)) + geom_rect(fill = "#448838", alpha=0.6) + annotate(geom = "point", x = test$coord, y = 1.5,fill = "#cb63a4ff",color = "#cb63a4ff", size = 1.5) +
  annotate("segment", x =  test$coord-1, xend = test$coord+1, y = 0, yend = 1.5,colour = "#cb63a4ff") + 
  annotate("text_repel", x = test$coord, y=2, label=test$name,size = 2,max.overlaps=20) + 
  theme_void() + scale_y_continuous(limits = c(0,3)) 

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()

supfig5c <- readcount_bait %>% fill(midpoint,.direction = "downup") %>% 
  mutate(rescaled_Nav=if_else(is.na(rescaled_Nav),0,rescaled_Nav)) %>%
  group_by(condition,name) %>% rm_sd(.,5) %>% ungroup() %>%
  mutate(rmean=replace(rmean,oeID_mid==bait_mid,NA)) %>% #remove bait read count
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20",size=0.3) + theme_bw() + 
  annotate(geom = "rect", xmin = 28465489, xmax = 28467972, ymin = -Inf, ymax = Inf, fill = "#629fd8ff") +
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = 78, xend = oeID_mid, yend = 78), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "point", x = enh$baitID_mid, y = 85,fill = "red2",color = "red2", size = 1,shape=25) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = 90, forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(3, "mm"), arrowhead_height = unit(3, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 90,label=unique(gid$symbol),size = 2) +
  annotate(geom = "tile", x = enh$baitID_mid, y = 90,fill = "#448838",width=enh$wid,height=3) +
  annotate(geom = "text", x=enh$baitID_mid+200000, y = 90,label=enh$name,size = 2) +
  scale_y_continuous(limits = c(0,100)) +

  labs(y = "Normalized\ninteraction\nfrequencies\nin forelimb") +
  coord_cartesian(xlim=gene_limit) +  
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
supfig5c <- plot_grid(p1,p0,supfig5c, ncol=1, align = "v", rel_heights = c(0.3,0.2,2))
supfig5c
```

```{r}
#hs1877
tis="CF"
targ="hs1877"

chrom="chr15"
gene_limit=c(61880003,63506895)

#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "any")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)

p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=1) +
  annotate(geom = "text",x=gene_limit[2]-100000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=1) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()

supfig5d <- readcount_bait %>% fill(midpoint,.direction = "downup") %>% 
  mutate(rescaled_Nav=if_else(is.na(rescaled_Nav),0,rescaled_Nav)) %>%
  group_by(condition,name) %>% rm_sd(.,5) %>% ungroup() %>%
  mutate(rmean=replace(rmean,oeID_mid==bait_mid,NA)) %>% #remove bait read count
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20",size=0.3) + theme_bw() + 
  annotate(geom = "rect", xmin = 61983835, xmax = 61987821, ymin = -Inf, ymax = Inf, fill = "#629fd8ff") +
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = 78, xend = oeID_mid, yend = 78), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "point", x = enh$baitID_mid, y = 87,fill = "red2",color = "red2", size = 1,shape=25) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = 90, forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(3, "mm"), arrowhead_height = unit(3, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 92,label=unique(gid$symbol),size = 2) +
  annotate(geom = "tile", x = enh$baitID_mid, y = 92,fill = "#448838",width=enh$wid,height=3) +
  annotate(geom = "text", x=enh$baitID_mid+200000, y = 90,label=enh$name,size = 2) +
  scale_y_continuous(limits = c(0,100)) +
  annotate(geom = "rect", xmin = 62844900, xmax = 63406895, ymin = 95, ymax = 100, fill = "#cb63a4ff") +
  annotate(geom = "text", x = 63406895-800000, y = 96,label="Region associated with cleft lip risk",size = 2, color = "#cb63a4ff") +
  
  labs(y = "Normalized\ninteraction\nfrequencies\nin craniofacial\nmesenchyme") +
  coord_cartesian(xlim=gene_limit) +  
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
supfig5d <- plot_grid(p0,supfig5d, ncol=1, align = "v", rel_heights = c(0.2,2))
supfig5d
```

```{r}
#hs737
tis="MB"
targ="hs737"

chrom="chr7"
gene_limit=c(136018204,137420338)

#remove reads around 10kb
bait_mid=dplyr::filter(bmap,name==targ)$baitID_mid
exid = dplyr::filter(bmap,chr==chrom,start>gene_limit[1],end <gene_limit[2])$baitID

readcount_bait <- read_tsv("~/enh_capC/track_hub/chicago_round3/mm10_DpnII_chicago3.rmap",col_names = F) %>% mutate(oeID_mid=as.integer((X2+X3)/2),width=X3-X2) %>% dplyr::select(otherEndID=X4,oe_chr=X1,oeID_mid,width) %>% dplyr::filter(oe_chr==chrom & oeID_mid>gene_limit[1] & oeID_mid<gene_limit[2]) %>% add_column(name=targ,condition=sample(tis,size=nrow(.),replace=T)) %>% complete(nesting(otherEndID,oe_chr,oeID_mid,name,width),condition=tis) %>% left_join(count_scaled,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>% left_join(ditag_peaks,by=c("name","otherEndID","oe_chr","oeID_mid","condition")) %>%
  arrange(condition,otherEndID) %>% mutate(midpoint=(start+end)/2, scaled_Nav=replace(scaled_Nav,(otherEndID %in% exid),NA)) #remove bait-to-bait reads
  
max_dep <- readcount_bait %>% mutate(scaled_Nav=replace(scaled_Nav, (oeID_mid>bait_mid-10000 & oeID_mid<bait_mid+10000), NA)) %>% #remove reads around 10kb of the bait_mid
  group_by(name,condition) %>% dplyr::filter(scaled_Nav==max(scaled_Nav,na.rm=T)) %>% ungroup() %>% dplyr::select(name,condition,max=scaled_Nav)
  
readcount_bait <- readcount_bait %>% dplyr::select(name,otherEndID,oe_chr,oeID_mid,condition,scaled_Nav,score=score.y,midpoint,width) %>%   left_join(max_dep,by=c("name","condition")) %>% mutate(scaled_Nav=if_else( (!is.na(scaled_Nav) & scaled_Nav>max), max,scaled_Nav)) %>% group_by(name,condition) %>% mutate(rescaled_Nav=(scaled_Nav-min(scaled_Nav,na.rm=T))/(max(scaled_Nav,na.rm=T)-min(scaled_Nav,na.rm=T))*100) %>% ungroup()  #rescale them to 1-100

enh = dplyr::filter(bmap,name==targ) %>% mutate(wid=end-start)

#add curve
df <- dplyr::filter(readcount_bait,score>=5) %>% left_join(tis_col,by=c("condition"="tissue")) %>% mutate(cur=if_else(midpoint>oeID_mid,-0.1,0.1),condition=paste0(condition,"-1"))

#add gene
edb <- EnsDb.Mmusculus.v79
num=str_remove(chrom, "chr")
grf <- GRangesFilter(GRanges(num, ranges = IRanges(gene_limit[1]-1e4, gene_limit[2]+1e4)), type = "any")
gid <- genes(edb, filter = grf,return.type="DataFrame") %>% as_tibble() %>% left_join(transcripts(edb, filter = grf,return.type="DataFrame") %>% as_tibble(), by=c("gene_id","seq_name","seq_strand")) %>% group_by(symbol,gene_seq_start,gene_seq_end,gene_id,seq_strand) %>% dplyr::filter(!is.na(tx_cds_seq_start)) %>% 
  summarise(tx_cds_seq_start=min(tx_cds_seq_start),tx_cds_seq_end=max(tx_cds_seq_end)) %>% mutate(wid=tx_cds_seq_end-tx_cds_seq_start,mid=(tx_cds_seq_start+tx_cds_seq_end)/2)

p0 <- data.frame(oeID_mid=rep(gene_limit[2]-25000,2), condition=c("scale_bar","text"), width=c(50000,0)) %>%
  ggplot(aes(oeID_mid, condition, width=width)) + geom_tile(fill="black",color="white",height=0.6) +  theme_void() + 
  annotate(geom = "text",x=gene_limit[2]-25000, y = "text", label="50kb",size=1) +
  annotate(geom = "text",x=gene_limit[2]-100000, y = "scale_bar",
           label=paste0(chrom,":",paste(gene_limit[1],gene_limit[2],sep="-")), size=1) +
  scale_x_continuous(limits = gene_limit, position = "top") +
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.text=element_blank(),
        axis.ticks=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))


test <- data.frame(coord=c(128568698,128569418,128569437), name=c("A>G","T>C","G>A"))
p1 <- data.frame(xmin=128568604,	xmax=128569741, ymin = 0, ymax = 1) %>% ggplot(aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax)) + geom_rect(fill = "#448838", alpha=0.6) + annotate(geom = "point", x = test$coord, y = 1.5,fill = "#cb63a4ff",color = "#cb63a4ff", size = 1.5) +
  annotate("segment", x = test$coord-1, xend = test$coord+1, y = 0, yend = 1.5,colour = "#cb63a4ff") + 
  annotate("text_repel", x = test$coord, y=2, label=test$name, size = 2) + 
  theme_void() + scale_y_continuous(limits = c(0,3)) 

#order tissues
ord=c("text","anno","gene",paste0(rep(tis,rep(2,length(tis))),rep(c("-1",""),length(tis)))) %>% rev()

supfig5e <- readcount_bait %>% fill(midpoint,.direction = "downup") %>% 
  mutate(rescaled_Nav=if_else(is.na(rescaled_Nav),0,rescaled_Nav)) %>%
  group_by(condition,name) %>% rm_sd(.,5) %>% ungroup() %>%
  mutate(rmean=replace(rmean,oeID_mid==bait_mid,NA)) %>% #remove bait read count
  ggplot(aes(oeID_mid, rmean)) + geom_line(color="gray20",size=0.3) + theme_bw() + 
  annotate(geom = "rect", xmin = 137312569, xmax = 137316184, ymin = -Inf, ymax = Inf, fill = "#629fd8ff") +
  {if(dim(df)[1]!=0) {lapply(split(df, 1:nrow(df)), function(dat) {    #arches on top
      geom_curve(data = dat, aes(x = midpoint, y = 78, xend = oeID_mid, yend = 78), 
                 curvature = -dat["cur"]*1, color="gray25",size=0.1 ) })}} + 
  annotate(geom = "point", x = enh$baitID_mid, y = 87,fill = "red2",color = "red2", size = 1,shape=25) +
  geom_gene_arrow(data=gid, aes(xmin = tx_cds_seq_start, xmax = tx_cds_seq_end, y = 90, forward =seq_strand), fill = "#868686FF",inherit.aes = F, arrow_body_height = unit(3, "mm"), arrowhead_height = unit(3, "mm"), arrowhead_width = unit(0.5, "mm")) +
  annotate(geom = "text_repel", x=unique(gid$mid), y= 92,label=unique(gid$symbol),size = 2) +
  annotate(geom = "tile", x = enh$baitID_mid, y = 92,fill = "#448838",width=enh$wid,height=3) +
  annotate(geom = "text", x=enh$baitID_mid+200000, y = 90,label=enh$name,size = 2) +
  scale_y_continuous(limits = c(0,100)) +
  annotate(geom = "rect", xmin = 62844900, xmax = 63406895, ymin = 95, ymax = 100, fill = "#cb63a4ff") +
  annotate(geom = "text", x = 63406895-800000, y = 96,label="Region associated with cleft lip risk",size = 2, color = "#cb63a4ff") +
  labs(y = "Normalized\ninteraction\nfrequencies\nin midbrain") +
  coord_cartesian(xlim=gene_limit) +  
  theme(legend.position="none", axis.title.x=element_blank(),
        axis.title.y=element_text(size=6,angle = 0, vjust = 0.5),
        axis.text.y=element_text(size=6),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0))
supfig5e <- plot_grid(p1,p0,supfig5e, ncol=1, align = "v", rel_heights = c(0.3,0.2,2))
supfig5e

```

```{r}
supfig5ab <- plot_grid(supfig5a,supfig5c,labels=c("A","B"),label_size = 12,nrow=1)
supfig5cde <- plot_grid(supfig5d,supfig5e,labels=c("C","D"),label_size = 12,nrow=1)
supfig5 <- plot_grid(supfig5ab,supfig5cde,ncol=1)
ggsave("../manuscript/version_4_figure/Supp_fig5.pdf",width=183,height = 120,units="mm")
supfig5
```

3D DNA-FISH analysis

zic1/4 and hs654 are ~600kb apart
```{r}
library(rstatix)

FL1 <- read_csv("FL-Zic-Cy3-hs654-488-2-27-23-9.csv",skip=4,col_names = c("dist","unit")) %>% mutate(tissue="FL") 
MB1 <- read_csv("MB-Zic-Cy3-hs654-488-2-27-23-9.csv",skip=4,col_names = c("dist","unit")) %>% mutate(tissue="MB") 

FL2 <- read_csv("FL-Zic-Cy3-hs654-488-2-27-23-4.csv",skip=4,col_names = c("dist","unit")) %>% mutate(tissue="FL") 
MB2 <- read_csv("MB-Zic-Cy3-hs654-488-2-27-23-4.csv",skip=4,col_names = c("dist","unit")) %>% mutate(tissue="MB") 

FL <- bind_rows(FL1,FL2)
MB <- bind_rows(MB1,MB2)

num=bind_rows(FL,MB) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("MB","FL"))) %>% count(tissue) #count dot number
stat.test <- bind_rows(FL,MB) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("MB","FL"))) %>% wilcox_test(dist ~ tissue, paired = F)
col_frac=bind_rows(FL,MB) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("MB","FL")),col=if_else(dist<0.25,"yes","no")) %>% dplyr::count(col,tissue) %>% group_by(tissue) %>% mutate(frac=formatC(n/sum(n)*100, format = "g", digits = 2)) %>% ungroup() %>% dplyr::filter(col=="yes")

fig4g_a <- bind_rows(FL,MB) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("MB","FL"))) %>%  ggviolin(x="tissue",y="dist",fill="tissue", palette=c("#68abdaff","grey"), trim=T) +
  geom_boxplot(fill="white", outlier.shape = NA,width=0.2) +
  stat_pvalue_manual(stat.test, label = "p",y.position=1.55,size=2)  + 
  annotate(geom = "text", x = col_frac$tissue, y = 0.15, label=paste0(col_frac$frac,"%"), size=2) + 
  scale_x_discrete(labels=c(paste0(num$tissue[1],"\n(n=",num$n[1],")"), paste0(num$tissue[2],"\n(n=",num$n[2],")"))) + 
  geom_hline(yintercept=0.25, color="red",linetype=2) + 
  labs(x="", y="inter-probe distance (µm)",title = "Zic1/4 and hs654") +  
  theme(legend.position = "None", 
        axis.title.x = element_text(size=7),
        axis.title.y=element_text(size=7),axis.text = element_text(size=6),
        plot.title = element_text(size=8))
fig4g_a

#pie chart
bind_rows(FL,MB) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("MB","FL"))) %>% group_by(gr=cut(dist, breaks= seq(0, 1.6, by = 0.2)) ) %>% count(tissue,gr) %>% group_by(tissue) %>% mutate(perc=n/sum(n)) %>% ungroup() %>% ggplot(aes(x = "", y = perc, fill = gr)) + 
  geom_bar(width = 1, stat = "identity", color = "white") + facet_wrap(~ tissue) +
  coord_polar("y", start = 0) +
  geom_text(aes(label=n, x = 1.3), #change the in or out of the label
    position=position_stack(0.5), #align the label in the middle of each pie
    size=2.5) +
  theme_void()

#ECDF (or Empirical cumulative distribution function) plot
fig4g_supa1 <- bind_rows(FL,MB) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("MB","FL"))) %>% 
  ggplot(aes(x = dist)) + stat_ecdf(aes(color = tissue),  geom = "step", size = 0.5) +
  scale_color_manual(values = c("#68abdaff","grey")) +
  labs(y = "cumulative frequency", x="E-P interprobe distance (µm)",title = "Zic1/4 and hs654") +
  theme(legend.key.size = unit(1, 'mm'),legend.title = element_text(size=7), 
        legend.text = element_text(size=6),legend.position = c(0.85,0.6),
        axis.title.x = element_text(size=7),
        axis.title.y=element_text(size=7),axis.text = element_text(size=6),
        plot.title = element_text(size=7))

fig4g_supa1

fig4g_supa <- bind_rows(FL,MB) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("MB","FL"))) %>% count(tissue,gr=cut(dist, breaks= seq(0, 1.6, by = 0.2))) %>% ggplot(aes(x = gr, y=n, fill=tissue)) + geom_bar(stat="identity", position = "dodge") + scale_fill_manual(values = c("#68abdaff","grey")) + labs(x="E-P interprobe distance (µm)",title = "Zic1/4 and hs654")


#histogram
data <- bind_rows(FL,MB) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("MB","FL"))) %>% count(tissue,gr=cut(dist, breaks= c(seq(0, 1, by = 0.25),1.5)))
  
stat.test <- data %>% group_by(tissue) %>% mutate(non=sum(n)-n) %>% unite("code",c("gr","tissue"),sep=":") %>% pivot_longer(!code, names_to = "cat", values_to = "count") %>% pivot_wider(names_from = "code", values_from = "count") %>% column_to_rownames("cat") %>% pairwise_prop_test(p.adjust.method = "fdr") %>% separate(group1, c("A", "B"),remove = F,sep=":") %>% separate(group2, c("C", "D"),remove = F,sep=":") %>% dplyr::filter(A==C) %>% arrange(A) %>% mutate(p.adjsig=if_else(p.adj<0.05,as.character(p.adj),p.adj.signif)) %>% mutate(xmin=row_number()-0.23, xmax=row_number()+0.23) 

fig4g_supa2 <- data %>% group_by(tissue) %>% mutate(frac=n/sum(n)) %>% ggbarplot(x = "gr", y="frac", fill="tissue",position = position_dodge(0.9)) + scale_fill_manual(values = c("#68abdaff","grey")) + labs(x="E-P interprobe distance (µm)",y="Percent of colocalization") + stat_pvalue_manual(stat.test, label = "p.adjsig",y.position=0.45,tip.length = 0.02,size=2)  +
  scale_y_continuous(labels = scales::percent) +
  theme(legend.key.size = unit(1, 'mm'),legend.title = element_text(size=7), 
        legend.text = element_text(size=6),legend.position = c(0.85,0.6),
        axis.title.x = element_text(size=7),
        axis.title.y=element_text(size=7),axis.text = element_text(size=6),
        plot.title = element_text(size=7))
fig4g_supa2

```


mir9-2 and hs266 are ~900kb apart
```{r}
Mir_FB1 <- read_csv("FB-Mir9-Cy3-hs266-488-2-27-23-4.csv",skip=4,col_names = c("dist","unit")) %>% mutate(tissue="FB") 
Mir_FL1 <- read_csv("FL-Mir9-Cy3-hs266-488-2-27-23-4.csv",skip=4,col_names = c("dist","unit")) %>% mutate(tissue="FL") 

Mir_FB2 <- read_csv("FB-Mir9-Cy3-hs266-488-2-27-23-1.csv",skip=4,col_names = c("dist","unit")) %>% mutate(tissue="FB") 
Mir_FL2 <- read_csv("FL-Mir9-Cy3-hs266-488-2-27-23-1",skip=4,col_names = c("dist","unit")) %>% mutate(tissue="FL") 

num=bind_rows(Mir_FB1,Mir_FL1,Mir_FB2,Mir_FL2) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("FB","FL"))) %>% count(tissue) #count dot number
stat.test <- bind_rows(Mir_FB1,Mir_FL1,Mir_FB2,Mir_FL2) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("FB","FL"))) %>%
  wilcox_test(dist ~ tissue, paired = F)
col_frac=bind_rows(Mir_FB1,Mir_FL1,Mir_FB2,Mir_FL2) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("FB","FL")),col=if_else(dist<0.25,"yes","no")) %>% dplyr::count(col,tissue) %>% group_by(tissue) %>% mutate(frac=formatC(n/sum(n)*100, format = "g", digits = 2)) %>% ungroup() %>% dplyr::filter(col=="yes")

fig4g_b <- bind_rows(Mir_FB1,Mir_FL1,Mir_FB2,Mir_FL2) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("FB","FL"))) %>%
  ggviolin(x="tissue",y="dist",fill="tissue", palette=c("#68abdaff","grey"), trim=T) + 
  geom_boxplot(fill="white", outlier.shape = NA,width=0.2) +
  stat_pvalue_manual(stat.test, label = "p",y.position=1.55,size=2)  + 
  annotate(geom = "text", x = col_frac$tissue, y = 0.15, label=paste0(col_frac$frac,"%"), size=2) + 
  scale_x_discrete(labels=c(paste0(num$tissue[1],"\n(n=",num$n[1],")"), paste0(num$tissue[2],"\n(n=",num$n[2],")"))) + 
  geom_hline(yintercept=0.25, color="red",linetype=2) + 
  labs(x="", y="inter-probe distance (µm)",title = "Mir9-2 and hs266") +  
  theme(legend.position = "None", plot.title = element_text(size = 8),
        axis.title.x = element_text(size=7),
        axis.title.y=element_text(size=7),axis.text = element_text(size=6))

fig4g_b

#ECDF (or Empirical cumulative distribution function) plot
fig4g_supb1 <- bind_rows(Mir_FB1,Mir_FL1,Mir_FB2,Mir_FL2) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("FB","FL"))) %>% 
  ggplot(aes(x = dist)) + stat_ecdf(aes(color = tissue),  geom = "step", size = 0.5) +
  scale_color_manual(values = c("#68abdaff","grey"))+
  labs(y = "cumulative frequency", x="E-P interprobe distance (µm)",title = "Mir9-2 and hs266") +
  theme(legend.key.size = unit(1, 'mm'),legend.title = element_text(size=7), 
        legend.text = element_text(size=6),legend.position = c(0.85,0.6),
        axis.title.x = element_text(size=7),
        axis.title.y=element_text(size=7),axis.text = element_text(size=6),
        plot.title = element_text(size=7))
fig4g_supb1

#histogram
data <- bind_rows(Mir_FB1,Mir_FL1,Mir_FB2,Mir_FL2) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("FB","FL"))) %>% count(tissue,gr=cut(dist, breaks= c(seq(0, 1, by = 0.25),1.5)))

stat.test <- data %>% group_by(tissue) %>% mutate(non=sum(n)-n) %>% unite("code",c("gr","tissue"),sep=":") %>% pivot_longer(!code, names_to = "cat", values_to = "count") %>% pivot_wider(names_from = "code", values_from = "count") %>% column_to_rownames("cat") %>% pairwise_prop_test(p.adjust.method = "fdr") %>% separate(group1, c("A", "B"),remove = F,sep=":") %>% separate(group2, c("C", "D"),remove = F,sep=":") %>% dplyr::filter(A==C) %>% arrange(A) %>% mutate(p.adjsig=if_else(p.adj<0.05,as.character(p.adj),p.adj.signif)) %>% mutate(xmin=row_number()-0.23, xmax=row_number()+0.23)

fig4g_supb2 <- data %>% group_by(tissue) %>% mutate(frac=n/sum(n)) %>% ggbarplot(x = "gr", y="frac", fill="tissue",position = position_dodge(0.9)) + scale_fill_manual(values = c("#68abdaff","grey")) + labs(x="E-P interprobe distance (µm)",y="Percent of colocalization") + stat_pvalue_manual(stat.test, label = "p.adjsig",y.position=0.55,tip.length = 0.02,size=2)  +
  scale_y_continuous(labels = scales::percent) +
  theme(legend.key.size = unit(1, 'mm'),legend.title = element_text(size=7), 
        legend.text = element_text(size=6),legend.position = c(0.85,0.6),
        axis.title.x = element_text(size=7),
        axis.title.y=element_text(size=7),axis.text = element_text(size=6),
        plot.title = element_text(size=7))
fig4g_supb2
```

snai2 and hs1431 are ~300kb apart


```{r}
Snai_CF1 <- read_csv("CF-snai2-cy3-hs1431-488-2-27-23-1.csv",skip=4,col_names = c("dist","unit")) %>% bind_rows(read_csv("CF-snai2-cy3-hs1431-488-2-27-23-2.csv",skip=4,col_names = c("dist","unit"))) %>% bind_rows(read_csv("CF-snai2-cy3-hs1431-488-2-27-23-3.csv",skip=4,col_names = c("dist","unit"))) %>% mutate(tissue="CF")  

Snai_FB1 <- read_csv("MB-snai2-cy3-hs1431-488-2-27-23-1.csv",skip=4,col_names = c("dist","unit")) %>% bind_rows(read_csv("MB-snai2-cy3-hs1431-488-2-27-23-2.csv",skip=4,col_names = c("dist","unit"))) %>% bind_rows(read_csv("MB-snai2-cy3-hs1431-488-2-27-23-3.csv",skip=4,col_names = c("dist","unit"))) %>% mutate(tissue="FB") 

#boxplot
num=bind_rows(Snai_CF1,Snai_FB1)  %>%  filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("CF","FB"))) %>% count(tissue) #count dot number
stat.test <- bind_rows(Snai_CF1,Snai_FB1)  %>%  filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("CF","FB"))) %>%
  wilcox_test(dist ~ tissue, paired = F)
col_frac=bind_rows(Snai_CF1,Snai_FB1) %>% filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("CF","FB")),col=if_else(dist<0.25,"yes","no")) %>% dplyr::count(col,tissue) %>% group_by(tissue) %>% mutate(frac=formatC(n/sum(n)*100, format = "g", digits = 2)) %>% ungroup() %>% dplyr::filter(col=="yes")

fig4g_c <- bind_rows(Snai_CF1,Snai_FB1)  %>%  filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("CF","FB"))) %>%
  ggviolin(x="tissue",y="dist",fill="tissue", palette=c("#68abdaff","grey"), trim=T) + 
  geom_boxplot(fill="white", outlier.shape = NA,width=0.2) +
  stat_pvalue_manual(stat.test, label = "p",y.position=1.55,size=2)  + 
  annotate(geom = "text", x = col_frac$tissue, y = 0.15, label=paste0(col_frac$frac,"%"), size=2) + 
  scale_x_discrete(labels=c(paste0(num$tissue[1],"\n(n=",num$n[1],")"), paste0(num$tissue[2],"\n(n=",num$n[2],")"))) + 
  geom_hline(yintercept=0.25, color="red",linetype=2) + 
  coord_cartesian(ylim= c(0,1.6)) +
  labs(x="", y="inter-probe distance (µm)",title = "Snai2 and hs1431") + 
  theme(legend.position = "None", plot.title = element_text(size = 8),
        axis.title.x = element_text(size=7),
        axis.title.y=element_text(size=7),axis.text = element_text(size=6))
fig4g_c

#CDF
fig4g_supc1 <- bind_rows(Snai_CF1,Snai_FB1)  %>%  filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("CF","FB"))) %>% 
  ggplot(aes(x = dist)) + stat_ecdf(aes(color = tissue),  geom = "step", size = 0.5) +
  scale_color_manual(values = c("#68abdaff","grey"))+
  labs(y = "cumulative frequency", x="E-P interprobe distance (µm)",title = "Snai2 and hs1431") +
  theme(legend.key.size = unit(1, 'mm'),legend.title = element_text(size=7), 
        legend.text = element_text(size=6),legend.position = c(0.85,0.6),
        axis.title.x = element_text(size=7),
        axis.title.y=element_text(size=7),axis.text = element_text(size=6),
        plot.title = element_text(size=7))
fig4g_supc1
#histogram
data <- bind_rows(Snai_CF1,Snai_FB1)  %>%  filter(dist<1.5) %>% mutate(tissue=factor(tissue,levels=c("CF","FB"))) %>% count(tissue,gr=cut(dist, breaks= c(seq(0, 1, by = 0.25),1.5)))
stat.test <- data %>% group_by(tissue) %>% mutate(non=sum(n)-n) %>% unite("code",c("gr","tissue"),sep=":") %>% pivot_longer(!code, names_to = "cat", values_to = "count") %>% pivot_wider(names_from = "code", values_from = "count") %>% column_to_rownames("cat") %>% pairwise_prop_test(p.adjust.method = "fdr") %>% separate(group1, c("A", "B"),remove = F,sep=":") %>% separate(group2, c("C", "D"),remove = F,sep=":") %>% dplyr::filter(A==C) %>% arrange(A) %>% mutate(p.adjsig=if_else(p.adj<0.05,as.character(p.adj),p.adj.signif)) %>% mutate(xmin=row_number()-0.23, xmax=row_number()+0.23)

fig4g_supc2 <- data %>% group_by(tissue) %>% mutate(frac=n/sum(n)) %>% ggbarplot(x = "gr", y="frac", fill="tissue",position = position_dodge(0.9)) + scale_fill_manual(values = c("#68abdaff","grey")) + labs(x="E-P interprobe distance (µm)",y="Percent of colocalization") + stat_pvalue_manual(stat.test, label = "p.adjsig",y.position=0.5,tip.length = 0.02,size=2)  +
  scale_y_continuous(labels = scales::percent) +
  theme(legend.key.size = unit(1, 'mm'),legend.title = element_text(size=7), 
        legend.text = element_text(size=6),legend.position = c(0.85,0.6),
        axis.title.x = element_text(size=7),
        axis.title.y=element_text(size=7),axis.text = element_text(size=6),
        plot.title = element_text(size=7))
fig4g_supc2

fig4g_sup <- plot_grid(fig4g_supa1,fig4g_supb1,fig4g_supc1,fig4g_supa2,fig4g_supb2,fig4g_supc2,labels = c("A","B","C","D","E","F"),label_size=12,nrow=2)
fig4g_sup

fig4g <- plot_grid(fig4g_a,fig4g_b,fig4g_c,nrow=1)
fig4g

```
